{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "import gc\n",
    "import anndata as ad\n",
    "import sys\n",
    "import traceback\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "from glob import glob\n",
    "from pathlib import Path\n",
    "import infercnvpy as cnv\n",
    "import genomic_features as gf\n",
    "import traceback\n",
    "from biomart import BiomartServer\n",
    "import mygene\n",
    "from scipy.stats import zscore\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc = sc.read_h5ad('../single_cell_int/adata_sc_int_outlier_genes.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = adata_sc.obs.batch_covariate.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.obsm['X_umap'] = adata_sc.obsm['X_umap_int']\n",
    "sc.pl.umap(adata_sc, color='Level_1', frameon=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset = sc.pp.subsample(adata_sc, fraction=0.1, copy=True)\n",
    "raw_counts = subset.layers['raw'].toarray()\n",
    "print(f\"Are raw counts integers? {np.all(raw_counts.astype(int) == raw_counts)}\")\n",
    "print(f\"Range of raw counts: {np.min(raw_counts)} to {np.max(raw_counts)}\")\n",
    "print(\"-\" * 50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.var_names = adata_sc.var.gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.combat(adata_sc, key='batch_covariate')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "try:\n",
    "    for dataset in datasets:\n",
    "        print(f'\\033[92mRunning on dataset {dataset}\\033[0m')\n",
    "    #     # Subset the data\n",
    "        adata_temp = adata_sc[adata_sc.obs['batch_covariate'] == dataset].copy()\n",
    "        print(f\"Filtering genes in dataset {dataset} with minimum 5 cells\")\n",
    "        sc.pp.filter_genes(adata_temp, min_cells=5)\n",
    "        # adata_temp.var_names = adata_temp.var.ensembl_id\n",
    "        malignant_cells = [\"Acinar Cell\", \"Ductal Cell\", \"Ductal Cell/Malignant\"]\n",
    "        adata_temp.obs[\"Reference\"] = np.where(adata_temp.obs.Level_1.isin(malignant_cells), 'Potentially Malignant', 'Reference')\n",
    "    #     # Normalize, log transform, and perform dimensionality reduction\n",
    "        # sc.pp.normalize_total(adata_sc, target_sum=1e4)\n",
    "        # sc.pp.log1p(adata_sc)\n",
    "        print(f\"Running PCA, neighbors, UMAP, and Leiden clustering for dataset {dataset}\")\n",
    "        sc.pp.pca(adata_temp, layer='raw')\n",
    "        sc.pp.neighbors(adata_temp)\n",
    "        sc.tl.umap(adata_temp)\n",
    "        sc.tl.leiden(adata_temp)\n",
    "        print(f'infering CNVs for {dataset}')\n",
    "    #     # Run inferCNV\n",
    "        cnv.tl.infercnv(\n",
    "            adata_temp,\n",
    "            reference_key=\"Reference\",\n",
    "            reference_cat=\"Reference\",\n",
    "            window_size=150,\n",
    "        )\n",
    "\n",
    "    #     # PCA and neighbor graph for CNV data\n",
    "        print(f\"Running PCA, neighbors, Leiden clustering for CNV data in dataset {dataset}\")\n",
    "        cnv.tl.pca(adata_temp)\n",
    "        cnv.pp.neighbors(adata_temp)\n",
    "        cnv.tl.leiden(adata_temp)\n",
    "\n",
    "    #     # Run UMAP and CNV scoring\n",
    "        print(f\"Running UMAP and CNV scoring for dataset {dataset}\")\n",
    "        cnv.tl.umap(adata_temp)\n",
    "        cnv.tl.cnv_score(adata_temp)\n",
    "\n",
    "    #     # Save results\n",
    "        print(f\"Saving results for dataset {dataset}\")\n",
    "        output_dir = f\"../inferCNV/{dataset}\"\n",
    "        os.makedirs(output_dir, exist_ok=True)\n",
    "        adata_temp.write(f\"{output_dir}/PDAC_{dataset}_inferCNV.h5ad\")\n",
    "        print('Saving Images')\n",
    "    #     # Save chromosome heatmap\n",
    "        cnv.pl.chromosome_heatmap(adata_temp, groupby=\"Level_1\", save=f\"{dataset}_chromosome_heatmap_labels_inferCNV.png\")\n",
    "        cnv.pl.chromosome_heatmap(adata_temp, groupby=\"cnv_leiden\", dendrogram=True, save=f\"{dataset}_chromosome_heatmap_cnvleiden_inferCNV.png\")\n",
    "\n",
    "    #     # Save combined UMAP plots\n",
    "        try:\n",
    "            fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(11, 11))\n",
    "            ax4.axis(\"off\")\n",
    "            np.random.seed(0)\n",
    "            random_indices = np.random.permutation(range(adata_temp.shape[0]))\n",
    "            cnv.pl.umap(adata_temp[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "            cnv.pl.umap(adata_temp[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "            cnv.pl.umap(adata_temp[random_indices,:], color=\"Level_1\", ax=ax3, size=5)\n",
    "            fig.savefig(f\"{dataset}_combined_umap.png\")\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while saving UMAP plots for {dataset}: {e}\")\n",
    "\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(12, 11), gridspec_kw={\"wspace\": 0.5})\n",
    "        ax4.axis(\"off\")\n",
    "        np.random.seed(0)\n",
    "        random_indices = np.random.permutation(list(range(adata_temp.shape[0])))\n",
    "        sc.pl.umap(adata_temp[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "        sc.pl.umap(adata_temp[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "        sc.pl.umap(adata_temp[random_indices,:], color=\"Level_1\", ax=ax3, size=5)\n",
    "        fig.savefig(f\"../inferCNV/{dataset}_combined_umap_transcriptomic.png\")\n",
    "        print(f'\\033[91mCompleted inferring CNVs for {dataset}\\033[0m')\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred in loop: {dataset}\")\n",
    "    traceback.print_exc()\n",
    "\n",
    "#TODO remove commented code if not necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.tl.infercnv(\n",
    "    adata_sc,\n",
    "    reference_key=\"batch_covariate\",\n",
    "    reference_cat=\"Peng_Normal_scRNA-seq\",\n",
    "    window_size=500, layer='raw'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.pp.neighbors(adata_sc, use_rep='scpoli')\n",
    "cnv.tl.leiden(adata_sc)\n",
    "cnv.tl.umap(adata_sc)\n",
    "cnv.tl.cnv_score(adata_sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"Level_1\") #save=\"chromosome_heatmap_labels_inferCNV.png\")\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"cnv_leiden\") #, dendrogram=True, save=\"chromosome_heatmap_cnvleiden_inferCNV.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"batch_covariate\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.write('../inferCNV/adata_sc_inferCNV.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(16, 11), gridspec_kw={\"wspace\": 0.75})\n",
    "ax4.axis(\"off\")\n",
    "np.random.seed(0)\n",
    "random_indices = np.random.permutation(range(adata_sc.shape[0]))\n",
    "cnv.pl.umap(adata_sc[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "cnv.pl.umap(adata_sc[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "cnv.pl.umap(adata_sc[random_indices,:], color=\"Level_1\", ax=ax3, size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(16, 11), gridspec_kw={\"wspace\": 0.75})\n",
    "ax4.axis(\"off\")\n",
    "np.random.seed(0)\n",
    "random_indices = np.random.permutation(list(range(adata_sc.shape[0])))\n",
    "sc.pl.umap(adata_sc[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "sc.pl.umap(adata_sc[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "sc.pl.umap(adata_sc[random_indices,:], color=\"Level_1\", ax=ax3, size=5)\n",
    "# fig.savefig(f\"../inferCNV/{dataset}_combined_umap_transcriptomic.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(16, 11), gridspec_kw={\"wspace\": 0.75})\n",
    "ax4.axis(\"off\")\n",
    "np.random.seed(0)\n",
    "random_indices = np.random.permutation(list(range(adata_sc.shape[0])))\n",
    "sc.pl.umap(adata_sc[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "sc.pl.umap(adata_sc[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "sc.pl.umap(adata_sc[random_indices,:], color=\"Level_1\", ax=ax3, size=5)\n",
    "# fig.savefig(f\"../inferCNV/{dataset}_combined_umap_transcriptomic.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "malignant_cells = [\"Ductal Cell\", \"Ductal Cell/Malignant\"]\n",
    "adata_sc.obs[\"Reference\"] = np.where(adata_sc.obs.Level_1.isin(malignant_cells), 'Potentially Malignant', 'Reference')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.tl.infercnv(\n",
    "    adata_sc,\n",
    "    reference_key=\"Reference\",\n",
    "    reference_cat=\"Reference\",\n",
    "    window_size=1000, layer='log_norm'\n",
    ")\n",
    "\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"Level_1\") #save=\"chromosome_heatmap_labels_inferCNV.png\")\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"batch_covariate\") #, dendrogram=True, save=\"chromosome_heatmap_cnvleiden_inferCNV.png\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.tl.infercnv(\n",
    "    adata_sc,\n",
    "    reference_key=\"Reference\",\n",
    "    reference_cat=\"Reference\",\n",
    "    window_size=50, layer='log_norm'\n",
    ")\n",
    "\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"Level_1\") #save=\"chromosome_heatmap_labels_inferCNV.png\")\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"batch_covariate\") #, dendrogram=True, save=\"chromosome_heatmap_cnvleiden_inferCNV.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.tl.infercnv(\n",
    "    adata_sc,\n",
    "    reference_key=\"batch_covariate\",\n",
    "    reference_cat=\"Peng_Normal_scRNA-seq\",\n",
    "    window_size=150, layer='log_norm'\n",
    ")\n",
    "\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"Level_1\") #save=\"chromosome_heatmap_labels_inferCNV.png\")\n",
    "cnv.pl.chromosome_heatmap(adata_sc, groupby=\"batch_covariate\") #, dendrogram=True, save=\"chromosome_heatmap_cnvleiden_inferCNV.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22",
   "metadata": {},
   "outputs": [],
   "source": [
    "cnv.pl.chromosome_heatmap(adata_sc[adata_sc.obs.Reference == 'Potentially Malignant'], groupby=\"Level_1\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "infercnv_dir = '../inferCNV/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25",
   "metadata": {},
   "outputs": [],
   "source": [
    "files = glob('../inferCNV/*/*.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "adatas = {}\n",
    "for i in files:\n",
    "    print(i)\n",
    "    adatas[i] = sc.read_h5ad(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in adatas:\n",
    "    print(i)\n",
    "    sc.pl.umap(adatas[i], color=['cnv_score', 'Level_1', 'cnv_leiden'], frameon=False, wspace=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_cnv_score = {}\n",
    "total_cells = []\n",
    "for i in adatas:\n",
    "    print(i)\n",
    "    total_cells.append(adatas[i].shape[0])\n",
    "    dict_cnv_score.update(dict(list(zip(adatas[i].obs.Dataset_Barcode, list(zip(adatas[i].obs['cnv_score'], adatas[i].obs.Dataset))))))\n",
    "print(sum(total_cells))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cnv_scores = pd.DataFrame.from_dict(dict_cnv_score, orient='index')\n",
    "\n",
    "df_cnv_scores['cnv_score_normalized'] = df_cnv_scores.groupby(1)[0].transform(zscore)\n",
    "\n",
    "adata_sc.obs['cnv_score_znormalized'] = adata_sc.obs.Dataset_Barcode.map(dict(list(zip(df_cnv_scores.index, df_cnv_scores.cnv_score_normalized))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_sc, color=['cnv_score_znormalized', 'Level_1'], frameon=False, vmax=5, size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in adatas:\n",
    "    print(i)\n",
    "    sc.pl.umap(adatas[i], color=['cnv_score', 'Level_1', 'cnv_leiden'], frameon=False, wspace=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32",
   "metadata": {},
   "outputs": [],
   "source": [
    "cutoff_values = {\n",
    "    '../inferCNV/Simeone_scRNA-seq/PDAC_Simeone_scRNA-seq_inferCNV.h5ad': 0.05,\n",
    "    '../inferCNV/Steele_scRNA-seq/PDAC_Steele_scRNA-seq_inferCNV.h5ad': 0.06,\n",
    "    '../inferCNV/Zhang_scRNA-seq/PDAC_Zhang_scRNA-seq_inferCNV.h5ad': 0.04,\n",
    "    '../inferCNV/Peng_Normal_scRNA-seq/PDAC_Peng_Normal_scRNA-seq_inferCNV.h5ad': 0.04,\n",
    "    '../inferCNV/Lin_scRNA-seq/PDAC_Lin_scRNA-seq_inferCNV.h5ad': 0.030,\n",
    "    '../inferCNV/Ding_scRNA-seq/PDAC_Ding_scRNA-seq_inferCNV.h5ad': 0.020,\n",
    "    '../inferCNV/Schlesinger_scRNA-seq/PDAC_Schlesinger_scRNA-seq_inferCNV.h5ad': 0.04,\n",
    "    '../inferCNV/Peng_scRNA-seq/PDAC_Peng_scRNA-seq_inferCNV.h5ad': 0.05,\n",
    "    '../inferCNV/Caronni_scRNA-seq/PDAC_Caronni_scRNA-seq_inferCNV.h5ad': 0.05,\n",
    "    '../inferCNV/Lee_scRNA-seq/PDAC_Lee_scRNA-seq_inferCNV.h5ad': 0.05,\n",
    "    '../inferCNV/Steele_Adj_Norm_scRNA-seq/PDAC_Steele_Adj_Norm_scRNA-seq_inferCNV.h5ad': 0.04\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_infercnv_score_malignant = {}\n",
    "for dataset_name, adata in adatas.items():\n",
    "    cutoff = cutoff_values[dataset_name]  \n",
    "    adata.obs['infercnv_score_malignant'] = adata.obs['cnv_score'] > cutoff\n",
    "    dict_infercnv_score_malignant.update(dict(list(zip(adata.obs.Dataset_Barcode, adata.obs['infercnv_score_malignant']))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.obs['infercnv_score_malignant'] = adata_sc.obs.Dataset_Barcode.map(dict_infercnv_score_malignant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.obs['infercnv_score_malignant'] = adata_sc.obs['infercnv_score_malignant'].map({True: 'Malignant', False: 'Non-Malignant'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_indices = np.random.permutation(list(range(adata_sc.shape[0])))\n",
    "sc.pl.umap(\n",
    "    adata_sc[random_indices, :], color=['infercnv_score_malignant', 'Level_1'], frameon=False, vmax=5, size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_sc[random_indices, :], color=['infercnv_score_malignant'], groups='Malignant', frameon=False, vmax=5, size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38",
   "metadata": {},
   "outputs": [],
   "source": [
    "malignant_cell_types = ['Acinar Cell', 'Ductal Cell', 'Ductal Cell/Malignant']\n",
    "adata_sc.obs['infercnv_score_malignant_refined'] = np.where(\n",
    "    (adata_sc.obs['infercnv_score_malignant'] == 'Malignant') & \n",
    "    (adata_sc.obs['Level_1'].isin(malignant_cell_types)),\n",
    "    'Malignant',\n",
    "    'Non-Malignant'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(\n",
    "    adata_sc[random_indices, :], color=['infercnv_score_malignant_refined'], groups='Malignant', frameon=False, vmax=5, size=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_infercnv_score = {}\n",
    "for dataset_name, adata in adatas.items():\n",
    "    dict_infercnv_score.update(dict(list(zip(adata.obs.Dataset_Barcode, adata.obs['cnv_score']))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.obs['cnv_score_abs'] = adata_sc.obs.Dataset_Barcode.map(dict_infercnv_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.write('../single_cell_int/adata_sc_int_cnv.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43",
   "metadata": {},
   "source": [
    "# look for single nuc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sn = sc.read_h5ad('../single_nuc_int/adata_nuc_int_outlier_genes.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45",
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = adata_sn.obs.Dataset.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sn.obs['Level_1'] = adata_sn.obs['scpoli_labels'].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_sn, color=['Level_1', 'Dataset'], frameon=False, wspace=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    for dataset in datasets:\n",
    "        print(f'\\033[92mRunning on dataset {dataset}\\033[0m')\n",
    "    #     # Subset the data\n",
    "        adata_temp = adata_sn[adata_sn.obs['Dataset'] == dataset].copy()\n",
    "        print(f\"Filtering genes in dataset {dataset} with minimum 5 cells\")\n",
    "        sc.pp.filter_genes(adata_temp, min_cells=5)\n",
    "        # adata_temp.var_names = adata_temp.var.ensembl_id\n",
    "        malignant_cells = [\"Acinar Cell\", \"Ductal Cell\", \"Malignant\"]\n",
    "        adata_temp.obs[\"Reference\"] = np.where(adata_temp.obs.Level_1.isin(malignant_cells), 'Potentially Malignant', 'Reference')\n",
    "    #     # Normalize, log transform, and perform dimensionality reduction\n",
    "        # sc.pp.normalize_total(adata_sc, target_sum=1e4)\n",
    "        # sc.pp.log1p(adata_sc)\n",
    "        print(f\"Running PCA, neighbors, UMAP, and Leiden clustering for dataset {dataset}\")\n",
    "        sc.pp.pca(adata_temp, layer='log_norm')\n",
    "        sc.pp.neighbors(adata_temp)\n",
    "        sc.tl.umap(adata_temp)\n",
    "        sc.tl.leiden(adata_temp)\n",
    "        print(f'infering CNVs for {dataset}')\n",
    "    #     # Run inferCNV\n",
    "        cnv.tl.infercnv(\n",
    "            adata_temp,\n",
    "            reference_key=\"Reference\",\n",
    "            reference_cat=\"Reference\",\n",
    "            window_size=150,\n",
    "        )\n",
    "\n",
    "    #     # PCA and neighbor graph for CNV data\n",
    "        print(f\"Running PCA, neighbors, Leiden clustering for CNV data in dataset {dataset}\")\n",
    "        cnv.tl.pca(adata_temp)\n",
    "        cnv.pp.neighbors(adata_temp)\n",
    "        cnv.tl.leiden(adata_temp)\n",
    "\n",
    "    #     # Run UMAP and CNV scoring\n",
    "        print(f\"Running UMAP and CNV scoring for dataset {dataset}\")\n",
    "        cnv.tl.umap(adata_temp)\n",
    "        cnv.tl.cnv_score(adata_temp)\n",
    "\n",
    "    #     # Save results\n",
    "        print(f\"Saving results for dataset {dataset}\")\n",
    "        output_dir = f\"../inferCNV/{dataset}\"\n",
    "        os.makedirs(output_dir, exist_ok=True)\n",
    "        adata_temp.write(f\"{output_dir}/PDAC_{dataset}_inferCNV.h5ad\")\n",
    "        print('Saving Images')\n",
    "    #     # Save chromosome heatmap\n",
    "        cnv.pl.chromosome_heatmap(adata_temp, groupby=\"Level_1\", save=f\"{dataset}_chromosome_heatmap_labels_inferCNV.png\")\n",
    "        cnv.pl.chromosome_heatmap(adata_temp, groupby=\"cnv_leiden\", dendrogram=True, save=f\"{dataset}_chromosome_heatmap_cnvleiden_inferCNV.png\")\n",
    "\n",
    "    #     # Save combined UMAP plots\n",
    "        try:\n",
    "            fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(11, 11))\n",
    "            ax4.axis(\"off\")\n",
    "            np.random.seed(0)\n",
    "            random_indices = np.random.permutation(range(adata_temp.shape[0]))\n",
    "            cnv.pl.umap(adata_temp[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "            cnv.pl.umap(adata_temp[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "            cnv.pl.umap(adata_temp[random_indices,:], color=\"Level_1\", ax=ax3, size=5)\n",
    "            fig.savefig(f\"{dataset}_combined_umap.png\")\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while saving UMAP plots for {dataset}: {e}\")\n",
    "\n",
    "        fig, ((ax1, ax2), (ax3, ax4)) = plt.subplots(2, 2, figsize=(12, 11), gridspec_kw={\"wspace\": 0.5})\n",
    "        ax4.axis(\"off\")\n",
    "        np.random.seed(0)\n",
    "        random_indices = np.random.permutation(list(range(adata_temp.shape[0])))\n",
    "        sc.pl.umap(adata_temp[random_indices,:], color=\"cnv_leiden\", ax=ax1, show=False, size=5)\n",
    "        sc.pl.umap(adata_temp[random_indices,:], color=\"cnv_score\", ax=ax2, show=False, size=5)\n",
    "        sc.pl.umap(adata_temp[random_indices,:], color=\"Level_1\", ax=ax3, size=5)\n",
    "        fig.savefig(f\"../inferCNV/{dataset}_combined_umap_transcriptomic.png\")\n",
    "        print(f'\\033[91mCompleted inferring CNVs for {dataset}\\033[0m')\n",
    "except Exception as e:\n",
    "    print(f\"An error occurred in loop: {dataset}\")\n",
    "    traceback.print_exc()\n",
    "\n",
    "#TODO remove commented code if not necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "inferCNV",
   "language": "python",
   "name": "infercnv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
