{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c296ed23-28e8-4cf2-a967-82da84c5e3f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9e9368c5-465e-4f88-85bd-488bf0e7b50d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ictstr01/groups/ml01/workspace/shrey.parikh/PDAC_Work_Dir/PDAC_Final/Downstream'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "228ff35b-0db2-4098-bdcb-e8f502f482b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "ddbe89b1-da3a-4c1f-964d-18871f8fc32d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id = pd.read_csv('ID_Harmonised.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "401dd57d-ce41-4ebd-9ef1-01c79af25096",
   "metadata": {},
   "source": [
    "# Ding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 734,
   "id": "224b6c42-a8c7-44a7-b345-8afa8751127b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding = pd.read_excel('metadata_publications/Ding_metadata.xlsx', sheet_name='Table2_Metadata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 735,
   "id": "9beaf907-7dd6-44e9-a9eb-637173d9d093",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset_Barcode</th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ding_AAACCCACAGAAGCTG-1-HT060P1-S1R1A1G1Z1_1B2...</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT060P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ding_AAACCCAAGTAACCGG-1-HT061P1-S1PAA1A1Z1_1B2...</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT061P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ding_AAAGGGCCAAGTCATC-1-HT064P1-S1H1A2A1Z1_1B2...</td>\n",
       "      <td>HT064P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT064P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ding_AAACGAATCTCTCTTC-1-HT071P1-XB1_1-1-0-0-0</td>\n",
       "      <td>HT071P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT071P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>Ding_AAACAGCCAATATAGG-1-HT270P1-S1H2Fc2A2N1Bmn...</td>\n",
       "      <td>HT270P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT270P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>Ding_AAACAGCCAGCTCATA-1-HT284P1-S1H2Fc2A2N1Z1_...</td>\n",
       "      <td>HT284P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT284P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>Ding_AAACAGCCAAACCCTA-1-HT288P1-S1H4Fc2A2N1Z1_...</td>\n",
       "      <td>HT288P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT288P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>Ding_AAACAGCCAATGCCCG-1-HT306P1-S1H1Fc2A2N1Z1_...</td>\n",
       "      <td>HT306P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT306P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>Ding_AAACAGCCACAACAAA-1-HT412P1-S1H1A3N1Z1_1Bm...</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Dataset_Barcode ID_harmonised  \\\n",
       "0        Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       HT056P1   \n",
       "1    Ding_AAACCCACAGAAGCTG-1-HT060P1-S1R1A1G1Z1_1B2...       HT060P1   \n",
       "2    Ding_AAACCCAAGTAACCGG-1-HT061P1-S1PAA1A1Z1_1B2...       HT061P1   \n",
       "3    Ding_AAAGGGCCAAGTCATC-1-HT064P1-S1H1A2A1Z1_1B2...       HT064P1   \n",
       "4        Ding_AAACGAATCTCTCTTC-1-HT071P1-XB1_1-1-0-0-0       HT071P1   \n",
       "..                                                 ...           ...   \n",
       "168  Ding_AAACAGCCAATATAGG-1-HT270P1-S1H2Fc2A2N1Bmn...       HT270P1   \n",
       "169  Ding_AAACAGCCAGCTCATA-1-HT284P1-S1H2Fc2A2N1Z1_...       HT284P1   \n",
       "170  Ding_AAACAGCCAAACCCTA-1-HT288P1-S1H4Fc2A2N1Z1_...       HT288P1   \n",
       "171  Ding_AAACAGCCAATGCCCG-1-HT306P1-S1H1Fc2A2N1Z1_...       HT306P1   \n",
       "172  Ding_AAACAGCCACAACAAA-1-HT412P1-S1H1A3N1Z1_1Bm...       HT412P1   \n",
       "\n",
       "            Dataset Unique_ID  \n",
       "0              Ding   HT056P1  \n",
       "1              Ding   HT060P1  \n",
       "2              Ding   HT061P1  \n",
       "3              Ding   HT064P1  \n",
       "4              Ding   HT071P1  \n",
       "..              ...       ...  \n",
       "168  Ding_snRNA-seq   HT270P1  \n",
       "169  Ding_snRNA-seq   HT284P1  \n",
       "170  Ding_snRNA-seq   HT288P1  \n",
       "171  Ding_snRNA-seq   HT306P1  \n",
       "172  Ding_snRNA-seq   HT412P1  \n",
       "\n",
       "[173 rows x 4 columns]"
      ]
     },
     "execution_count": 735,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 736,
   "id": "a961a7c1-fd53-4043-a423-e2affbeaeb53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample</th>\n",
       "      <th>Case</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>scRNA_available</th>\n",
       "      <th>WES_available</th>\n",
       "      <th>RNA-Seq_available</th>\n",
       "      <th>Proteome_available</th>\n",
       "      <th>Phosphoproteome_available</th>\n",
       "      <th>Hyperion_available</th>\n",
       "      <th>H&amp;E_available</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>KRAS_VAF</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>scRNA_tumor_percentage</th>\n",
       "      <th>Histology_tumor_percentage</th>\n",
       "      <th>ABSOLUTE_Cancer_DNA_fraction</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Stroma_Score</th>\n",
       "      <th>Immune_Score</th>\n",
       "      <th>ESTIMATE_Score</th>\n",
       "      <th>Immune_subtype_2k</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>HT200P1_S1H2</td>\n",
       "      <td>HT200P1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>0.230769</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>5.723704</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.25</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>1312.272569</td>\n",
       "      <td>1385.296908</td>\n",
       "      <td>2697.569478</td>\n",
       "      <td>High</td>\n",
       "      <td>Classical</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>HT200P1_S1H4</td>\n",
       "      <td>HT200P1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>12.137097</td>\n",
       "      <td>50.0</td>\n",
       "      <td>0.29</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>1819.804456</td>\n",
       "      <td>1172.983804</td>\n",
       "      <td>2992.788260</td>\n",
       "      <td>High</td>\n",
       "      <td>Classical</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>HT204P1_S1H2</td>\n",
       "      <td>HT204P1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.281407</td>\n",
       "      <td>17.5</td>\n",
       "      <td>0.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>919.460912</td>\n",
       "      <td>1132.259029</td>\n",
       "      <td>2051.719941</td>\n",
       "      <td>High</td>\n",
       "      <td>Classical</td>\n",
       "      <td>ADEX</td>\n",
       "      <td>Exocrine-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>HT204P1_S1H4</td>\n",
       "      <td>HT204P1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.158730</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1741.273344</td>\n",
       "      <td>1715.367115</td>\n",
       "      <td>3456.640459</td>\n",
       "      <td>High</td>\n",
       "      <td>Classical</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Exocrine-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>HT204P1_S1H5</td>\n",
       "      <td>HT204P1</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.869565</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1662.948871</td>\n",
       "      <td>1077.558971</td>\n",
       "      <td>2740.507842</td>\n",
       "      <td>Low</td>\n",
       "      <td>Classical</td>\n",
       "      <td>Squamous</td>\n",
       "      <td>Exocrine-like</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Sample     Case Tissue  scRNA_available  WES_available  \\\n",
       "68  HT200P1_S1H2  HT200P1  Tumor             True           True   \n",
       "69  HT200P1_S1H4  HT200P1  Tumor             True           True   \n",
       "70  HT204P1_S1H2  HT204P1  Tumor             True          False   \n",
       "71  HT204P1_S1H4  HT204P1  Tumor             True          False   \n",
       "72  HT204P1_S1H5  HT204P1  Tumor             True          False   \n",
       "\n",
       "    RNA-Seq_available  Proteome_available  Phosphoproteome_available  \\\n",
       "68               True               False                      False   \n",
       "69               True               False                      False   \n",
       "70               True               False                      False   \n",
       "71               True               False                      False   \n",
       "72               True               False                      False   \n",
       "\n",
       "    Hyperion_available  H&E_available   Treatment  KRAS_VAF KRAS_mutation  \\\n",
       "68               False           True  FOLFIRINOX  0.230769        p.G12V   \n",
       "69               False           True  FOLFIRINOX  0.150000        p.G12V   \n",
       "70               False           True  FOLFIRINOX       NaN           NaN   \n",
       "71               False           True  FOLFIRINOX       NaN           NaN   \n",
       "72               False           True  FOLFIRINOX       NaN           NaN   \n",
       "\n",
       "    scRNA_tumor_percentage  Histology_tumor_percentage  \\\n",
       "68                5.723704                        45.0   \n",
       "69               12.137097                        50.0   \n",
       "70                6.281407                        17.5   \n",
       "71               20.158730                        35.0   \n",
       "72               10.869565                        25.0   \n",
       "\n",
       "    ABSOLUTE_Cancer_DNA_fraction KRAS_type TP53_type SMAD4_type CDKN2A_type  \\\n",
       "68                          0.25  Missense  Nonsense   Missense    Missense   \n",
       "69                          0.29  Missense  Nonsense   Missense    Missense   \n",
       "70                          0.18       NaN       NaN        NaN         NaN   \n",
       "71                          0.71       NaN       NaN        NaN         NaN   \n",
       "72                          0.20       NaN       NaN        NaN         NaN   \n",
       "\n",
       "   Germline_ATM Germline_BRCA2 Germline_FANCC  Stroma_Score  Immune_Score  \\\n",
       "68    Wild_Type      Wild_Type      Wild_Type   1312.272569   1385.296908   \n",
       "69    Wild_Type      Wild_Type      Wild_Type   1819.804456   1172.983804   \n",
       "70          NaN            NaN            NaN    919.460912   1132.259029   \n",
       "71          NaN            NaN            NaN   1741.273344   1715.367115   \n",
       "72          NaN            NaN            NaN   1662.948871   1077.558971   \n",
       "\n",
       "    ESTIMATE_Score Immune_subtype_2k Moffitt_Subtype Bailey_Subtype  \\\n",
       "68     2697.569478              High       Classical       Squamous   \n",
       "69     2992.788260              High       Classical       Squamous   \n",
       "70     2051.719941              High       Classical           ADEX   \n",
       "71     3456.640459              High       Classical       Squamous   \n",
       "72     2740.507842               Low       Classical       Squamous   \n",
       "\n",
       "   Collisson_Subtype  \n",
       "68         Classical  \n",
       "69         Classical  \n",
       "70     Exocrine-like  \n",
       "71     Exocrine-like  \n",
       "72     Exocrine-like  "
      ]
     },
     "execution_count": 736,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 737,
   "id": "d5c95503-9a32-4fa7-a57d-dc8da27c2d31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sample', 'Case', 'Tissue', 'scRNA_available', 'WES_available',\n",
       "       'RNA-Seq_available', 'Proteome_available', 'Phosphoproteome_available',\n",
       "       'Hyperion_available', 'H&E_available', 'Treatment', 'KRAS_VAF',\n",
       "       'KRAS_mutation', 'scRNA_tumor_percentage', 'Histology_tumor_percentage',\n",
       "       'ABSOLUTE_Cancer_DNA_fraction', 'KRAS_type', 'TP53_type', 'SMAD4_type',\n",
       "       'CDKN2A_type', 'Germline_ATM', 'Germline_BRCA2', 'Germline_FANCC',\n",
       "       'Stroma_Score', 'Immune_Score', 'ESTIMATE_Score', 'Immune_subtype_2k',\n",
       "       'Moffitt_Subtype', 'Bailey_Subtype', 'Collisson_Subtype'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 737,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 738,
   "id": "404bcf16-e448-40b9-b543-d808640342a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_dedup = ding[ding.Case.isin(df_id.ID_harmonised.tolist())].drop_duplicates(subset='Case', keep='first')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 739,
   "id": "9929fa88-f0fa-4c20-b6c5-75ebee816271",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Case', 'KRAS_mutation', 'KRAS_type', 'TP53_type', 'SMAD4_type',\n",
    "       'CDKN2A_type', 'Germline_ATM', 'Germline_BRCA2', 'Germline_FANCC', 'Moffitt_Subtype', 'Bailey_Subtype', 'Collisson_Subtype']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 740,
   "id": "9746353e-725a-41b5-8f1b-779e552650ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged = df_id.merge(ding_dedup[interesing_cols], left_on='ID_harmonised', right_on='Case', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 741,
   "id": "6b2da3fc-07e9-4399-9ed4-e26980916e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged.drop('Dataset_Barcode', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 742,
   "id": "b78f6194-74e4-4cfc-b4c4-11ffa78ed268",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Classical</td>\n",
       "      <td>ADEX</td>\n",
       "      <td>Exocrine-like</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HT061P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HT064P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT064P1</td>\n",
       "      <td>HT064P1</td>\n",
       "      <td>p.G12R</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Pancreatic_progenitor</td>\n",
       "      <td>Classical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HT071P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT071P1</td>\n",
       "      <td>HT071P1</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>HT270P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT270P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>HT284P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT284P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>HT288P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT288P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>HT306P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT306P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID_harmonised         Dataset Unique_ID     Case KRAS_mutation KRAS_type  \\\n",
       "0         HT056P1            Ding   HT056P1  HT056P1        p.G12V  Missense   \n",
       "1         HT060P1            Ding   HT060P1  HT060P1        p.G12V  Missense   \n",
       "2         HT061P1            Ding   HT061P1  HT061P1        p.G12D  Missense   \n",
       "3         HT064P1            Ding   HT064P1  HT064P1        p.G12R  Missense   \n",
       "4         HT071P1            Ding   HT071P1  HT071P1        p.G12D  Missense   \n",
       "..            ...             ...       ...      ...           ...       ...   \n",
       "168       HT270P1  Ding_snRNA-seq   HT270P1      NaN           NaN       NaN   \n",
       "169       HT284P1  Ding_snRNA-seq   HT284P1      NaN           NaN       NaN   \n",
       "170       HT288P1  Ding_snRNA-seq   HT288P1      NaN           NaN       NaN   \n",
       "171       HT306P1  Ding_snRNA-seq   HT306P1      NaN           NaN       NaN   \n",
       "172       HT412P1  Ding_snRNA-seq   HT412P1      NaN           NaN       NaN   \n",
       "\n",
       "     TP53_type SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2  \\\n",
       "0     Missense  Wild_Type   Wild_Type    Wild_Type      Wild_Type   \n",
       "1    Wild_Type  Wild_Type   Wild_Type     Nonsense       Nonsense   \n",
       "2     Nonsense  Wild_Type   Wild_Type    Wild_Type      Wild_Type   \n",
       "3    Wild_Type  Wild_Type   Wild_Type    Wild_Type      Wild_Type   \n",
       "4     Missense   Nonsense   Wild_Type    Wild_Type      Wild_Type   \n",
       "..         ...        ...         ...          ...            ...   \n",
       "168        NaN        NaN         NaN          NaN            NaN   \n",
       "169        NaN        NaN         NaN          NaN            NaN   \n",
       "170        NaN        NaN         NaN          NaN            NaN   \n",
       "171        NaN        NaN         NaN          NaN            NaN   \n",
       "172        NaN        NaN         NaN          NaN            NaN   \n",
       "\n",
       "    Germline_FANCC Moffitt_Subtype         Bailey_Subtype Collisson_Subtype  \n",
       "0        Wild_Type      Basal-like            Immunogenic  Quasimesenchymal  \n",
       "1        Wild_Type       Classical                   ADEX     Exocrine-like  \n",
       "2        Wild_Type             NaN                    NaN               NaN  \n",
       "3        Wild_Type      Basal-like  Pancreatic_progenitor         Classical  \n",
       "4        Wild_Type      Basal-like            Immunogenic  Quasimesenchymal  \n",
       "..             ...             ...                    ...               ...  \n",
       "168            NaN             NaN                    NaN               NaN  \n",
       "169            NaN             NaN                    NaN               NaN  \n",
       "170            NaN             NaN                    NaN               NaN  \n",
       "171            NaN             NaN                    NaN               NaN  \n",
       "172            NaN             NaN                    NaN               NaN  \n",
       "\n",
       "[173 rows x 15 columns]"
      ]
     },
     "execution_count": 742,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 743,
   "id": "0d8fe1e9-2fea-4ce4-a8d8-ec71a1ec70ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_2 = pd.read_excel('metadata_publications/Ding_metadata.xlsx', sheet_name='Table1_Clinical_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 744,
   "id": "767da0e0-fb6b-4311-805c-799ae6261e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_2_dedup = ding_2[ding_2.Patient_ID.isin(df_id.ID_harmonised.tolist())].drop_duplicates(subset='Patient_ID', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 745,
   "id": "cba3fe9c-b8c7-444c-809a-83583d967f4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Treatment_notes</th>\n",
       "      <th>Cancer_type</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Tumor_viability_%</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Family_history_HPB_first_degree</th>\n",
       "      <th>Other_family_history_first_degree</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Other_Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>48</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>6</td>\n",
       "      <td>Last dose 7 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>60.0</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>61</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>6</td>\n",
       "      <td>Last dose 4 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>70.0</td>\n",
       "      <td>ypT2, extends to duodenum and vascular groove</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>IIb</td>\n",
       "      <td>Type II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>breast cancer, diabetes-sister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HT061P1</td>\n",
       "      <td>74</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>male</td>\n",
       "      <td>12</td>\n",
       "      <td>Naive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2, poorly differentiated tumor arising in IPMN</td>\n",
       "      <td>3/7, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Multiple myeloma (remission)</td>\n",
       "      <td>A-fib, chronic kidney disease</td>\n",
       "      <td>Former</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HT064P1</td>\n",
       "      <td>66</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>male</td>\n",
       "      <td>16</td>\n",
       "      <td>Naive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2, invasive carcinoma intermixed with IPMN</td>\n",
       "      <td>19/32, pN2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>breast cancer-sister, colon cancer-mother, lun...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HT071P1</td>\n",
       "      <td>68</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>16</td>\n",
       "      <td>Chemo-RT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5-FU and radiation 11 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>70.0</td>\n",
       "      <td>ypT2, poorly differentiated</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Breast cancer (chemo 13 years prior)</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>Hodgkin's lymphoma-fater, neuroblastoma-brother</td>\n",
       "      <td>Deceased</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient_ID  Age   Race     Ethnicity     Sex  \\\n",
       "0    HT056P1   48  white  non-hispanic  female   \n",
       "1    HT060P1   61  white  non-hispanic  female   \n",
       "2    HT061P1   74  white  non-hispanic    male   \n",
       "3    HT064P1   66  white  non-hispanic    male   \n",
       "4    HT071P1   68  white  non-hispanic  female   \n",
       "\n",
       "   Time_from_diagnosis_to_surgery_(weeks)   Treatment Cycles  \\\n",
       "0                                      20  FOLFIRINOX      6   \n",
       "1                                      20  FOLFIRINOX      6   \n",
       "2                                      12       Naive    NaN   \n",
       "3                                      16       Naive    NaN   \n",
       "4                                      16    Chemo-RT    NaN   \n",
       "\n",
       "                                Treatment_notes Cancer_type  \\\n",
       "0            Last dose 7 weeks prior to surgery        PDAC   \n",
       "1            Last dose 4 weeks prior to surgery        PDAC   \n",
       "2                                           NaN        PDAC   \n",
       "3                                           NaN        PDAC   \n",
       "4  5-FU and radiation 11 weeks prior to surgery        PDAC   \n",
       "\n",
       "                 Location  Grade  Size_(cm)  Tumor_viability_%  \\\n",
       "0  head, uncinate_process    3.0        3.2               60.0   \n",
       "1  head, uncinate_process    3.0        3.2               70.0   \n",
       "2                    head    3.0        3.5                NaN   \n",
       "3                    head    1.0        3.2                NaN   \n",
       "4                    head    3.0        3.4               70.0   \n",
       "\n",
       "                                    Pathology_notes Lymph_node_met Stage  \\\n",
       "0            ypT2, focal sarcomatoid transformation      2/21, pN1   NaN   \n",
       "1     ypT2, extends to duodenum and vascular groove      2/25, pN1   IIb   \n",
       "2  pT2, poorly differentiated tumor arising in IPMN       3/7, pN1   NaN   \n",
       "3      pT2, invasive carcinoma intermixed with IPMN     19/32, pN2   NaN   \n",
       "4                       ypT2, poorly differentiated      2/25, pN1   NaN   \n",
       "\n",
       "  Diabetes Pancreatitis               Previous_cancer_history  \\\n",
       "0       No          Yes                                    No   \n",
       "1  Type II          NaN                                   NaN   \n",
       "2       No          NaN          Multiple myeloma (remission)   \n",
       "3       No          Yes                                    No   \n",
       "4       No          NaN  Breast cancer (chemo 13 years prior)   \n",
       "\n",
       "            Other_co-morbidities  Smoker     Alcohol  \\\n",
       "0                             No      No  Occasional   \n",
       "1                             No      No  Occasional   \n",
       "2  A-fib, chronic kidney disease  Former          No   \n",
       "3                             No      No          No   \n",
       "4                             No      No  Occasional   \n",
       "\n",
       "  Family_history_HPB_first_degree  \\\n",
       "0                              No   \n",
       "1                              No   \n",
       "2                              No   \n",
       "3                              No   \n",
       "4                              No   \n",
       "\n",
       "                   Other_family_history_first_degree Vital_status  \\\n",
       "0                                                 No          NaN   \n",
       "1                     breast cancer, diabetes-sister          NaN   \n",
       "2                                                NaN          NaN   \n",
       "3  breast cancer-sister, colon cancer-mother, lun...          NaN   \n",
       "4    Hodgkin's lymphoma-fater, neuroblastoma-brother     Deceased   \n",
       "\n",
       "   Survival_post_surgery_(weeks) Other_Notes  \n",
       "0                            NaN         NaN  \n",
       "1                            NaN         NaN  \n",
       "2                            NaN         NaN  \n",
       "3                            NaN         NaN  \n",
       "4                            5.0         NaN  "
      ]
     },
     "execution_count": 745,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding_2_dedup.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 746,
   "id": "5cc563de-c008-4720-9b03-3a85644df2e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Patient_ID', 'Age', 'Race', 'Ethnicity', 'Sex',\n",
       "       'Time_from_diagnosis_to_surgery_(weeks)', 'Treatment', 'Cycles',\n",
       "       'Treatment_notes', 'Cancer_type', 'Location', 'Grade', 'Size_(cm)',\n",
       "       'Tumor_viability_%', 'Pathology_notes', 'Lymph_node_met', 'Stage',\n",
       "       'Diabetes', 'Pancreatitis', 'Previous_cancer_history',\n",
       "       'Other_co-morbidities', 'Smoker', 'Alcohol',\n",
       "       'Family_history_HPB_first_degree', 'Other_family_history_first_degree',\n",
       "       'Vital_status', 'Survival_post_surgery_(weeks)', 'Other_Notes'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 746,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding_2_dedup.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 747,
   "id": "711fda96-117e-4966-8cad-c46db881dc69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract the T stage and N stage\n",
    "def extract_tnm(row):\n",
    "    # Extract T stage from Pathology_notes\n",
    "    t_stage = row['Pathology_notes'].split(',')[0]  # Get the first part before the comma\n",
    "    # Extract N stage from Lymph_node_met\n",
    "    n_stage = row['Lymph_node_met'].split(',')[1].strip()  # Get the part after the comma\n",
    "    return f\"{t_stage} {n_stage}\"\n",
    "\n",
    "# Create the new TNM classification column\n",
    "ding_2_dedup['TNM_classification'] = ding_2_dedup.apply(extract_tnm, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 748,
   "id": "90088f53-9653-412f-a226-31612aef84d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Patient_ID', 'Time_from_diagnosis_to_surgery_(weeks)', 'Cycles', 'Location', 'Grade', 'Size_(cm)', 'Stage','Diabetes', \n",
    "                   'Pancreatitis', 'Previous_cancer_history',  'Other_co-morbidities', 'Smoker', 'Alcohol', 'Pathology_notes', 'Lymph_node_met',\n",
    "                   'Survival_post_surgery_(weeks)', 'Vital_status', 'TNM_classification']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 749,
   "id": "c67d2e2b-ebef-4431-ae26-be8850130f30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Treatment_notes</th>\n",
       "      <th>Cancer_type</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Tumor_viability_%</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Family_history_HPB_first_degree</th>\n",
       "      <th>Other_family_history_first_degree</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Other_Notes</th>\n",
       "      <th>TNM_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>48</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>6</td>\n",
       "      <td>Last dose 7 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>60.0</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>61</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>6</td>\n",
       "      <td>Last dose 4 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>70.0</td>\n",
       "      <td>ypT2, extends to duodenum and vascular groove</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>IIb</td>\n",
       "      <td>Type II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>breast cancer, diabetes-sister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient_ID  Age   Race     Ethnicity     Sex  \\\n",
       "0    HT056P1   48  white  non-hispanic  female   \n",
       "1    HT060P1   61  white  non-hispanic  female   \n",
       "\n",
       "   Time_from_diagnosis_to_surgery_(weeks)   Treatment Cycles  \\\n",
       "0                                      20  FOLFIRINOX      6   \n",
       "1                                      20  FOLFIRINOX      6   \n",
       "\n",
       "                      Treatment_notes Cancer_type                Location  \\\n",
       "0  Last dose 7 weeks prior to surgery        PDAC  head, uncinate_process   \n",
       "1  Last dose 4 weeks prior to surgery        PDAC  head, uncinate_process   \n",
       "\n",
       "   Grade  Size_(cm)  Tumor_viability_%  \\\n",
       "0    3.0        3.2               60.0   \n",
       "1    3.0        3.2               70.0   \n",
       "\n",
       "                                 Pathology_notes Lymph_node_met Stage  \\\n",
       "0         ypT2, focal sarcomatoid transformation      2/21, pN1   NaN   \n",
       "1  ypT2, extends to duodenum and vascular groove      2/25, pN1   IIb   \n",
       "\n",
       "  Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities Smoker  \\\n",
       "0       No          Yes                      No                   No     No   \n",
       "1  Type II          NaN                     NaN                   No     No   \n",
       "\n",
       "      Alcohol Family_history_HPB_first_degree  \\\n",
       "0  Occasional                              No   \n",
       "1  Occasional                              No   \n",
       "\n",
       "  Other_family_history_first_degree Vital_status  \\\n",
       "0                                No          NaN   \n",
       "1    breast cancer, diabetes-sister          NaN   \n",
       "\n",
       "   Survival_post_surgery_(weeks) Other_Notes TNM_classification  \n",
       "0                            NaN         NaN           ypT2 pN1  \n",
       "1                            NaN         NaN           ypT2 pN1  "
      ]
     },
     "execution_count": 749,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding_2_dedup.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 750,
   "id": "3d23e86c-e909-4246-8b89-ae60d456d6de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged = df_id_merged.merge(ding_2_dedup[interesing_cols], left_on='ID_harmonised', right_on='Patient_ID', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 751,
   "id": "0a6ad194-6c51-4f6b-9f6b-d9c57ab54acd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Classical</td>\n",
       "      <td>ADEX</td>\n",
       "      <td>Exocrine-like</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>IIb</td>\n",
       "      <td>Type II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, extends to duodenum and vascular groove</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HT061P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Multiple myeloma (remission)</td>\n",
       "      <td>A-fib, chronic kidney disease</td>\n",
       "      <td>Former</td>\n",
       "      <td>No</td>\n",
       "      <td>pT2, poorly differentiated tumor arising in IPMN</td>\n",
       "      <td>3/7, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HT064P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT064P1</td>\n",
       "      <td>HT064P1</td>\n",
       "      <td>p.G12R</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Pancreatic_progenitor</td>\n",
       "      <td>Classical</td>\n",
       "      <td>HT064P1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>head</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>pT2, invasive carcinoma intermixed with IPMN</td>\n",
       "      <td>19/32, pN2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2 pN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HT071P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT071P1</td>\n",
       "      <td>HT071P1</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>HT071P1</td>\n",
       "      <td>16.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Breast cancer (chemo 13 years prior)</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, poorly differentiated</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>5.0</td>\n",
       "      <td>Deceased</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ID_harmonised Dataset Unique_ID     Case KRAS_mutation KRAS_type  TP53_type  \\\n",
       "0       HT056P1    Ding   HT056P1  HT056P1        p.G12V  Missense   Missense   \n",
       "1       HT060P1    Ding   HT060P1  HT060P1        p.G12V  Missense  Wild_Type   \n",
       "2       HT061P1    Ding   HT061P1  HT061P1        p.G12D  Missense   Nonsense   \n",
       "3       HT064P1    Ding   HT064P1  HT064P1        p.G12R  Missense  Wild_Type   \n",
       "4       HT071P1    Ding   HT071P1  HT071P1        p.G12D  Missense   Missense   \n",
       "\n",
       "  SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "0  Wild_Type   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "1  Wild_Type   Wild_Type     Nonsense       Nonsense      Wild_Type   \n",
       "2  Wild_Type   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "3  Wild_Type   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "4   Nonsense   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "\n",
       "  Moffitt_Subtype         Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "0      Basal-like            Immunogenic  Quasimesenchymal    HT056P1   \n",
       "1       Classical                   ADEX     Exocrine-like    HT060P1   \n",
       "2             NaN                    NaN               NaN    HT061P1   \n",
       "3      Basal-like  Pancreatic_progenitor         Classical    HT064P1   \n",
       "4      Basal-like            Immunogenic  Quasimesenchymal    HT071P1   \n",
       "\n",
       "   Time_from_diagnosis_to_surgery_(weeks) Cycles                Location  \\\n",
       "0                                    20.0      6  head, uncinate_process   \n",
       "1                                    20.0      6  head, uncinate_process   \n",
       "2                                    12.0    NaN                    head   \n",
       "3                                    16.0    NaN                    head   \n",
       "4                                    16.0    NaN                    head   \n",
       "\n",
       "   Grade  Size_(cm) Stage Diabetes Pancreatitis  \\\n",
       "0    3.0        3.2   NaN       No          Yes   \n",
       "1    3.0        3.2   IIb  Type II          NaN   \n",
       "2    3.0        3.5   NaN       No          NaN   \n",
       "3    1.0        3.2   NaN       No          Yes   \n",
       "4    3.0        3.4   NaN       No          NaN   \n",
       "\n",
       "                Previous_cancer_history           Other_co-morbidities  \\\n",
       "0                                    No                             No   \n",
       "1                                   NaN                             No   \n",
       "2          Multiple myeloma (remission)  A-fib, chronic kidney disease   \n",
       "3                                    No                             No   \n",
       "4  Breast cancer (chemo 13 years prior)                             No   \n",
       "\n",
       "   Smoker     Alcohol                                   Pathology_notes  \\\n",
       "0      No  Occasional            ypT2, focal sarcomatoid transformation   \n",
       "1      No  Occasional     ypT2, extends to duodenum and vascular groove   \n",
       "2  Former          No  pT2, poorly differentiated tumor arising in IPMN   \n",
       "3      No          No      pT2, invasive carcinoma intermixed with IPMN   \n",
       "4      No  Occasional                       ypT2, poorly differentiated   \n",
       "\n",
       "  Lymph_node_met  Survival_post_surgery_(weeks) Vital_status  \\\n",
       "0      2/21, pN1                            NaN          NaN   \n",
       "1      2/25, pN1                            NaN          NaN   \n",
       "2       3/7, pN1                            NaN          NaN   \n",
       "3     19/32, pN2                            NaN          NaN   \n",
       "4      2/25, pN1                            5.0     Deceased   \n",
       "\n",
       "  TNM_classification  \n",
       "0           ypT2 pN1  \n",
       "1           ypT2 pN1  \n",
       "2            pT2 pN1  \n",
       "3            pT2 pN2  \n",
       "4           ypT2 pN1  "
      ]
     },
     "execution_count": 751,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "id": "9d0ce77b-0579-457f-bcfe-ac79c484ac5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(173, 33)"
      ]
     },
     "execution_count": 608,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e860d24f-20b7-42a1-9e79-6d7b3d8f7c51",
   "metadata": {},
   "source": [
    "# Hwang"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 752,
   "id": "2b46b134-58bd-40e8-856e-ae13c60685e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang = pd.read_excel('metadata_publications/Hwang_metadata.xlsx', sheet_name='Supplementary Table 1', header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 753,
   "id": "94168499-d3e5-4ad5-93d4-7f94b289a566",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Age (decade)</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Stage/Grade</th>\n",
       "      <th>Margin</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Neoadjuvant</th>\n",
       "      <th>Status Last FUP</th>\n",
       "      <th>PFS (d)</th>\n",
       "      <th>OS (d)</th>\n",
       "      <th>Storage (d)</th>\n",
       "      <th>10x Chemistry</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Untreated (U)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PDAC_U_1</td>\n",
       "      <td>20s</td>\n",
       "      <td>F</td>\n",
       "      <td>T3N1M0/g2-3</td>\n",
       "      <td>R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DWD</td>\n",
       "      <td>70</td>\n",
       "      <td>919.0</td>\n",
       "      <td>638.0</td>\n",
       "      <td>v3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PDAC_U_2</td>\n",
       "      <td>60s</td>\n",
       "      <td>F</td>\n",
       "      <td>T3N2M0/g2</td>\n",
       "      <td>R1</td>\n",
       "      <td>AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DWD</td>\n",
       "      <td>78</td>\n",
       "      <td>297.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>v3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PDAC_U_3</td>\n",
       "      <td>60s</td>\n",
       "      <td>F</td>\n",
       "      <td>T3N0M0/gX</td>\n",
       "      <td>R0</td>\n",
       "      <td>FG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MET</td>\n",
       "      <td>578</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>341.0</td>\n",
       "      <td>v3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PDAC_U_4</td>\n",
       "      <td>60s</td>\n",
       "      <td>M</td>\n",
       "      <td>T3N1M0/g3</td>\n",
       "      <td>R0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DWD</td>\n",
       "      <td>267</td>\n",
       "      <td>538.0</td>\n",
       "      <td>634.0</td>\n",
       "      <td>v2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID Age (decade)  Sex  Stage/Grade Margin Histology Neoadjuvant  \\\n",
       "0  Untreated (U)          NaN  NaN          NaN    NaN       NaN         NaN   \n",
       "1       PDAC_U_1          20s    F  T3N1M0/g2-3     R1       NaN         NaN   \n",
       "2       PDAC_U_2          60s    F    T3N2M0/g2     R1        AS         NaN   \n",
       "3       PDAC_U_3          60s    F    T3N0M0/gX     R0        FG         NaN   \n",
       "4       PDAC_U_4          60s    M    T3N1M0/g3     R0       NaN         NaN   \n",
       "\n",
       "  Status Last FUP PFS (d)  OS (d)  Storage (d) 10x Chemistry  \n",
       "0             NaN     NaN     NaN          NaN           NaN  \n",
       "1             DWD      70   919.0        638.0            v3  \n",
       "2             DWD      78   297.0         17.0            v3  \n",
       "3             MET     578  1082.0        341.0            v3  \n",
       "4             DWD     267   538.0        634.0            v2  "
      ]
     },
     "execution_count": 753,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hwang.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 754,
   "id": "785adb01-ee3f-4f2c-a2fc-c8f148831424",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang['Vital_status'] = hwang['Status Last FUP'].map({\n",
    "    'DWD': 'Deceased - Disease',\n",
    "    'DWOD': 'Deceased - Other',\n",
    "    'MET': 'Alive - Metastatic',\n",
    "    'NED': 'Alive - Disease-free'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 755,
   "id": "b5d1bc34-2b5c-4c04-9e1c-8e0e002ee328",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang[['TNM_classification', 'Grade']] = hwang['Stage/Grade'].str.split('/', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 756,
   "id": "7e92e471-2f31-4da7-b105-3e0d58a3bf97",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang.rename(columns={'PFS (d)': 'Progression-Free Survival', 'OS (d)': 'Overall Survival'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 757,
   "id": "bcaab6c9-6100-4e9b-9489-a88d3aafc591",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Age (decade)</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Stage/Grade</th>\n",
       "      <th>Margin</th>\n",
       "      <th>Histology</th>\n",
       "      <th>Neoadjuvant</th>\n",
       "      <th>Status Last FUP</th>\n",
       "      <th>Progression-Free Survival</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Storage (d)</th>\n",
       "      <th>10x Chemistry</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Grade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Untreated (U)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PDAC_U_1</td>\n",
       "      <td>20s</td>\n",
       "      <td>F</td>\n",
       "      <td>T3N1M0/g2-3</td>\n",
       "      <td>R1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DWD</td>\n",
       "      <td>70</td>\n",
       "      <td>919.0</td>\n",
       "      <td>638.0</td>\n",
       "      <td>v3</td>\n",
       "      <td>Deceased - Disease</td>\n",
       "      <td>T3N1M0</td>\n",
       "      <td>g2-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PDAC_U_2</td>\n",
       "      <td>60s</td>\n",
       "      <td>F</td>\n",
       "      <td>T3N2M0/g2</td>\n",
       "      <td>R1</td>\n",
       "      <td>AS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DWD</td>\n",
       "      <td>78</td>\n",
       "      <td>297.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>v3</td>\n",
       "      <td>Deceased - Disease</td>\n",
       "      <td>T3N2M0</td>\n",
       "      <td>g2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PDAC_U_3</td>\n",
       "      <td>60s</td>\n",
       "      <td>F</td>\n",
       "      <td>T3N0M0/gX</td>\n",
       "      <td>R0</td>\n",
       "      <td>FG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MET</td>\n",
       "      <td>578</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>341.0</td>\n",
       "      <td>v3</td>\n",
       "      <td>Alive - Metastatic</td>\n",
       "      <td>T3N0M0</td>\n",
       "      <td>gX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PDAC_U_4</td>\n",
       "      <td>60s</td>\n",
       "      <td>M</td>\n",
       "      <td>T3N1M0/g3</td>\n",
       "      <td>R0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DWD</td>\n",
       "      <td>267</td>\n",
       "      <td>538.0</td>\n",
       "      <td>634.0</td>\n",
       "      <td>v2</td>\n",
       "      <td>Deceased - Disease</td>\n",
       "      <td>T3N1M0</td>\n",
       "      <td>g3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID Age (decade)  Sex  Stage/Grade Margin Histology Neoadjuvant  \\\n",
       "0  Untreated (U)          NaN  NaN          NaN    NaN       NaN         NaN   \n",
       "1       PDAC_U_1          20s    F  T3N1M0/g2-3     R1       NaN         NaN   \n",
       "2       PDAC_U_2          60s    F    T3N2M0/g2     R1        AS         NaN   \n",
       "3       PDAC_U_3          60s    F    T3N0M0/gX     R0        FG         NaN   \n",
       "4       PDAC_U_4          60s    M    T3N1M0/g3     R0       NaN         NaN   \n",
       "\n",
       "  Status Last FUP Progression-Free Survival  Overall Survival  Storage (d)  \\\n",
       "0             NaN                       NaN               NaN          NaN   \n",
       "1             DWD                        70             919.0        638.0   \n",
       "2             DWD                        78             297.0         17.0   \n",
       "3             MET                       578            1082.0        341.0   \n",
       "4             DWD                       267             538.0        634.0   \n",
       "\n",
       "  10x Chemistry        Vital_status TNM_classification Grade  \n",
       "0           NaN                 NaN                NaN   NaN  \n",
       "1            v3  Deceased - Disease             T3N1M0  g2-3  \n",
       "2            v3  Deceased - Disease             T3N2M0    g2  \n",
       "3            v3  Alive - Metastatic             T3N0M0    gX  \n",
       "4            v2  Deceased - Disease             T3N1M0    g3  "
      ]
     },
     "execution_count": 757,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hwang.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 758,
   "id": "b2a0b4ee-2656-4c14-82b0-ac76b910d181",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>PDAC_T_25</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>PDAC_U_2</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>PDAC_U_16</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>PDAC_T_4</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>PDAC_U_15</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>PDAC_U_18</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>PDAC_T_16</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>PDAC_T_23</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T23</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>PDAC_T_7</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>PDAC_T_22</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>PDAC_U_7</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>PDAC_T_5</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>PDAC_U_12</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>PDAC_U_10</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>PDAC_U_5</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>PDAC_U_3</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>PDAC_T_18</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>PDAC_U_11</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>PDAC_T_20</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>PDAC_U_13</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>PDAC_U_14</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>PDAC_T_6</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>PDAC_T_17</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>PDAC_T_2</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>PDAC_U_1</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>PDAC_U_9</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>PDAC_U_6</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>PDAC_U_4</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PDAC_T_13</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>PDAC_T_14</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>PDAC_T_15</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>PDAC_T_3</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>PDAC_T_1</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>PDAC_T_10</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>PDAC_T_12</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>PDAC_T_24</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>155</th>\n",
       "      <td>PDAC_T_21</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>PDAC_T_11</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>PDAC_U_17</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>PDAC_U_8</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>U8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>159</th>\n",
       "      <td>PDAC_T_19</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160</th>\n",
       "      <td>PDAC_T_8</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>PDAC_T_9</td>\n",
       "      <td>Hwang</td>\n",
       "      <td>T9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID_harmonised Dataset Unique_ID Case KRAS_mutation KRAS_type TP53_type  \\\n",
       "119     PDAC_T_25   Hwang       T25  NaN           NaN       NaN       NaN   \n",
       "120      PDAC_U_2   Hwang        U2  NaN           NaN       NaN       NaN   \n",
       "121     PDAC_U_16   Hwang       U16  NaN           NaN       NaN       NaN   \n",
       "122      PDAC_T_4   Hwang        T4  NaN           NaN       NaN       NaN   \n",
       "123     PDAC_U_15   Hwang       U15  NaN           NaN       NaN       NaN   \n",
       "124     PDAC_U_18   Hwang       U18  NaN           NaN       NaN       NaN   \n",
       "125     PDAC_T_16   Hwang       T16  NaN           NaN       NaN       NaN   \n",
       "126     PDAC_T_23   Hwang       T23  NaN           NaN       NaN       NaN   \n",
       "127      PDAC_T_7   Hwang        T7  NaN           NaN       NaN       NaN   \n",
       "128     PDAC_T_22   Hwang       T22  NaN           NaN       NaN       NaN   \n",
       "129      PDAC_U_7   Hwang        U7  NaN           NaN       NaN       NaN   \n",
       "130      PDAC_T_5   Hwang        T5  NaN           NaN       NaN       NaN   \n",
       "131     PDAC_U_12   Hwang       U12  NaN           NaN       NaN       NaN   \n",
       "132     PDAC_U_10   Hwang       U10  NaN           NaN       NaN       NaN   \n",
       "133      PDAC_U_5   Hwang        U5  NaN           NaN       NaN       NaN   \n",
       "134      PDAC_U_3   Hwang        U3  NaN           NaN       NaN       NaN   \n",
       "135     PDAC_T_18   Hwang       T18  NaN           NaN       NaN       NaN   \n",
       "136     PDAC_U_11   Hwang       U11  NaN           NaN       NaN       NaN   \n",
       "137     PDAC_T_20   Hwang       T20  NaN           NaN       NaN       NaN   \n",
       "138     PDAC_U_13   Hwang       U13  NaN           NaN       NaN       NaN   \n",
       "139     PDAC_U_14   Hwang       U14  NaN           NaN       NaN       NaN   \n",
       "140      PDAC_T_6   Hwang        T6  NaN           NaN       NaN       NaN   \n",
       "141     PDAC_T_17   Hwang       T17  NaN           NaN       NaN       NaN   \n",
       "142      PDAC_T_2   Hwang        T2  NaN           NaN       NaN       NaN   \n",
       "143      PDAC_U_1   Hwang        U1  NaN           NaN       NaN       NaN   \n",
       "144      PDAC_U_9   Hwang        U9  NaN           NaN       NaN       NaN   \n",
       "145      PDAC_U_6   Hwang        U6  NaN           NaN       NaN       NaN   \n",
       "146      PDAC_U_4   Hwang        U4  NaN           NaN       NaN       NaN   \n",
       "147     PDAC_T_13   Hwang       T13  NaN           NaN       NaN       NaN   \n",
       "148     PDAC_T_14   Hwang       T14  NaN           NaN       NaN       NaN   \n",
       "149     PDAC_T_15   Hwang       T15  NaN           NaN       NaN       NaN   \n",
       "150      PDAC_T_3   Hwang        T3  NaN           NaN       NaN       NaN   \n",
       "151      PDAC_T_1   Hwang        T1  NaN           NaN       NaN       NaN   \n",
       "152     PDAC_T_10   Hwang       T10  NaN           NaN       NaN       NaN   \n",
       "153     PDAC_T_12   Hwang       T12  NaN           NaN       NaN       NaN   \n",
       "154     PDAC_T_24   Hwang       T24  NaN           NaN       NaN       NaN   \n",
       "155     PDAC_T_21   Hwang       T21  NaN           NaN       NaN       NaN   \n",
       "156     PDAC_T_11   Hwang       T11  NaN           NaN       NaN       NaN   \n",
       "157     PDAC_U_17   Hwang       U17  NaN           NaN       NaN       NaN   \n",
       "158      PDAC_U_8   Hwang        U8  NaN           NaN       NaN       NaN   \n",
       "159     PDAC_T_19   Hwang       T19  NaN           NaN       NaN       NaN   \n",
       "160      PDAC_T_8   Hwang        T8  NaN           NaN       NaN       NaN   \n",
       "161      PDAC_T_9   Hwang        T9  NaN           NaN       NaN       NaN   \n",
       "\n",
       "    SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "119        NaN         NaN          NaN            NaN            NaN   \n",
       "120        NaN         NaN          NaN            NaN            NaN   \n",
       "121        NaN         NaN          NaN            NaN            NaN   \n",
       "122        NaN         NaN          NaN            NaN            NaN   \n",
       "123        NaN         NaN          NaN            NaN            NaN   \n",
       "124        NaN         NaN          NaN            NaN            NaN   \n",
       "125        NaN         NaN          NaN            NaN            NaN   \n",
       "126        NaN         NaN          NaN            NaN            NaN   \n",
       "127        NaN         NaN          NaN            NaN            NaN   \n",
       "128        NaN         NaN          NaN            NaN            NaN   \n",
       "129        NaN         NaN          NaN            NaN            NaN   \n",
       "130        NaN         NaN          NaN            NaN            NaN   \n",
       "131        NaN         NaN          NaN            NaN            NaN   \n",
       "132        NaN         NaN          NaN            NaN            NaN   \n",
       "133        NaN         NaN          NaN            NaN            NaN   \n",
       "134        NaN         NaN          NaN            NaN            NaN   \n",
       "135        NaN         NaN          NaN            NaN            NaN   \n",
       "136        NaN         NaN          NaN            NaN            NaN   \n",
       "137        NaN         NaN          NaN            NaN            NaN   \n",
       "138        NaN         NaN          NaN            NaN            NaN   \n",
       "139        NaN         NaN          NaN            NaN            NaN   \n",
       "140        NaN         NaN          NaN            NaN            NaN   \n",
       "141        NaN         NaN          NaN            NaN            NaN   \n",
       "142        NaN         NaN          NaN            NaN            NaN   \n",
       "143        NaN         NaN          NaN            NaN            NaN   \n",
       "144        NaN         NaN          NaN            NaN            NaN   \n",
       "145        NaN         NaN          NaN            NaN            NaN   \n",
       "146        NaN         NaN          NaN            NaN            NaN   \n",
       "147        NaN         NaN          NaN            NaN            NaN   \n",
       "148        NaN         NaN          NaN            NaN            NaN   \n",
       "149        NaN         NaN          NaN            NaN            NaN   \n",
       "150        NaN         NaN          NaN            NaN            NaN   \n",
       "151        NaN         NaN          NaN            NaN            NaN   \n",
       "152        NaN         NaN          NaN            NaN            NaN   \n",
       "153        NaN         NaN          NaN            NaN            NaN   \n",
       "154        NaN         NaN          NaN            NaN            NaN   \n",
       "155        NaN         NaN          NaN            NaN            NaN   \n",
       "156        NaN         NaN          NaN            NaN            NaN   \n",
       "157        NaN         NaN          NaN            NaN            NaN   \n",
       "158        NaN         NaN          NaN            NaN            NaN   \n",
       "159        NaN         NaN          NaN            NaN            NaN   \n",
       "160        NaN         NaN          NaN            NaN            NaN   \n",
       "161        NaN         NaN          NaN            NaN            NaN   \n",
       "\n",
       "    Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "119             NaN            NaN               NaN        NaN   \n",
       "120             NaN            NaN               NaN        NaN   \n",
       "121             NaN            NaN               NaN        NaN   \n",
       "122             NaN            NaN               NaN        NaN   \n",
       "123             NaN            NaN               NaN        NaN   \n",
       "124             NaN            NaN               NaN        NaN   \n",
       "125             NaN            NaN               NaN        NaN   \n",
       "126             NaN            NaN               NaN        NaN   \n",
       "127             NaN            NaN               NaN        NaN   \n",
       "128             NaN            NaN               NaN        NaN   \n",
       "129             NaN            NaN               NaN        NaN   \n",
       "130             NaN            NaN               NaN        NaN   \n",
       "131             NaN            NaN               NaN        NaN   \n",
       "132             NaN            NaN               NaN        NaN   \n",
       "133             NaN            NaN               NaN        NaN   \n",
       "134             NaN            NaN               NaN        NaN   \n",
       "135             NaN            NaN               NaN        NaN   \n",
       "136             NaN            NaN               NaN        NaN   \n",
       "137             NaN            NaN               NaN        NaN   \n",
       "138             NaN            NaN               NaN        NaN   \n",
       "139             NaN            NaN               NaN        NaN   \n",
       "140             NaN            NaN               NaN        NaN   \n",
       "141             NaN            NaN               NaN        NaN   \n",
       "142             NaN            NaN               NaN        NaN   \n",
       "143             NaN            NaN               NaN        NaN   \n",
       "144             NaN            NaN               NaN        NaN   \n",
       "145             NaN            NaN               NaN        NaN   \n",
       "146             NaN            NaN               NaN        NaN   \n",
       "147             NaN            NaN               NaN        NaN   \n",
       "148             NaN            NaN               NaN        NaN   \n",
       "149             NaN            NaN               NaN        NaN   \n",
       "150             NaN            NaN               NaN        NaN   \n",
       "151             NaN            NaN               NaN        NaN   \n",
       "152             NaN            NaN               NaN        NaN   \n",
       "153             NaN            NaN               NaN        NaN   \n",
       "154             NaN            NaN               NaN        NaN   \n",
       "155             NaN            NaN               NaN        NaN   \n",
       "156             NaN            NaN               NaN        NaN   \n",
       "157             NaN            NaN               NaN        NaN   \n",
       "158             NaN            NaN               NaN        NaN   \n",
       "159             NaN            NaN               NaN        NaN   \n",
       "160             NaN            NaN               NaN        NaN   \n",
       "161             NaN            NaN               NaN        NaN   \n",
       "\n",
       "     Time_from_diagnosis_to_surgery_(weeks) Cycles Location  Grade  Size_(cm)  \\\n",
       "119                                     NaN    NaN      NaN    NaN        NaN   \n",
       "120                                     NaN    NaN      NaN    NaN        NaN   \n",
       "121                                     NaN    NaN      NaN    NaN        NaN   \n",
       "122                                     NaN    NaN      NaN    NaN        NaN   \n",
       "123                                     NaN    NaN      NaN    NaN        NaN   \n",
       "124                                     NaN    NaN      NaN    NaN        NaN   \n",
       "125                                     NaN    NaN      NaN    NaN        NaN   \n",
       "126                                     NaN    NaN      NaN    NaN        NaN   \n",
       "127                                     NaN    NaN      NaN    NaN        NaN   \n",
       "128                                     NaN    NaN      NaN    NaN        NaN   \n",
       "129                                     NaN    NaN      NaN    NaN        NaN   \n",
       "130                                     NaN    NaN      NaN    NaN        NaN   \n",
       "131                                     NaN    NaN      NaN    NaN        NaN   \n",
       "132                                     NaN    NaN      NaN    NaN        NaN   \n",
       "133                                     NaN    NaN      NaN    NaN        NaN   \n",
       "134                                     NaN    NaN      NaN    NaN        NaN   \n",
       "135                                     NaN    NaN      NaN    NaN        NaN   \n",
       "136                                     NaN    NaN      NaN    NaN        NaN   \n",
       "137                                     NaN    NaN      NaN    NaN        NaN   \n",
       "138                                     NaN    NaN      NaN    NaN        NaN   \n",
       "139                                     NaN    NaN      NaN    NaN        NaN   \n",
       "140                                     NaN    NaN      NaN    NaN        NaN   \n",
       "141                                     NaN    NaN      NaN    NaN        NaN   \n",
       "142                                     NaN    NaN      NaN    NaN        NaN   \n",
       "143                                     NaN    NaN      NaN    NaN        NaN   \n",
       "144                                     NaN    NaN      NaN    NaN        NaN   \n",
       "145                                     NaN    NaN      NaN    NaN        NaN   \n",
       "146                                     NaN    NaN      NaN    NaN        NaN   \n",
       "147                                     NaN    NaN      NaN    NaN        NaN   \n",
       "148                                     NaN    NaN      NaN    NaN        NaN   \n",
       "149                                     NaN    NaN      NaN    NaN        NaN   \n",
       "150                                     NaN    NaN      NaN    NaN        NaN   \n",
       "151                                     NaN    NaN      NaN    NaN        NaN   \n",
       "152                                     NaN    NaN      NaN    NaN        NaN   \n",
       "153                                     NaN    NaN      NaN    NaN        NaN   \n",
       "154                                     NaN    NaN      NaN    NaN        NaN   \n",
       "155                                     NaN    NaN      NaN    NaN        NaN   \n",
       "156                                     NaN    NaN      NaN    NaN        NaN   \n",
       "157                                     NaN    NaN      NaN    NaN        NaN   \n",
       "158                                     NaN    NaN      NaN    NaN        NaN   \n",
       "159                                     NaN    NaN      NaN    NaN        NaN   \n",
       "160                                     NaN    NaN      NaN    NaN        NaN   \n",
       "161                                     NaN    NaN      NaN    NaN        NaN   \n",
       "\n",
       "    Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "119   NaN      NaN          NaN                     NaN                  NaN   \n",
       "120   NaN      NaN          NaN                     NaN                  NaN   \n",
       "121   NaN      NaN          NaN                     NaN                  NaN   \n",
       "122   NaN      NaN          NaN                     NaN                  NaN   \n",
       "123   NaN      NaN          NaN                     NaN                  NaN   \n",
       "124   NaN      NaN          NaN                     NaN                  NaN   \n",
       "125   NaN      NaN          NaN                     NaN                  NaN   \n",
       "126   NaN      NaN          NaN                     NaN                  NaN   \n",
       "127   NaN      NaN          NaN                     NaN                  NaN   \n",
       "128   NaN      NaN          NaN                     NaN                  NaN   \n",
       "129   NaN      NaN          NaN                     NaN                  NaN   \n",
       "130   NaN      NaN          NaN                     NaN                  NaN   \n",
       "131   NaN      NaN          NaN                     NaN                  NaN   \n",
       "132   NaN      NaN          NaN                     NaN                  NaN   \n",
       "133   NaN      NaN          NaN                     NaN                  NaN   \n",
       "134   NaN      NaN          NaN                     NaN                  NaN   \n",
       "135   NaN      NaN          NaN                     NaN                  NaN   \n",
       "136   NaN      NaN          NaN                     NaN                  NaN   \n",
       "137   NaN      NaN          NaN                     NaN                  NaN   \n",
       "138   NaN      NaN          NaN                     NaN                  NaN   \n",
       "139   NaN      NaN          NaN                     NaN                  NaN   \n",
       "140   NaN      NaN          NaN                     NaN                  NaN   \n",
       "141   NaN      NaN          NaN                     NaN                  NaN   \n",
       "142   NaN      NaN          NaN                     NaN                  NaN   \n",
       "143   NaN      NaN          NaN                     NaN                  NaN   \n",
       "144   NaN      NaN          NaN                     NaN                  NaN   \n",
       "145   NaN      NaN          NaN                     NaN                  NaN   \n",
       "146   NaN      NaN          NaN                     NaN                  NaN   \n",
       "147   NaN      NaN          NaN                     NaN                  NaN   \n",
       "148   NaN      NaN          NaN                     NaN                  NaN   \n",
       "149   NaN      NaN          NaN                     NaN                  NaN   \n",
       "150   NaN      NaN          NaN                     NaN                  NaN   \n",
       "151   NaN      NaN          NaN                     NaN                  NaN   \n",
       "152   NaN      NaN          NaN                     NaN                  NaN   \n",
       "153   NaN      NaN          NaN                     NaN                  NaN   \n",
       "154   NaN      NaN          NaN                     NaN                  NaN   \n",
       "155   NaN      NaN          NaN                     NaN                  NaN   \n",
       "156   NaN      NaN          NaN                     NaN                  NaN   \n",
       "157   NaN      NaN          NaN                     NaN                  NaN   \n",
       "158   NaN      NaN          NaN                     NaN                  NaN   \n",
       "159   NaN      NaN          NaN                     NaN                  NaN   \n",
       "160   NaN      NaN          NaN                     NaN                  NaN   \n",
       "161   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "    Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "119    NaN     NaN             NaN            NaN   \n",
       "120    NaN     NaN             NaN            NaN   \n",
       "121    NaN     NaN             NaN            NaN   \n",
       "122    NaN     NaN             NaN            NaN   \n",
       "123    NaN     NaN             NaN            NaN   \n",
       "124    NaN     NaN             NaN            NaN   \n",
       "125    NaN     NaN             NaN            NaN   \n",
       "126    NaN     NaN             NaN            NaN   \n",
       "127    NaN     NaN             NaN            NaN   \n",
       "128    NaN     NaN             NaN            NaN   \n",
       "129    NaN     NaN             NaN            NaN   \n",
       "130    NaN     NaN             NaN            NaN   \n",
       "131    NaN     NaN             NaN            NaN   \n",
       "132    NaN     NaN             NaN            NaN   \n",
       "133    NaN     NaN             NaN            NaN   \n",
       "134    NaN     NaN             NaN            NaN   \n",
       "135    NaN     NaN             NaN            NaN   \n",
       "136    NaN     NaN             NaN            NaN   \n",
       "137    NaN     NaN             NaN            NaN   \n",
       "138    NaN     NaN             NaN            NaN   \n",
       "139    NaN     NaN             NaN            NaN   \n",
       "140    NaN     NaN             NaN            NaN   \n",
       "141    NaN     NaN             NaN            NaN   \n",
       "142    NaN     NaN             NaN            NaN   \n",
       "143    NaN     NaN             NaN            NaN   \n",
       "144    NaN     NaN             NaN            NaN   \n",
       "145    NaN     NaN             NaN            NaN   \n",
       "146    NaN     NaN             NaN            NaN   \n",
       "147    NaN     NaN             NaN            NaN   \n",
       "148    NaN     NaN             NaN            NaN   \n",
       "149    NaN     NaN             NaN            NaN   \n",
       "150    NaN     NaN             NaN            NaN   \n",
       "151    NaN     NaN             NaN            NaN   \n",
       "152    NaN     NaN             NaN            NaN   \n",
       "153    NaN     NaN             NaN            NaN   \n",
       "154    NaN     NaN             NaN            NaN   \n",
       "155    NaN     NaN             NaN            NaN   \n",
       "156    NaN     NaN             NaN            NaN   \n",
       "157    NaN     NaN             NaN            NaN   \n",
       "158    NaN     NaN             NaN            NaN   \n",
       "159    NaN     NaN             NaN            NaN   \n",
       "160    NaN     NaN             NaN            NaN   \n",
       "161    NaN     NaN             NaN            NaN   \n",
       "\n",
       "     Survival_post_surgery_(weeks) Vital_status TNM_classification  \n",
       "119                            NaN          NaN                NaN  \n",
       "120                            NaN          NaN                NaN  \n",
       "121                            NaN          NaN                NaN  \n",
       "122                            NaN          NaN                NaN  \n",
       "123                            NaN          NaN                NaN  \n",
       "124                            NaN          NaN                NaN  \n",
       "125                            NaN          NaN                NaN  \n",
       "126                            NaN          NaN                NaN  \n",
       "127                            NaN          NaN                NaN  \n",
       "128                            NaN          NaN                NaN  \n",
       "129                            NaN          NaN                NaN  \n",
       "130                            NaN          NaN                NaN  \n",
       "131                            NaN          NaN                NaN  \n",
       "132                            NaN          NaN                NaN  \n",
       "133                            NaN          NaN                NaN  \n",
       "134                            NaN          NaN                NaN  \n",
       "135                            NaN          NaN                NaN  \n",
       "136                            NaN          NaN                NaN  \n",
       "137                            NaN          NaN                NaN  \n",
       "138                            NaN          NaN                NaN  \n",
       "139                            NaN          NaN                NaN  \n",
       "140                            NaN          NaN                NaN  \n",
       "141                            NaN          NaN                NaN  \n",
       "142                            NaN          NaN                NaN  \n",
       "143                            NaN          NaN                NaN  \n",
       "144                            NaN          NaN                NaN  \n",
       "145                            NaN          NaN                NaN  \n",
       "146                            NaN          NaN                NaN  \n",
       "147                            NaN          NaN                NaN  \n",
       "148                            NaN          NaN                NaN  \n",
       "149                            NaN          NaN                NaN  \n",
       "150                            NaN          NaN                NaN  \n",
       "151                            NaN          NaN                NaN  \n",
       "152                            NaN          NaN                NaN  \n",
       "153                            NaN          NaN                NaN  \n",
       "154                            NaN          NaN                NaN  \n",
       "155                            NaN          NaN                NaN  \n",
       "156                            NaN          NaN                NaN  \n",
       "157                            NaN          NaN                NaN  \n",
       "158                            NaN          NaN                NaN  \n",
       "159                            NaN          NaN                NaN  \n",
       "160                            NaN          NaN                NaN  \n",
       "161                            NaN          NaN                NaN  "
      ]
     },
     "execution_count": 758,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset.str.contains('Hwa')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 759,
   "id": "65faca86-de36-4044-8caa-8d3c2bdb95a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43, 33)"
      ]
     },
     "execution_count": 759,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset.str.contains('Hwa')].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 760,
   "id": "c2aeb065-41ec-4d80-9c7f-82e1eb2d6dbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang_clean = hwang[hwang['ID'].notna()].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 761,
   "id": "2662794f-48e6-41a4-af1d-f73a5544a46b",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang_clean['ID'] = (\n",
    "    hwang_clean['ID']\n",
    "    .astype(str)                   # ensure all are strings\n",
    "    .str.strip()                   # remove leading/trailing spaces\n",
    "    .str.replace('†', '', regex=False)  # remove dagger symbol\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 762,
   "id": "cab4cc7d-0dbe-4d54-a2c7-fc49b5658f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang_clean = hwang_clean[hwang_clean['ID'].str.startswith('PDAC_')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 763,
   "id": "fa45ebc7-7720-42a1-bc92-50f2dec5fa43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Age (decade)', 'Sex', 'Stage/Grade', 'Margin', 'Histology',\n",
       "       'Neoadjuvant', 'Status Last FUP', 'Progression-Free Survival',\n",
       "       'Overall Survival', 'Storage (d)', '10x Chemistry', 'Vital_status',\n",
       "       'TNM_classification', 'Grade'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 763,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hwang_clean.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 764,
   "id": "d933310a-36a7-4d8e-9e1e-d34c4d26c5bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_id_merged['Overall Survival'] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 765,
   "id": "b74fd11a-1247-4869-ad1d-c8dae6984fa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged['Overall Survival'] = df_id_merged['Survival_post_surgery_(weeks)'] * 7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 766,
   "id": "ce83e8cb-92a8-494e-a062-63caad34768a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# interesing_cols = ['ID', 'Grade', 'TNM_classification' 'Vital_status', 'Overall Survival']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 767,
   "id": "eb6969a6-cd90-459c-b2bb-58e974adae15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43"
      ]
     },
     "execution_count": 767,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(hwang_clean.ID) & set(df_id_merged.ID_harmonised))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 768,
   "id": "f6f88a2f-91db-44b4-976c-2f51ef9537ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "hwang_map = hwang_clean.set_index('ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 771,
   "id": "4fbf84be-2499-42ad-9303-01acd308fc11",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Grade','TNM_classification', 'Vital_status', 'Overall Survival']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 772,
   "id": "327e95c3-bf11-4a70-a18c-a3f76239768a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in interesing_cols:\n",
    "    if col in df_id_merged.columns:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(hwang_map[col]).fillna(df_id_merged[col])\n",
    "    else:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(hwang_map[col])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 773,
   "id": "ace208bf-c942-4aa0-9625-cc057579c541",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Classical</td>\n",
       "      <td>ADEX</td>\n",
       "      <td>Exocrine-like</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>IIb</td>\n",
       "      <td>Type II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, extends to duodenum and vascular groove</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HT061P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>p.G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HT061P1</td>\n",
       "      <td>12.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Multiple myeloma (remission)</td>\n",
       "      <td>A-fib, chronic kidney disease</td>\n",
       "      <td>Former</td>\n",
       "      <td>No</td>\n",
       "      <td>pT2, poorly differentiated tumor arising in IPMN</td>\n",
       "      <td>3/7, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ID_harmonised Dataset Unique_ID     Case KRAS_mutation KRAS_type  TP53_type  \\\n",
       "0       HT056P1    Ding   HT056P1  HT056P1        p.G12V  Missense   Missense   \n",
       "1       HT060P1    Ding   HT060P1  HT060P1        p.G12V  Missense  Wild_Type   \n",
       "2       HT061P1    Ding   HT061P1  HT061P1        p.G12D  Missense   Nonsense   \n",
       "\n",
       "  SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "0  Wild_Type   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "1  Wild_Type   Wild_Type     Nonsense       Nonsense      Wild_Type   \n",
       "2  Wild_Type   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "\n",
       "  Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "0      Basal-like    Immunogenic  Quasimesenchymal    HT056P1   \n",
       "1       Classical           ADEX     Exocrine-like    HT060P1   \n",
       "2             NaN            NaN               NaN    HT061P1   \n",
       "\n",
       "   Time_from_diagnosis_to_surgery_(weeks) Cycles                Location  \\\n",
       "0                                    20.0      6  head, uncinate_process   \n",
       "1                                    20.0      6  head, uncinate_process   \n",
       "2                                    12.0    NaN                    head   \n",
       "\n",
       "  Grade  Size_(cm) Stage Diabetes Pancreatitis       Previous_cancer_history  \\\n",
       "0   3.0        3.2   NaN       No          Yes                            No   \n",
       "1   3.0        3.2   IIb  Type II          NaN                           NaN   \n",
       "2   3.0        3.5   NaN       No          NaN  Multiple myeloma (remission)   \n",
       "\n",
       "            Other_co-morbidities  Smoker     Alcohol  \\\n",
       "0                             No      No  Occasional   \n",
       "1                             No      No  Occasional   \n",
       "2  A-fib, chronic kidney disease  Former          No   \n",
       "\n",
       "                                    Pathology_notes Lymph_node_met  \\\n",
       "0            ypT2, focal sarcomatoid transformation      2/21, pN1   \n",
       "1     ypT2, extends to duodenum and vascular groove      2/25, pN1   \n",
       "2  pT2, poorly differentiated tumor arising in IPMN       3/7, pN1   \n",
       "\n",
       "   Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "0                            NaN          NaN           ypT2 pN1   \n",
       "1                            NaN          NaN           ypT2 pN1   \n",
       "2                            NaN          NaN            pT2 pN1   \n",
       "\n",
       "   Overall Survival  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  "
      ]
     },
     "execution_count": 773,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 774,
   "id": "62166361-d8ce-439b-8b35-35bc950ecde1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Hwang</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Grade</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1.0</th>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2.0</th>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.0</th>\n",
       "      <td>7.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g2-3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gX</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Ding  Ding_snRNA-seq  Hwang\n",
       "Grade                               \n",
       "1.0       4.0             2.0    NaN\n",
       "2.0       8.0             4.0    NaN\n",
       "3.0       7.0             4.0    NaN\n",
       "g2        NaN             NaN   22.0\n",
       "g2-3      NaN             NaN    4.0\n",
       "g3        NaN             NaN    7.0\n",
       "gX        NaN             NaN   10.0"
      ]
     },
     "execution_count": 774,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'Grade']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 775,
   "id": "5bccf4e2-7791-4b23-b4e7-c75c6aa9d8bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Hwang</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNM_classification</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T2N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT1a pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN2</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT3 pN2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpT2 moderately differentiated pN1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yT2 pN1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1 moderately differentiated positive for perineural invasion negative for lymphovascular pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1aN0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1aN1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1cN0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1cN1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN0</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                                             Ding  Ding_snRNA-seq  \\\n",
       "TNM_classification                                                         \n",
       "T2N0M0                                               NaN             NaN   \n",
       "T2N1M0                                               NaN             NaN   \n",
       "T2N2M0                                               NaN             NaN   \n",
       "T3N0M0                                               NaN             NaN   \n",
       "T3N1M0                                               NaN             NaN   \n",
       "T3N2M0                                               NaN             NaN   \n",
       "pT1a pN0                                             NaN             1.0   \n",
       "pT2 pN0                                              NaN             2.0   \n",
       "pT2 pN1                                              2.0             1.0   \n",
       "pT2 pN2                                              4.0             NaN   \n",
       "pT3 pN2                                              1.0             NaN   \n",
       "rpT2 moderately differentiated pN1                   NaN             1.0   \n",
       "yT2 pN1                                              1.0             NaN   \n",
       "ypT1 moderately differentiated positive for per...   NaN             1.0   \n",
       "ypT1aN0M0                                            NaN             NaN   \n",
       "ypT1aN1M0                                            NaN             NaN   \n",
       "ypT1cN0M0                                            NaN             NaN   \n",
       "ypT1cN1M0                                            NaN             NaN   \n",
       "ypT2 pN0                                             3.0             1.0   \n",
       "ypT2 pN1                                             5.0             1.0   \n",
       "ypT2 pN2                                             3.0             NaN   \n",
       "ypT2N0M0                                             NaN             NaN   \n",
       "ypT2N1M0                                             NaN             NaN   \n",
       "ypT2N2M0                                             NaN             NaN   \n",
       "ypT3 pN0                                             NaN             2.0   \n",
       "ypT3 pN2                                             2.0             1.0   \n",
       "ypT3N0M0                                             NaN             NaN   \n",
       "ypT3N1M0                                             NaN             NaN   \n",
       "\n",
       "Dataset                                             Hwang  \n",
       "TNM_classification                                         \n",
       "T2N0M0                                                2.0  \n",
       "T2N1M0                                                6.0  \n",
       "T2N2M0                                                1.0  \n",
       "T3N0M0                                                3.0  \n",
       "T3N1M0                                                3.0  \n",
       "T3N2M0                                                3.0  \n",
       "pT1a pN0                                              NaN  \n",
       "pT2 pN0                                               NaN  \n",
       "pT2 pN1                                               NaN  \n",
       "pT2 pN2                                               NaN  \n",
       "pT3 pN2                                               NaN  \n",
       "rpT2 moderately differentiated pN1                    NaN  \n",
       "yT2 pN1                                               NaN  \n",
       "ypT1 moderately differentiated positive for per...    NaN  \n",
       "ypT1aN0M0                                             2.0  \n",
       "ypT1aN1M0                                             1.0  \n",
       "ypT1cN0M0                                             2.0  \n",
       "ypT1cN1M0                                             1.0  \n",
       "ypT2 pN0                                              NaN  \n",
       "ypT2 pN1                                              NaN  \n",
       "ypT2 pN2                                              NaN  \n",
       "ypT2N0M0                                              6.0  \n",
       "ypT2N1M0                                              2.0  \n",
       "ypT2N2M0                                              1.0  \n",
       "ypT3 pN0                                              NaN  \n",
       "ypT3 pN2                                              NaN  \n",
       "ypT3N0M0                                              8.0  \n",
       "ypT3N1M0                                              2.0  "
      ]
     },
     "execution_count": 775,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'TNM_classification']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31b58d80-1bd3-4614-95e1-827d04ef8ee7",
   "metadata": {},
   "source": [
    "# Simeone"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 776,
   "id": "cd4bb986-53fe-449a-9e87-2f98cbe2e759",
   "metadata": {},
   "outputs": [],
   "source": [
    "simeone = pd.read_excel('metadata_publications/Simeone_metadata.xlsx', sheet_name='Genomic Mutations')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 777,
   "id": "d9148ab5-0411-4f5d-9954-ed245cad8b9f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Source</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Seq Platform</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>TP53</th>\n",
       "      <th>CDKN2A</th>\n",
       "      <th>SMAD4</th>\n",
       "      <th>other genetic features</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P01</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12R</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P02</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P03</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>WES</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P04</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P05</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P06</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P07</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P08</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>P09</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>P10</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>P11</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12L</td>\n",
       "      <td>Missense</td>\n",
       "      <td>LGR</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>P12</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>Missense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>P13</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>WES</td>\n",
       "      <td>WT</td>\n",
       "      <td>Frameshift</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>P14</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>WES</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>P15</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>WT</td>\n",
       "      <td>MSI-H (MSH6 mut, MSH2 mut)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>P16</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>P17</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>P18</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Loss</td>\n",
       "      <td>WT</td>\n",
       "      <td>germline BRCA-1 mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P19</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>germline BRCA-2 mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>P20</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>P21</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>P22</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P23</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P24</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>Q61K</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P25</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P26</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P27</td>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>N/A = lack of sufficient tissue for sequencing</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WT = wildtype</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mut = mutation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient    Source                                       Procedure  \\\n",
       "0      P01     Liver                                          Biopsy   \n",
       "1      P02     Liver                                          Biopsy   \n",
       "2      P03  Pancreas                                       Resection   \n",
       "3      P04  Pancreas                                       Resection   \n",
       "4      P05  Pancreas                                       Resection   \n",
       "5      P06  Pancreas                                          Biopsy   \n",
       "6      P07  Pancreas                                       Resection   \n",
       "7      P08  Pancreas                                       Resection   \n",
       "8      P09  Pancreas                                          Biopsy   \n",
       "9      P10  Pancreas                                       Resection   \n",
       "10     P11     Liver                                          Biopsy   \n",
       "11     P12  Pancreas                                          Biopsy   \n",
       "12     P13  Pancreas                                          Biopsy   \n",
       "13     P14  Pancreas                                       Resection   \n",
       "14     P15  Pancreas                                       Resection   \n",
       "15     P16     Liver                                          Biopsy   \n",
       "16     P17     Liver                                          Biopsy   \n",
       "17     P18     Liver                                          Biopsy   \n",
       "18     P19  Pancreas                                       Resection   \n",
       "19     P20  Pancreas                                          Biopsy   \n",
       "20     P21     Liver                                          Biopsy   \n",
       "21     P22  Pancreas                                          Biopsy   \n",
       "22     P23  Pancreas                                       Resection   \n",
       "23     P24     Liver                                          Biopsy   \n",
       "24     P25     Liver                                          Biopsy   \n",
       "25     P26  Pancreas                                          Biopsy   \n",
       "26     P27     Liver                                          Biopsy   \n",
       "27     NaN       NaN                                             NaN   \n",
       "28     NaN       NaN  N/A = lack of sufficient tissue for sequencing   \n",
       "29     NaN       NaN                                   WT = wildtype   \n",
       "30     NaN       NaN                                  Mut = mutation   \n",
       "\n",
       "   Seq Platform  KRAS         TP53       CDKN2A        SMAD4  \\\n",
       "0        Tempus  G12R  Stop gained           WT           WT   \n",
       "1        Tempus  G12D  Splice site           WT           WT   \n",
       "2           WES    WT           WT           WT           WT   \n",
       "3        Tempus  G12D     Missense           WT           WT   \n",
       "4    Foundation  G12D     Missense           WT  Stop gained   \n",
       "5           NaN   NaN          NaN          NaN          NaN   \n",
       "6        Tempus  G12V           WT           WT           WT   \n",
       "7        Tempus  G12D     Missense           WT           WT   \n",
       "8           NaN   NaN          NaN          NaN          NaN   \n",
       "9        Tempus  G12V           WT           WT           WT   \n",
       "10   Foundation  G12L     Missense          LGR           WT   \n",
       "11   Foundation  G12V  Stop gained  Stop gained     Missense   \n",
       "12          WES    WT   Frameshift           WT           WT   \n",
       "13          WES    WT           WT           WT           WT   \n",
       "14       Tempus  G12V     Missense  Stop gained           WT   \n",
       "15       Tempus  G12V     Missense           WT           WT   \n",
       "16          NaN   NaN          NaN          NaN          NaN   \n",
       "17   Foundation  G12D     Missense         Loss           WT   \n",
       "18   Foundation  G12V  Splice site           WT           WT   \n",
       "19       Tempus  G12D     Missense           WT           WT   \n",
       "20       Tempus  G12D     Missense           WT           WT   \n",
       "21          NaN   NaN          NaN          NaN          NaN   \n",
       "22       Tempus  G12D     Missense           WT           WT   \n",
       "23       Tempus  Q61K     Missense           WT           WT   \n",
       "24       Tempus  G12V           WT           WT           WT   \n",
       "25       Tempus  G12V           WT           WT           WT   \n",
       "26       Tempus    WT           WT           WT           WT   \n",
       "27          NaN   NaN          NaN          NaN          NaN   \n",
       "28          NaN   NaN          NaN          NaN          NaN   \n",
       "29          NaN   NaN          NaN          NaN          NaN   \n",
       "30          NaN   NaN          NaN          NaN          NaN   \n",
       "\n",
       "        other genetic features  \n",
       "0                          NaN  \n",
       "1                          NaN  \n",
       "2                          NaN  \n",
       "3                          NaN  \n",
       "4                          NaN  \n",
       "5                          NaN  \n",
       "6                          NaN  \n",
       "7                          NaN  \n",
       "8                          NaN  \n",
       "9                          NaN  \n",
       "10                         NaN  \n",
       "11                         NaN  \n",
       "12                         NaN  \n",
       "13                         NaN  \n",
       "14  MSI-H (MSH6 mut, MSH2 mut)  \n",
       "15                         NaN  \n",
       "16                         NaN  \n",
       "17         germline BRCA-1 mut  \n",
       "18         germline BRCA-2 mut  \n",
       "19                         NaN  \n",
       "20                         NaN  \n",
       "21                         NaN  \n",
       "22                         NaN  \n",
       "23                         NaN  \n",
       "24                         NaN  \n",
       "25                         NaN  \n",
       "26                         NaN  \n",
       "27                         NaN  \n",
       "28                         NaN  \n",
       "29                         NaN  \n",
       "30                         NaN  "
      ]
     },
     "execution_count": 777,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simeone #.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 778,
   "id": "a56c99fd-fb63-490e-989b-5dcd5169a0d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "simeone.rename(columns={'KRAS': 'KRAS_mutation', 'TP53': 'TP53_type', 'CDKN2A': 'CDKN2A_type', 'SMAD4': 'SMAD4_type', 'Source': 'Tissue'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 779,
   "id": "e54921ae-2d9d-4de5-8b15-ff1d6b17a8f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Patient', 'Tissue', 'Procedure', 'Seq Platform', 'KRAS_mutation',\n",
       "       'TP53_type', 'CDKN2A_type', 'SMAD4_type', 'other genetic features'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 779,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simeone.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 780,
   "id": "f8492708-eab6-40f6-a940-d58c9c7a82a3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P22</td>\n",
       "      <td>Simeone</td>\n",
       "      <td>GSM6204130_P22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P04</td>\n",
       "      <td>Simeone</td>\n",
       "      <td>GSM6204112_P04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID_harmonised  Dataset       Unique_ID Case KRAS_mutation KRAS_type  \\\n",
       "27           P22  Simeone  GSM6204130_P22  NaN           NaN       NaN   \n",
       "28           P04  Simeone  GSM6204112_P04  NaN           NaN       NaN   \n",
       "\n",
       "   TP53_type SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2  \\\n",
       "27       NaN        NaN         NaN          NaN            NaN   \n",
       "28       NaN        NaN         NaN          NaN            NaN   \n",
       "\n",
       "   Germline_FANCC Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "27            NaN             NaN            NaN               NaN        NaN   \n",
       "28            NaN             NaN            NaN               NaN        NaN   \n",
       "\n",
       "    Time_from_diagnosis_to_surgery_(weeks) Cycles Location Grade  Size_(cm)  \\\n",
       "27                                     NaN    NaN      NaN   NaN        NaN   \n",
       "28                                     NaN    NaN      NaN   NaN        NaN   \n",
       "\n",
       "   Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "27   NaN      NaN          NaN                     NaN                  NaN   \n",
       "28   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "   Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "27    NaN     NaN             NaN            NaN   \n",
       "28    NaN     NaN             NaN            NaN   \n",
       "\n",
       "    Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "27                            NaN          NaN                NaN   \n",
       "28                            NaN          NaN                NaN   \n",
       "\n",
       "    Overall Survival  \n",
       "27               NaN  \n",
       "28               NaN  "
      ]
     },
     "execution_count": 780,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset.str.contains('Simeo')].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 781,
   "id": "7e9bfe18-8b1c-4b9e-92b5-d6e4ea8af81f",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Procedure', 'KRAS_mutation',\n",
    "       'TP53_type', 'CDKN2A_type', 'SMAD4_type']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 782,
   "id": "e85432b4-d368-4e6e-970b-7885ca2c9b8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "simeone_map = simeone.set_index('Patient')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 783,
   "id": "44e45e3a-3be9-4dd6-8fb6-5d4deb5fbc05",
   "metadata": {},
   "outputs": [],
   "source": [
    "simeone_map = simeone_map.iloc[:27, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 784,
   "id": "3e3f0c20-2f4a-457a-8e55-1485513ce408",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Seq Platform</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>other genetic features</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P01</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12R</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P02</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P03</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>WES</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P04</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P05</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P06</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P07</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P08</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P09</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P10</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P11</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12L</td>\n",
       "      <td>Missense</td>\n",
       "      <td>LGR</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P12</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>Missense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P13</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>WES</td>\n",
       "      <td>WT</td>\n",
       "      <td>Frameshift</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P14</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>WES</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P15</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Stop gained</td>\n",
       "      <td>WT</td>\n",
       "      <td>MSI-H (MSH6 mut, MSH2 mut)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P16</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P17</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P18</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Loss</td>\n",
       "      <td>WT</td>\n",
       "      <td>germline BRCA-1 mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P19</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Foundation</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Splice site</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>germline BRCA-2 mut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P20</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P21</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P22</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P23</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Resection</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P24</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>Q61K</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P25</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P26</th>\n",
       "      <td>Pancreas</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>G12V</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P27</th>\n",
       "      <td>Liver</td>\n",
       "      <td>Biopsy</td>\n",
       "      <td>Tempus</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Tissue  Procedure Seq Platform KRAS_mutation    TP53_type  \\\n",
       "Patient                                                                \n",
       "P01         Liver     Biopsy       Tempus          G12R  Stop gained   \n",
       "P02         Liver     Biopsy       Tempus          G12D  Splice site   \n",
       "P03      Pancreas  Resection          WES            WT           WT   \n",
       "P04      Pancreas  Resection       Tempus          G12D     Missense   \n",
       "P05      Pancreas  Resection   Foundation          G12D     Missense   \n",
       "P06      Pancreas     Biopsy          NaN           NaN          NaN   \n",
       "P07      Pancreas  Resection       Tempus          G12V           WT   \n",
       "P08      Pancreas  Resection       Tempus          G12D     Missense   \n",
       "P09      Pancreas     Biopsy          NaN           NaN          NaN   \n",
       "P10      Pancreas  Resection       Tempus          G12V           WT   \n",
       "P11         Liver     Biopsy   Foundation          G12L     Missense   \n",
       "P12      Pancreas     Biopsy   Foundation          G12V  Stop gained   \n",
       "P13      Pancreas     Biopsy          WES            WT   Frameshift   \n",
       "P14      Pancreas  Resection          WES            WT           WT   \n",
       "P15      Pancreas  Resection       Tempus          G12V     Missense   \n",
       "P16         Liver     Biopsy       Tempus          G12V     Missense   \n",
       "P17         Liver     Biopsy          NaN           NaN          NaN   \n",
       "P18         Liver     Biopsy   Foundation          G12D     Missense   \n",
       "P19      Pancreas  Resection   Foundation          G12V  Splice site   \n",
       "P20      Pancreas     Biopsy       Tempus          G12D     Missense   \n",
       "P21         Liver     Biopsy       Tempus          G12D     Missense   \n",
       "P22      Pancreas     Biopsy          NaN           NaN          NaN   \n",
       "P23      Pancreas  Resection       Tempus          G12D     Missense   \n",
       "P24         Liver     Biopsy       Tempus          Q61K     Missense   \n",
       "P25         Liver     Biopsy       Tempus          G12V           WT   \n",
       "P26      Pancreas     Biopsy       Tempus          G12V           WT   \n",
       "P27         Liver     Biopsy       Tempus            WT           WT   \n",
       "\n",
       "         CDKN2A_type   SMAD4_type      other genetic features  \n",
       "Patient                                                        \n",
       "P01               WT           WT                         NaN  \n",
       "P02               WT           WT                         NaN  \n",
       "P03               WT           WT                         NaN  \n",
       "P04               WT           WT                         NaN  \n",
       "P05               WT  Stop gained                         NaN  \n",
       "P06              NaN          NaN                         NaN  \n",
       "P07               WT           WT                         NaN  \n",
       "P08               WT           WT                         NaN  \n",
       "P09              NaN          NaN                         NaN  \n",
       "P10               WT           WT                         NaN  \n",
       "P11              LGR           WT                         NaN  \n",
       "P12      Stop gained     Missense                         NaN  \n",
       "P13               WT           WT                         NaN  \n",
       "P14               WT           WT                         NaN  \n",
       "P15      Stop gained           WT  MSI-H (MSH6 mut, MSH2 mut)  \n",
       "P16               WT           WT                         NaN  \n",
       "P17              NaN          NaN                         NaN  \n",
       "P18             Loss           WT         germline BRCA-1 mut  \n",
       "P19               WT           WT         germline BRCA-2 mut  \n",
       "P20               WT           WT                         NaN  \n",
       "P21               WT           WT                         NaN  \n",
       "P22              NaN          NaN                         NaN  \n",
       "P23               WT           WT                         NaN  \n",
       "P24               WT           WT                         NaN  \n",
       "P25               WT           WT                         NaN  \n",
       "P26               WT           WT                         NaN  \n",
       "P27               WT           WT                         NaN  "
      ]
     },
     "execution_count": 784,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simeone_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 785,
   "id": "d7c69b52-88f1-4356-907c-5e54729fc507",
   "metadata": {},
   "outputs": [],
   "source": [
    "simeone_map['Unique_ID'] = simeone_map.index.map(dict(zip(df_id_merged[df_id_merged.Dataset.str.contains('Simeo')].ID_harmonised, df_id_merged[df_id_merged.Dataset.str.contains('Simeo')].Unique_ID)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 786,
   "id": "769b1257-5e73-40d3-b711-1a3b852e4798",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "simeone_map = simeone_map[~simeone_map.Unique_ID.isna()].set_index('Unique_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 787,
   "id": "21a0dd31-cdac-4a3a-8542-24f3a77f3f02",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in interesing_cols:\n",
    "    if col in df_id_merged.columns:\n",
    "        df_id_merged[col] = df_id_merged['Unique_ID'].map(simeone_map[col]).fillna(df_id_merged[col])\n",
    "    else:\n",
    "        df_id_merged[col] = df_id_merged['Unique_ID'].map(simeone_map[col])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 788,
   "id": "15374987-03eb-432c-b30f-7e30a9062c00",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>P22</td>\n",
       "      <td>Simeone</td>\n",
       "      <td>GSM6204130_P22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Biopsy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>P04</td>\n",
       "      <td>Simeone</td>\n",
       "      <td>GSM6204112_P04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G12D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense</td>\n",
       "      <td>WT</td>\n",
       "      <td>WT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Resection</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID_harmonised  Dataset       Unique_ID Case KRAS_mutation KRAS_type  \\\n",
       "27           P22  Simeone  GSM6204130_P22  NaN           NaN       NaN   \n",
       "28           P04  Simeone  GSM6204112_P04  NaN          G12D       NaN   \n",
       "\n",
       "   TP53_type SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2  \\\n",
       "27       NaN        NaN         NaN          NaN            NaN   \n",
       "28  Missense         WT          WT          NaN            NaN   \n",
       "\n",
       "   Germline_FANCC Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "27            NaN             NaN            NaN               NaN        NaN   \n",
       "28            NaN             NaN            NaN               NaN        NaN   \n",
       "\n",
       "    Time_from_diagnosis_to_surgery_(weeks) Cycles Location Grade  Size_(cm)  \\\n",
       "27                                     NaN    NaN      NaN   NaN        NaN   \n",
       "28                                     NaN    NaN      NaN   NaN        NaN   \n",
       "\n",
       "   Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "27   NaN      NaN          NaN                     NaN                  NaN   \n",
       "28   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "   Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "27    NaN     NaN             NaN            NaN   \n",
       "28    NaN     NaN             NaN            NaN   \n",
       "\n",
       "    Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "27                            NaN          NaN                NaN   \n",
       "28                            NaN          NaN                NaN   \n",
       "\n",
       "    Overall Survival  Procedure  \n",
       "27               NaN     Biopsy  \n",
       "28               NaN  Resection  "
      ]
     },
     "execution_count": 788,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset.str.contains('Simeo')].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 789,
   "id": "613f948d-7ab8-4912-83bf-063858622bd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Simeone</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G12D</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12V</th>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WT</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G12D</th>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G12R</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G12V</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q61H</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset        Ding  Simeone\n",
       "KRAS_mutation               \n",
       "G12D            NaN      5.0\n",
       "G12V            NaN      6.0\n",
       "WT              NaN      3.0\n",
       "p.G12D          6.0      NaN\n",
       "p.G12R          4.0      NaN\n",
       "p.G12V          4.0      NaN\n",
       "p.Q61H          1.0      NaN"
      ]
     },
     "execution_count": 789,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'KRAS_mutation']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d25833e-027f-4fe1-b988-08e93eca2a65",
   "metadata": {},
   "source": [
    "# Steele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 650,
   "id": "f3034f13-ae52-4b99-bb61-53cc6ff03c4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 651,
   "id": "9ae7dbab-02fa-457a-bf5c-7ef9c77ba998",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ictstr01/groups/ml01/workspace/shrey.parikh/PDAC_Work_Dir/PDAC_Final/Downstream'"
      ]
     },
     "execution_count": 651,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 379,
   "id": "b398faa7-045f-40d0-ba6b-92141048787c",
   "metadata": {},
   "outputs": [],
   "source": [
    "steele_atlas = sc.read_h5ad('../../Steele_Atlas.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 400,
   "id": "3192ac87-51d1-4e4a-b567-5b890d0f9df3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>Count</th>\n",
       "      <th>Study..Citation..PMID.</th>\n",
       "      <th>GSE.SRA..Study.</th>\n",
       "      <th>Name</th>\n",
       "      <th>If.metastatic..location</th>\n",
       "      <th>Clusters</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>DiseaseState</th>\n",
       "      <th>TreatmentType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACTCGTTT-1_1</th>\n",
       "      <td>2480.0</td>\n",
       "      <td>962</td>\n",
       "      <td>12.661290</td>\n",
       "      <td>4</td>\n",
       "      <td>241.0</td>\n",
       "      <td>Lin W, Noel P, Borazanci EH, Lee J et al. Sing...</td>\n",
       "      <td>GSE154778</td>\n",
       "      <td>T1</td>\n",
       "      <td>NA</td>\n",
       "      <td>DUCTAL</td>\n",
       "      <td>Treatment naïve</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGGGTAGG-1_1</th>\n",
       "      <td>690.0</td>\n",
       "      <td>338</td>\n",
       "      <td>2.173913</td>\n",
       "      <td>3</td>\n",
       "      <td>241.0</td>\n",
       "      <td>Lin W, Noel P, Borazanci EH, Lee J et al. Sing...</td>\n",
       "      <td>GSE154778</td>\n",
       "      <td>T1</td>\n",
       "      <td>NA</td>\n",
       "      <td>FIBROBLASTS</td>\n",
       "      <td>Treatment naïve</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    nCount_RNA  nFeature_RNA  percent.mt seurat_clusters  \\\n",
       "AAACATACTCGTTT-1_1      2480.0           962   12.661290               4   \n",
       "AAACCGTGGGTAGG-1_1       690.0           338    2.173913               3   \n",
       "\n",
       "                    Count                             Study..Citation..PMID.  \\\n",
       "AAACATACTCGTTT-1_1  241.0  Lin W, Noel P, Borazanci EH, Lee J et al. Sing...   \n",
       "AAACCGTGGGTAGG-1_1  241.0  Lin W, Noel P, Borazanci EH, Lee J et al. Sing...   \n",
       "\n",
       "                   GSE.SRA..Study. Name If.metastatic..location     Clusters  \\\n",
       "AAACATACTCGTTT-1_1       GSE154778   T1                      NA       DUCTAL   \n",
       "AAACCGTGGGTAGG-1_1       GSE154778   T1                      NA  FIBROBLASTS   \n",
       "\n",
       "                          Treatment   DiseaseState TreatmentType  \n",
       "AAACATACTCGTTT-1_1  Treatment naïve  Primary tumor           NaN  \n",
       "AAACCGTGGGTAGG-1_1  Treatment naïve  Primary tumor           NaN  "
      ]
     },
     "execution_count": 400,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "steele_atlas.obs.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 401,
   "id": "ddf711aa-4cd3-4a80-9e0d-d36aa431ea90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>Count</th>\n",
       "      <th>Study..Citation..PMID.</th>\n",
       "      <th>GSE.SRA..Study.</th>\n",
       "      <th>Name</th>\n",
       "      <th>If.metastatic..location</th>\n",
       "      <th>Clusters</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>DiseaseState</th>\n",
       "      <th>TreatmentType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGCTTCTTGGATG-1_220</th>\n",
       "      <td>6861.0</td>\n",
       "      <td>1695</td>\n",
       "      <td>4.124763</td>\n",
       "      <td>1</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Steele NG, Carpenter ES, Kemp SB, Sirihorachai...</td>\n",
       "      <td>GSE155698</td>\n",
       "      <td>SRR12461530</td>\n",
       "      <td>NA</td>\n",
       "      <td>TNK</td>\n",
       "      <td>Treatment naïve</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGAACAGCGCTGAA-1_220</th>\n",
       "      <td>4993.0</td>\n",
       "      <td>1513</td>\n",
       "      <td>4.686561</td>\n",
       "      <td>1</td>\n",
       "      <td>96.0</td>\n",
       "      <td>Steele NG, Carpenter ES, Kemp SB, Sirihorachai...</td>\n",
       "      <td>GSE155698</td>\n",
       "      <td>SRR12461530</td>\n",
       "      <td>NA</td>\n",
       "      <td>TNK</td>\n",
       "      <td>Treatment naïve</td>\n",
       "      <td>Primary tumor</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        nCount_RNA  nFeature_RNA  percent.mt seurat_clusters  \\\n",
       "AAACGCTTCTTGGATG-1_220      6861.0          1695    4.124763               1   \n",
       "AAAGAACAGCGCTGAA-1_220      4993.0          1513    4.686561               1   \n",
       "\n",
       "                        Count  \\\n",
       "AAACGCTTCTTGGATG-1_220   96.0   \n",
       "AAAGAACAGCGCTGAA-1_220   96.0   \n",
       "\n",
       "                                                   Study..Citation..PMID.  \\\n",
       "AAACGCTTCTTGGATG-1_220  Steele NG, Carpenter ES, Kemp SB, Sirihorachai...   \n",
       "AAAGAACAGCGCTGAA-1_220  Steele NG, Carpenter ES, Kemp SB, Sirihorachai...   \n",
       "\n",
       "                       GSE.SRA..Study.         Name If.metastatic..location  \\\n",
       "AAACGCTTCTTGGATG-1_220       GSE155698  SRR12461530                      NA   \n",
       "AAAGAACAGCGCTGAA-1_220       GSE155698  SRR12461530                      NA   \n",
       "\n",
       "                       Clusters        Treatment   DiseaseState TreatmentType  \n",
       "AAACGCTTCTTGGATG-1_220      TNK  Treatment naïve  Primary tumor            NA  \n",
       "AAAGAACAGCGCTGAA-1_220      TNK  Treatment naïve  Primary tumor            NA  "
      ]
     },
     "execution_count": 401,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "steele_atlas.obs[steele_atlas.obs['GSE.SRA..Study.'].str.contains('GSE155698')].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 403,
   "id": "77ae70d4-8637-4be0-a393-83dfc0c80906",
   "metadata": {},
   "outputs": [],
   "source": [
    "steele = sc.read_h5ad('/lustre/groups/ml01/workspace/shrey.parikh/PDAC/raw_data/Steele/PDAC_TISSUE_1/filtered_feature_bc_matrix_output.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "id": "da1534c5-37ed-4f0c-8fef-93a4e5132c67",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGAAAGTGGAAAG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTAGGGTAC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTCATAGTC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGCTGTAATCAGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGAACCATTAAAGG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCGATGTCGCATCG-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGATCAGAGAGTGA-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGAGCAGCAAGAC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGGTTAGTATCTGC-1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTTGAGCAGAAAG-1</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1633 rows × 0 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AAACGAAAGTGGAAAG-1, AAACGAAGTAGGGTAC-1, AAACGAAGTCATAGTC-1, AAACGCTGTAATCAGA-1, AAAGAACCATTAAAGG-1, AAAGGATTCGGCTTGG-1, AAAGGGCAGTAGCAAT-1, AAAGGGCAGTGAATAC-1, AAAGGGCCATGAAGGC-1, AAAGGTACAATAGTAG-1, AAAGGTACACATACGT-1, AAAGTCCTCAAGCCTA-1, AAAGTGAGTCACCTTC-1, AACAACCGTCTTCAAG-1, AACAACCGTTGCCTAA-1, AACAACCTCTTGTTAC-1, AACAAGAAGGGTTAGC-1, AACAAGAGTGTATCCA-1, AACAAGATCCATGAGT-1, AACACACGTTTGTTCT-1, AACCAACAGCGCTTCG-1, AACCAACCACTGGCCA-1, AACCACACACTTGAAC-1, AACCATGGTCATCCGG-1, AACCCAACACTAGGCC-1, AACCTGACATCTCATT-1, AACCTGAGTACACGCC-1, AACCTGAGTCGGCTAC-1, AACCTTTCAACCGTGC-1, AACCTTTGTATGACAA-1, AACCTTTTCATTTGTC-1, AACCTTTTCTGCTCTG-1, AACGAAAGTAGTCGTT-1, AACGGGACAGCTCCTT-1, AACGGGAGTGTTGCCG-1, AACGGGAGTTCCTAAG-1, AACGTCAGTTGTTTGG-1, AACGTCATCGGACGTC-1, AACTTCTCACGGCGTT-1, AAGAACAAGTCTTGGT-1, AAGAACAGTAGACGTG-1, AAGACAAAGCTGTGCC-1, AAGACAATCACCCTGT-1, AAGACTCCATCAGCAT-1, AAGACTCTCCCGAATA-1, AAGATAGAGCCTGTCG-1, AAGATAGGTCTCCCTA-1, AAGCATCGTTACACAC-1, AAGCCATAGAAACCCG-1, AAGCCATCATCACAGT-1, AAGCCATCATTGCTGA-1, AAGCCATGTTAGAAGT-1, AAGCGAGAGTGATTCC-1, AAGCGTTGTTAAGAAC-1, AAGCGTTTCACCATAG-1, AAGGAATAGTGTACCT-1, AAGGTAACACTTCCTG-1, AAGTACCCAAGTGTCT-1, AAGTACCCACAAACGG-1, AAGTACCGTGCACATT-1, AAGTCGTAGCGCCTTG-1, AAGTCGTAGTAGCAAT-1, AAGTTCGAGTGGCGAT-1, AAGTTCGCACTTCCTG-1, AAGTTCGTCGGCCAAC-1, AATAGAGAGTTGTCAC-1, AATAGAGCAGGATCTT-1, AATCACGCACCCTGTT-1, AATCACGTCACGGGAA-1, AATCACGTCCCAAGTA-1, AATCGACAGCACTCAT-1, AATCGACGTCGAGCTC-1, AATCGTGCAGTAACCT-1, AATGACCCAGAAATTG-1, AATGACCTCTATGCCC-1, AATGCCATCATGGAGG-1, AATGGAAAGAACCCGA-1, AATGGCTGTATAGGAT-1, AATGGCTGTTAATGAG-1, AATTCCTCATGAAGGC-1, AATTCCTGTCGAATTC-1, AATTCCTGTTTCGATG-1, AATTTCCGTTCGAAGG-1, AATTTCCTCGGCTTGG-1, ACAAAGAAGTAGAATC-1, ACAAAGACAGGATTCT-1, ACAACCACACATCATG-1, ACAACCACACATTCGA-1, ACAAGCTAGGGATGTC-1, ACAAGCTGTCGCCACA-1, ACACCAAAGCCGCACT-1, ACACCAAGTATGGGAC-1, ACACCAATCGCTGATA-1, ACACGCGGTATACCCA-1, ACACTGAGTGCGACAA-1, ACAGCCGTCCCTCTTT-1, ACATCCCAGCGTGCTC-1, ACATCGAGTGGTCTGC-1, ACATCGATCTGGGCCA-1, ACATGCAAGCTGTGCC-1, ...]\n",
       "\n",
       "[1633 rows x 0 columns]"
      ]
     },
     "execution_count": 407,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "steele.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 409,
   "id": "ecb23d0c-6427-4169-b53b-99ad652ad676",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Steele atlas is useless again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "id": "ce6c49ab-60d9-465b-920b-90292ab9ab23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the zenodo "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 420,
   "id": "d612526d-888a-46f7-aa7f-d0a1940fdeba",
   "metadata": {},
   "outputs": [],
   "source": [
    "zenodo = sc.read_h5ad('/lustre/groups/ml01/workspace/shrey.parikh/PDAC/raw_data/Zenodo/Data/pk_all.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 425,
   "id": "b265368d-fdfb-412a-a863-d158b2cefdca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CELL                                   50943\n",
       "Patient                                    0\n",
       "Type                                       0\n",
       "Cell_type                                  0\n",
       "celltype0                              50943\n",
       "celltype1                              50943\n",
       "celltype2                              50943\n",
       "celltype3                              50943\n",
       "Patient2                                   0\n",
       "nCount_RNA                                 0\n",
       "nFeature_RNA                               0\n",
       "Project                                    0\n",
       "orig.ident                                 0\n",
       "Stage                                  50943\n",
       "Grade                                  50943\n",
       "Gender                                 50943\n",
       "Age                                    50943\n",
       "Percent_mito                               0\n",
       "Percent_ribo                               0\n",
       "Percent_hemo                               0\n",
       "percent.mt                             50943\n",
       "predicted.id                               0\n",
       "prediction.score.Fibroblast.cell           0\n",
       "prediction.score.Stellate.cell             0\n",
       "prediction.score.Macrophage.cell           0\n",
       "prediction.score.Endothelial.cell          0\n",
       "prediction.score.T.cell                    0\n",
       "prediction.score.B.cell                    0\n",
       "prediction.score.Ductal.cell.type.2        0\n",
       "prediction.score.Endocrine.cell            0\n",
       "prediction.score.Ductal.cell.type.1        0\n",
       "prediction.score.Acinar.cell               0\n",
       "prediction.score.max                       0\n",
       "classical_score1                           0\n",
       "basal_score1                               0\n",
       "classical_score21                          0\n",
       "basal_score21                              0\n",
       "endocrine_score_1                          0\n",
       "immune_score_1                             0\n",
       "exocrine_score_1                           0\n",
       "activated_stroma_score_1                   0\n",
       "histone_score_1                            0\n",
       "normal_stroma_score_1                      0\n",
       "olfactory_score_1                          0\n",
       "na_score_1                                 0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 425,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zenodo.obs[zenodo.obs.Project == 'GSE155698'].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40fe0950-b5ca-47f2-b122-67039eff9512",
   "metadata": {},
   "source": [
    "### PDAC\n",
    "- All samples are treatment naive\n",
    "- Average age 60.2 (47-72)\n",
    "- Std Dev BMI 23.5 (4.24)\n",
    "### Adj_Norm\n",
    "- Average age 66.4 (45-81)\n",
    "- Std Dev BMI 25 (5.54)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 790,
   "id": "23535419-1a8e-4b37-870d-97636466e4b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = [\n",
    "    (df_id_merged['Dataset'] == 'Steele'),\n",
    "    (df_id_merged['Dataset'] == 'Steele_Adj_Norm')\n",
    "]\n",
    "values = [60.2, 66.4]\n",
    "df_id_merged['Age'] = np.select(conditions, values, default=np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 791,
   "id": "3819fe89-618e-4332-b866-247ab25769c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = [\n",
    "    (df_id_merged['Dataset'] == 'Steele'),\n",
    "    (df_id_merged['Dataset'] == 'Steele_Adj_Norm')\n",
    "]\n",
    "values = [23.5, 25]\n",
    "df_id_merged['BMI'] = np.select(conditions, values, default=np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 792,
   "id": "8f1c616d-5a5a-42dc-9e0b-66cbf7ed1e1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged['Treatment'] = np.where(df_id_merged.Dataset == 'Steele', 'Treatment Naive', np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 793,
   "id": "abe415d4-0239-479c-bb5d-d0675694047e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>BMI</th>\n",
       "      <th>23.5</th>\n",
       "      <th>25.0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Steele</th>\n",
       "      <td>16.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "BMI              23.5  25.0\n",
       "Dataset                    \n",
       "Steele           16.0   NaN\n",
       "Steele_Adj_Norm   NaN   3.0"
      ]
     },
     "execution_count": 793,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'BMI']).size().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 794,
   "id": "415eea0c-2238-4eb5-8315-b6ce2b1355e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Treatment</th>\n",
       "      <th>Treatment Naive</th>\n",
       "      <th>nan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Caronni</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding</th>\n",
       "      <td>NaN</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hwang</th>\n",
       "      <td>NaN</td>\n",
       "      <td>43.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lee</th>\n",
       "      <td>NaN</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lin</th>\n",
       "      <td>NaN</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng</th>\n",
       "      <td>NaN</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_Normal</th>\n",
       "      <td>NaN</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schlesinger</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Simeone</th>\n",
       "      <td>NaN</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele</th>\n",
       "      <td>16.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Treatment        Treatment Naive   nan\n",
       "Dataset                               \n",
       "Caronni                      NaN  10.0\n",
       "Ding                         NaN  22.0\n",
       "Ding_snRNA-seq               NaN  11.0\n",
       "Hwang                        NaN  43.0\n",
       "Lee                          NaN   5.0\n",
       "Lin                          NaN  10.0\n",
       "Peng                         NaN  24.0\n",
       "Peng_Normal                  NaN  11.0\n",
       "Schlesinger                  NaN   1.0\n",
       "Simeone                      NaN  17.0\n",
       "Steele                      16.0   NaN\n",
       "Steele_Adj_Norm              NaN   3.0"
      ]
     },
     "execution_count": 794,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'Treatment']).size().unstack()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e281ac41-5998-4283-bcdf-5589a6981648",
   "metadata": {},
   "source": [
    "# Lee"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44d4bcca-5a1e-4369-b909-871a5fa89743",
   "metadata": {},
   "source": [
    " P2-P5 primary samples came from treatment-naïve patients, while P1 was a second FNA from a patient who had been treated with gemcitabine/paclitaxel. \n",
    "For metastatic samples, the most recent therapies prior to sample acquisition were 5-fluorouracil/liposmal irinotecan, evofosfamide/ipilimumab and \n",
    "capecitabine for VM, LuM, and PM, respectively. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "id": "2cf70b71-f469-46cc-b32c-06f28b0182ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>P1</td>\n",
       "      <td>Lee</td>\n",
       "      <td>P1_filtered_feature_bc_matrix</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>P3</td>\n",
       "      <td>Lee</td>\n",
       "      <td>P3_filtered_feature_bc_matrix</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>P4</td>\n",
       "      <td>Lee</td>\n",
       "      <td>P4_filtered_feature_bc_matrix</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>P5</td>\n",
       "      <td>Lee</td>\n",
       "      <td>P5_filtered_feature_bc_matrix</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>P2</td>\n",
       "      <td>Lee</td>\n",
       "      <td>P2_filtered_feature_bc_matrix</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID_harmonised Dataset                      Unique_ID Case KRAS_mutation  \\\n",
       "22            P1     Lee  P1_filtered_feature_bc_matrix  NaN           NaN   \n",
       "23            P3     Lee  P3_filtered_feature_bc_matrix  NaN           NaN   \n",
       "24            P4     Lee  P4_filtered_feature_bc_matrix  NaN           NaN   \n",
       "25            P5     Lee  P5_filtered_feature_bc_matrix  NaN           NaN   \n",
       "26            P2     Lee  P2_filtered_feature_bc_matrix  NaN           NaN   \n",
       "\n",
       "   KRAS_type TP53_type SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2  \\\n",
       "22       NaN       NaN        NaN         NaN          NaN            NaN   \n",
       "23       NaN       NaN        NaN         NaN          NaN            NaN   \n",
       "24       NaN       NaN        NaN         NaN          NaN            NaN   \n",
       "25       NaN       NaN        NaN         NaN          NaN            NaN   \n",
       "26       NaN       NaN        NaN         NaN          NaN            NaN   \n",
       "\n",
       "   Germline_FANCC Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "22            NaN             NaN            NaN               NaN        NaN   \n",
       "23            NaN             NaN            NaN               NaN        NaN   \n",
       "24            NaN             NaN            NaN               NaN        NaN   \n",
       "25            NaN             NaN            NaN               NaN        NaN   \n",
       "26            NaN             NaN            NaN               NaN        NaN   \n",
       "\n",
       "    Time_from_diagnosis_to_surgery_(weeks) Cycles Location  Grade  Size_(cm)  \\\n",
       "22                                     NaN    NaN      NaN    NaN        NaN   \n",
       "23                                     NaN    NaN      NaN    NaN        NaN   \n",
       "24                                     NaN    NaN      NaN    NaN        NaN   \n",
       "25                                     NaN    NaN      NaN    NaN        NaN   \n",
       "26                                     NaN    NaN      NaN    NaN        NaN   \n",
       "\n",
       "   Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "22   NaN      NaN          NaN                     NaN                  NaN   \n",
       "23   NaN      NaN          NaN                     NaN                  NaN   \n",
       "24   NaN      NaN          NaN                     NaN                  NaN   \n",
       "25   NaN      NaN          NaN                     NaN                  NaN   \n",
       "26   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "   Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "22    NaN     NaN             NaN            NaN   \n",
       "23    NaN     NaN             NaN            NaN   \n",
       "24    NaN     NaN             NaN            NaN   \n",
       "25    NaN     NaN             NaN            NaN   \n",
       "26    NaN     NaN             NaN            NaN   \n",
       "\n",
       "    Survival_post_surgery_(weeks) Vital_status  Overall Survival Procedure  \n",
       "22                            NaN          NaN               NaN       NaN  \n",
       "23                            NaN          NaN               NaN       NaN  \n",
       "24                            NaN          NaN               NaN       NaN  \n",
       "25                            NaN          NaN               NaN       NaN  \n",
       "26                            NaN          NaN               NaN       NaN  "
      ]
     },
     "execution_count": 376,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset == 'Lee']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d75b7121-e845-4eae-bf4a-7b4f0602d03a",
   "metadata": {},
   "source": [
    "### Added already by Daniele"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e12e0d52-af40-4158-a8b3-e13368b5e6fa",
   "metadata": {},
   "source": [
    "# Peng"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "id": "2e2d6ee7-af9f-46ac-ab1d-5cb7e89ce883",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting xlrd\n",
      "  Downloading xlrd-2.0.1-py2.py3-none-any.whl.metadata (3.4 kB)\n",
      "Downloading xlrd-2.0.1-py2.py3-none-any.whl (96 kB)\n",
      "Installing collected packages: xlrd\n",
      "Successfully installed xlrd-2.0.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install xlrd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 795,
   "id": "aaaf9fd6-7732-4ffe-8ce0-098d72cf4c7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "peng = pd.read_excel('metadata_publications/Peng_metadata.xlsx', header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 796,
   "id": "ad269450-2a86-49e8-ac46-daac236ef6aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Number', 'pathologic diagnoses', 'Gender', 'Age(y/o)', 'CA199(U/ml)',\n",
       "       'Diabetes Mellitus', 'Procedure', 'Location', 'Maximum Diameter(mm)',\n",
       "       'TNM Classification', 'Staging', 'perineural invasion',\n",
       "       'vascular invasion', 'peripancreatic infiltarion', 'note',\n",
       "       'Unnamed: 15'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 796,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 797,
   "id": "ee55bbb2-6245-487c-ab85-a317afad4723",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number</th>\n",
       "      <th>pathologic diagnoses</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age(y/o)</th>\n",
       "      <th>CA199(U/ml)</th>\n",
       "      <th>Diabetes Mellitus</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Location</th>\n",
       "      <th>Maximum Diameter(mm)</th>\n",
       "      <th>TNM Classification</th>\n",
       "      <th>Staging</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PUMCH-T1</td>\n",
       "      <td>moderately-poorly differentiated PDAC</td>\n",
       "      <td>M</td>\n",
       "      <td>64.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>N</td>\n",
       "      <td>LDP</td>\n",
       "      <td>body</td>\n",
       "      <td>26.0</td>\n",
       "      <td>T4N2M0</td>\n",
       "      <td>III</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PUMCH-T2</td>\n",
       "      <td>well differentiated PDAC</td>\n",
       "      <td>M</td>\n",
       "      <td>52.0</td>\n",
       "      <td>46.3</td>\n",
       "      <td>N</td>\n",
       "      <td>PD</td>\n",
       "      <td>head</td>\n",
       "      <td>20.0</td>\n",
       "      <td>T1cN1M0</td>\n",
       "      <td>IIB</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Number                   pathologic diagnoses Gender  Age(y/o)  \\\n",
       "0  PUMCH-T1  moderately-poorly differentiated PDAC      M      64.0   \n",
       "1  PUMCH-T2               well differentiated PDAC      M      52.0   \n",
       "\n",
       "  CA199(U/ml) Diabetes Mellitus Procedure Location  Maximum Diameter(mm)  \\\n",
       "0        86.0                 N       LDP     body                  26.0   \n",
       "1        46.3                 N        PD     head                  20.0   \n",
       "\n",
       "  TNM Classification Staging perineural invasion vascular invasion  \\\n",
       "0             T4N2M0     III                   Y                 Y   \n",
       "1            T1cN1M0     IIB                   Y                 N   \n",
       "\n",
       "  peripancreatic infiltarion note Unnamed: 15  \n",
       "0                          Y  NaN         NaN  \n",
       "1                          Y  NaN         NaN  "
      ]
     },
     "execution_count": 797,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 798,
   "id": "50e0278c-47dd-4676-89fd-d04a89383ce7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>T1</td>\n",
       "      <td>Peng</td>\n",
       "      <td>T1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>T2</td>\n",
       "      <td>Peng</td>\n",
       "      <td>T2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID_harmonised Dataset Unique_ID Case KRAS_mutation KRAS_type TP53_type  \\\n",
       "60            T1    Peng        T1  NaN           NaN       NaN       NaN   \n",
       "61            T2    Peng        T2  NaN           NaN       NaN       NaN   \n",
       "\n",
       "   SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "60        NaN         NaN          NaN            NaN            NaN   \n",
       "61        NaN         NaN          NaN            NaN            NaN   \n",
       "\n",
       "   Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "60             NaN            NaN               NaN        NaN   \n",
       "61             NaN            NaN               NaN        NaN   \n",
       "\n",
       "    Time_from_diagnosis_to_surgery_(weeks) Cycles Location Grade  Size_(cm)  \\\n",
       "60                                     NaN    NaN      NaN   NaN        NaN   \n",
       "61                                     NaN    NaN      NaN   NaN        NaN   \n",
       "\n",
       "   Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "60   NaN      NaN          NaN                     NaN                  NaN   \n",
       "61   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "   Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "60    NaN     NaN             NaN            NaN   \n",
       "61    NaN     NaN             NaN            NaN   \n",
       "\n",
       "    Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "60                            NaN          NaN                NaN   \n",
       "61                            NaN          NaN                NaN   \n",
       "\n",
       "    Overall Survival Procedure  Age  BMI Treatment  \n",
       "60               NaN       NaN  NaN  NaN       nan  \n",
       "61               NaN       NaN  NaN  NaN       nan  "
      ]
     },
     "execution_count": 798,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset == 'Peng'].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 799,
   "id": "f4f59a74-879a-4758-abfc-f6093552bfaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "peng.rename(columns={'pathologic diagnoses': 'Pathology_notes', 'Staging': 'Stage', 'TNM Classification': 'TNM_classification', 'Age(y/o)': 'Age'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 800,
   "id": "e1979ad9-e8fc-4f85-9247-b12e2e867932",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Number', 'Pathology_notes', 'Gender', 'Age', 'CA199(U/ml)',\n",
       "       'Diabetes Mellitus', 'Procedure', 'Location', 'Maximum Diameter(mm)',\n",
       "       'TNM_classification', 'Stage', 'perineural invasion',\n",
       "       'vascular invasion', 'peripancreatic infiltarion', 'note',\n",
       "       'Unnamed: 15'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 800,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 801,
   "id": "dc81d106-6ac8-478d-b175-092ecdaa4d44",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Pathology_notes', 'Age', 'Procedure', \n",
    "       'TNM_classification',  'Stage', 'perineural invasion',\n",
    "       'vascular invasion', 'peripancreatic infiltarion']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 802,
   "id": "e9dc3fa9-1896-4ddf-945f-5ee2089aadff",
   "metadata": {},
   "outputs": [],
   "source": [
    "peng[['Random', 'Unique_ID']] = peng.Number.str.split('-', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 803,
   "id": "0acdb91d-cf27-479a-8f1c-e75ee4ac312b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged.ID_harmonised = np.where(df_id_merged.Dataset.str.contains('Peng'), 'Peng' + '_' + df_id_merged.ID_harmonised, df_id_merged.ID_harmonised)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 804,
   "id": "0229fd0e-9d15-4b54-9b7a-3b79b9a8a84d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_id_merged.ID_harmonised = df_id_merged.ID_harmonised.str.replace(r'^Peng_Peng_Peng_', 'Peng_', regex=True)\n",
    "\n",
    "# df_id_merged.ID_harmonised = df_id_merged.ID_harmonised.str.replace(r'^Peng_Peng_', 'Peng_', regex=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 805,
   "id": "c57ebc8c-bf6f-4f17-844f-ae8e560fb4ef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Treatment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Peng_T1</td>\n",
       "      <td>Peng</td>\n",
       "      <td>T1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>Peng_T2</td>\n",
       "      <td>Peng</td>\n",
       "      <td>T2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID_harmonised Dataset Unique_ID Case KRAS_mutation KRAS_type TP53_type  \\\n",
       "60       Peng_T1    Peng        T1  NaN           NaN       NaN       NaN   \n",
       "61       Peng_T2    Peng        T2  NaN           NaN       NaN       NaN   \n",
       "\n",
       "   SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "60        NaN         NaN          NaN            NaN            NaN   \n",
       "61        NaN         NaN          NaN            NaN            NaN   \n",
       "\n",
       "   Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "60             NaN            NaN               NaN        NaN   \n",
       "61             NaN            NaN               NaN        NaN   \n",
       "\n",
       "    Time_from_diagnosis_to_surgery_(weeks) Cycles Location Grade  Size_(cm)  \\\n",
       "60                                     NaN    NaN      NaN   NaN        NaN   \n",
       "61                                     NaN    NaN      NaN   NaN        NaN   \n",
       "\n",
       "   Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "60   NaN      NaN          NaN                     NaN                  NaN   \n",
       "61   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "   Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "60    NaN     NaN             NaN            NaN   \n",
       "61    NaN     NaN             NaN            NaN   \n",
       "\n",
       "    Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "60                            NaN          NaN                NaN   \n",
       "61                            NaN          NaN                NaN   \n",
       "\n",
       "    Overall Survival Procedure  Age  BMI Treatment  \n",
       "60               NaN       NaN  NaN  NaN       nan  \n",
       "61               NaN       NaN  NaN  NaN       nan  "
      ]
     },
     "execution_count": 805,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset.str.contains('Peng')].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 806,
   "id": "672181f5-2011-4aff-b0d5-285f2674540e",
   "metadata": {},
   "outputs": [],
   "source": [
    "peng.Unique_ID = 'Peng' + '_' + peng.Unique_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 807,
   "id": "eca7f750-bda2-4f94-bd08-eb9582f129cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41"
      ]
     },
     "execution_count": 807,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(peng.Unique_ID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 808,
   "id": "83e210fa-bd72-4f50-a7a8-5b67db70a2ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 808,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(peng.Unique_ID) & set(df_id_merged.ID_harmonised))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 809,
   "id": "3e7b00c2-2a6d-459b-b318-ffff6a7c72f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "peng_map = peng.iloc[:40, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 810,
   "id": "a70a41f4-01e1-4f63-9747-db55d1ce13a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "35"
      ]
     },
     "execution_count": 810,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(peng_map.Unique_ID) & set(df_id_merged.ID_harmonised))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 811,
   "id": "77211397-8fc3-4219-8dd4-fd667cf97d8d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "peng_map = peng_map[~peng_map.Unique_ID.isna()].set_index('Unique_ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 812,
   "id": "c0ae01f7-c78e-4f7a-8c5f-de5f6dbb8525",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Number</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>CA199(U/ml)</th>\n",
       "      <th>Diabetes Mellitus</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Location</th>\n",
       "      <th>Maximum Diameter(mm)</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Stage</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "      <th>note</th>\n",
       "      <th>Unnamed: 15</th>\n",
       "      <th>Random</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unique_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Peng_T1</th>\n",
       "      <td>PUMCH-T1</td>\n",
       "      <td>moderately-poorly differentiated PDAC</td>\n",
       "      <td>M</td>\n",
       "      <td>64.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>N</td>\n",
       "      <td>LDP</td>\n",
       "      <td>body</td>\n",
       "      <td>26.0</td>\n",
       "      <td>T4N2M0</td>\n",
       "      <td>III</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PUMCH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_T2</th>\n",
       "      <td>PUMCH-T2</td>\n",
       "      <td>well differentiated PDAC</td>\n",
       "      <td>M</td>\n",
       "      <td>52.0</td>\n",
       "      <td>46.3</td>\n",
       "      <td>N</td>\n",
       "      <td>PD</td>\n",
       "      <td>head</td>\n",
       "      <td>20.0</td>\n",
       "      <td>T1cN1M0</td>\n",
       "      <td>IIB</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PUMCH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Number                        Pathology_notes Gender   Age  \\\n",
       "Unique_ID                                                                 \n",
       "Peng_T1    PUMCH-T1  moderately-poorly differentiated PDAC      M  64.0   \n",
       "Peng_T2    PUMCH-T2               well differentiated PDAC      M  52.0   \n",
       "\n",
       "          CA199(U/ml) Diabetes Mellitus Procedure Location  \\\n",
       "Unique_ID                                                    \n",
       "Peng_T1          86.0                 N       LDP     body   \n",
       "Peng_T2          46.3                 N        PD     head   \n",
       "\n",
       "           Maximum Diameter(mm) TNM_classification Stage perineural invasion  \\\n",
       "Unique_ID                                                                      \n",
       "Peng_T1                    26.0             T4N2M0   III                   Y   \n",
       "Peng_T2                    20.0            T1cN1M0   IIB                   Y   \n",
       "\n",
       "          vascular invasion peripancreatic infiltarion note Unnamed: 15 Random  \n",
       "Unique_ID                                                                       \n",
       "Peng_T1                   Y                          Y  NaN         NaN  PUMCH  \n",
       "Peng_T2                   N                          Y  NaN         NaN  PUMCH  "
      ]
     },
     "execution_count": 812,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng_map.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 815,
   "id": "569f2f3d-7e97-4703-8609-a0bd3adcbca4",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in interesing_cols:\n",
    "    if col in df_id_merged.columns:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(peng_map[col]).fillna(df_id_merged[col])\n",
    "    else:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(peng_map[col])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 816,
   "id": "9ffcfa0b-9072-40ed-bef5-5c302bde857a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNM_classification</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1cN0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1cN1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1cN2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T4N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT1a pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN1</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN2</th>\n",
       "      <td>4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT3 pN2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpT2 moderately differentiated pN1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yT2 pN1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1 moderately differentiated positive for perineural invasion negative for lymphovascular pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1aN0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1aN1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1cN0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1cN1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN0</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN1</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN2</th>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2N2M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN2</th>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3N0M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3N1M0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                                             Ding  Ding_snRNA-seq  \\\n",
       "TNM_classification                                                         \n",
       "T1cN0M0                                              NaN             NaN   \n",
       "T1cN1M0                                              NaN             NaN   \n",
       "T1cN2M0                                              NaN             NaN   \n",
       "T2N0M0                                               NaN             NaN   \n",
       "T2N1M0                                               NaN             NaN   \n",
       "T2N2M0                                               NaN             NaN   \n",
       "T3N0M0                                               NaN             NaN   \n",
       "T3N1M0                                               NaN             NaN   \n",
       "T3N2M0                                               NaN             NaN   \n",
       "T4N2M0                                               NaN             NaN   \n",
       "pT1a pN0                                             NaN             1.0   \n",
       "pT2 pN0                                              NaN             2.0   \n",
       "pT2 pN1                                              2.0             1.0   \n",
       "pT2 pN2                                              4.0             NaN   \n",
       "pT3 pN2                                              1.0             NaN   \n",
       "rpT2 moderately differentiated pN1                   NaN             1.0   \n",
       "yT2 pN1                                              1.0             NaN   \n",
       "ypT1 moderately differentiated positive for per...   NaN             1.0   \n",
       "ypT1aN0M0                                            NaN             NaN   \n",
       "ypT1aN1M0                                            NaN             NaN   \n",
       "ypT1cN0M0                                            NaN             NaN   \n",
       "ypT1cN1M0                                            NaN             NaN   \n",
       "ypT2 pN0                                             3.0             1.0   \n",
       "ypT2 pN1                                             5.0             1.0   \n",
       "ypT2 pN2                                             3.0             NaN   \n",
       "ypT2N0M0                                             NaN             NaN   \n",
       "ypT2N1M0                                             NaN             NaN   \n",
       "ypT2N2M0                                             NaN             NaN   \n",
       "ypT3 pN0                                             NaN             2.0   \n",
       "ypT3 pN2                                             2.0             1.0   \n",
       "ypT3N0M0                                             NaN             NaN   \n",
       "ypT3N1M0                                             NaN             NaN   \n",
       "\n",
       "Dataset                                             Hwang  Peng  Peng_Normal  \n",
       "TNM_classification                                                            \n",
       "T1cN0M0                                               NaN   1.0          NaN  \n",
       "T1cN1M0                                               NaN   2.0          NaN  \n",
       "T1cN2M0                                               NaN   1.0          NaN  \n",
       "T2N0M0                                                2.0   8.0          NaN  \n",
       "T2N1M0                                                6.0   6.0          NaN  \n",
       "T2N2M0                                                1.0   NaN          NaN  \n",
       "T3N0M0                                                3.0   1.0          1.0  \n",
       "T3N1M0                                                3.0   3.0          NaN  \n",
       "T3N2M0                                                3.0   1.0          NaN  \n",
       "T4N2M0                                                NaN   1.0          NaN  \n",
       "pT1a pN0                                              NaN   NaN          NaN  \n",
       "pT2 pN0                                               NaN   NaN          NaN  \n",
       "pT2 pN1                                               NaN   NaN          NaN  \n",
       "pT2 pN2                                               NaN   NaN          NaN  \n",
       "pT3 pN2                                               NaN   NaN          NaN  \n",
       "rpT2 moderately differentiated pN1                    NaN   NaN          NaN  \n",
       "yT2 pN1                                               NaN   NaN          NaN  \n",
       "ypT1 moderately differentiated positive for per...    NaN   NaN          NaN  \n",
       "ypT1aN0M0                                             2.0   NaN          NaN  \n",
       "ypT1aN1M0                                             1.0   NaN          NaN  \n",
       "ypT1cN0M0                                             2.0   NaN          NaN  \n",
       "ypT1cN1M0                                             1.0   NaN          NaN  \n",
       "ypT2 pN0                                              NaN   NaN          NaN  \n",
       "ypT2 pN1                                              NaN   NaN          NaN  \n",
       "ypT2 pN2                                              NaN   NaN          NaN  \n",
       "ypT2N0M0                                              6.0   NaN          NaN  \n",
       "ypT2N1M0                                              2.0   NaN          NaN  \n",
       "ypT2N2M0                                              1.0   NaN          NaN  \n",
       "ypT3 pN0                                              NaN   NaN          NaN  \n",
       "ypT3 pN2                                              NaN   NaN          NaN  \n",
       "ypT3N0M0                                              8.0   NaN          NaN  \n",
       "ypT3N1M0                                              2.0   NaN          NaN  "
      ]
     },
     "execution_count": 816,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'TNM_classification']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5bff289-4a8b-4000-9225-b0aa53c8b63c",
   "metadata": {},
   "source": [
    "# Caronni"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 818,
   "id": "49e2f1a0-461d-4be7-b0f5-45f76e06dc9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Ding', 'Lee', 'Simeone', 'Steele', 'Peng', 'Caronni', 'Lin',\n",
       "       'Schlesinger', 'Steele_Adj_Norm', 'Peng_Normal', 'Hwang',\n",
       "       'Ding_snRNA-seq'], dtype=object)"
      ]
     },
     "execution_count": 818,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.Dataset.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 825,
   "id": "5ca1279d-5e21-4058-b9d2-262b24c4a64d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# columns_to_load = ['ID', 'age', 'sex', 'Site of lesion', 'Chemotherapy schedule', 'Grading', 'TNM']\n",
    "caronni = pd.read_excel('metadata_publications/Caronni_metadata.xlsx', \n",
    "                         sheet_name='Patients clinical data')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 836,
   "id": "d12ed6f0-f144-4655-a16a-cb53c34e0505",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      head, uncinate_process\n",
       "1      head, uncinate_process\n",
       "2                        head\n",
       "3                        head\n",
       "4                        head\n",
       "                ...          \n",
       "168                      Head\n",
       "169          neck/body margin\n",
       "170                      Head\n",
       "171                      Head\n",
       "172             Head/Uncinate\n",
       "Name: Location, Length: 173, dtype: object"
      ]
     },
     "execution_count": 836,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.Location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 831,
   "id": "4929441e-bb17-46b4-85c8-e4d278f484ec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "caronni = caronni[~caronni.ID.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 837,
   "id": "cf34affb-c142-4ec1-8da0-dd180add724a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2834644/4079294514.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  caronni.rename(columns={'Site of lesion':'Location', 'TNM':'TNM_classification', 'Grading': 'Grade'}, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "caronni.rename(columns={'Site of lesion':'Location', 'TNM':'TNM_classification', 'Grading': 'Grade'}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 838,
   "id": "0ae95027-03a8-46ec-af8f-d18706e5e669",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>age</th>\n",
       "      <th>sex</th>\n",
       "      <th>Location</th>\n",
       "      <th>Chemotherapy schedule</th>\n",
       "      <th>Grade</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Down Stream Analysis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PDAC47</td>\n",
       "      <td>78.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Head</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>G2</td>\n",
       "      <td>T2; N2; M0</td>\n",
       "      <td>Tumor Resection (fresh)</td>\n",
       "      <td>scRNA Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PDAC48</td>\n",
       "      <td>61.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Head</td>\n",
       "      <td>GEMCITABINA + NAB PACLITAXEL</td>\n",
       "      <td>G2</td>\n",
       "      <td>T2; N2; M0</td>\n",
       "      <td>Tumor Resection (fresh)</td>\n",
       "      <td>scRNA Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PDAC50</td>\n",
       "      <td>72.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Uncinated process</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>T2; N2</td>\n",
       "      <td>Tumor Resection (fresh)</td>\n",
       "      <td>scRNA Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>PDAC51</td>\n",
       "      <td>68.0</td>\n",
       "      <td>F</td>\n",
       "      <td>Head, Uncinated process</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>G2</td>\n",
       "      <td>T2; N2; M0</td>\n",
       "      <td>Tumor Resection (fresh)</td>\n",
       "      <td>scRNA Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>PDAC55</td>\n",
       "      <td>62.0</td>\n",
       "      <td>M</td>\n",
       "      <td>Head</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>T2; N1</td>\n",
       "      <td>Tumor Resection (fresh)</td>\n",
       "      <td>scRNA Seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID   age sex                 Location         Chemotherapy schedule  \\\n",
       "0   PDAC47  78.0   M                     Head                    FOLFIRINOX   \n",
       "3   PDAC48  61.0   M                     Head  GEMCITABINA + NAB PACLITAXEL   \n",
       "6   PDAC50  72.0   M        Uncinated process                           NaN   \n",
       "10  PDAC51  68.0   F  Head, Uncinated process                    FOLFIRINOX   \n",
       "13  PDAC55  62.0   M                     Head                           NaN   \n",
       "\n",
       "   Grade TNM_classification                   Tissue Down Stream Analysis  \n",
       "0     G2         T2; N2; M0  Tumor Resection (fresh)            scRNA Seq  \n",
       "3     G2         T2; N2; M0  Tumor Resection (fresh)            scRNA Seq  \n",
       "6     G3             T2; N2  Tumor Resection (fresh)            scRNA Seq  \n",
       "10    G2         T2; N2; M0  Tumor Resection (fresh)            scRNA Seq  \n",
       "13    G3             T2; N1  Tumor Resection (fresh)            scRNA Seq  "
      ]
     },
     "execution_count": 838,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "caronni.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 843,
   "id": "b6a7b467-52fe-4258-8833-4202e5a19815",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>LPDAC15</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727542_LPDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>PDAC60</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727551_PDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>PDAC50</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727548_PDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>PDAC48</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727547_PDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>LPDAC30</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727545_LPDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>PDAC51</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727549_PDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>PDAC55</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727550_PDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>LPDAC26</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727544_LPDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>LPDAC25</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727543_LPDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>PDAC47</td>\n",
       "      <td>Caronni</td>\n",
       "      <td>GSM6727546_PDAC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID_harmonised  Dataset         Unique_ID Case KRAS_mutation KRAS_type  \\\n",
       "84       LPDAC15  Caronni  GSM6727542_LPDAC  NaN           NaN       NaN   \n",
       "85        PDAC60  Caronni   GSM6727551_PDAC  NaN           NaN       NaN   \n",
       "86        PDAC50  Caronni   GSM6727548_PDAC  NaN           NaN       NaN   \n",
       "87        PDAC48  Caronni   GSM6727547_PDAC  NaN           NaN       NaN   \n",
       "88       LPDAC30  Caronni  GSM6727545_LPDAC  NaN           NaN       NaN   \n",
       "89        PDAC51  Caronni   GSM6727549_PDAC  NaN           NaN       NaN   \n",
       "90        PDAC55  Caronni   GSM6727550_PDAC  NaN           NaN       NaN   \n",
       "91       LPDAC26  Caronni  GSM6727544_LPDAC  NaN           NaN       NaN   \n",
       "92       LPDAC25  Caronni  GSM6727543_LPDAC  NaN           NaN       NaN   \n",
       "93        PDAC47  Caronni   GSM6727546_PDAC  NaN           NaN       NaN   \n",
       "\n",
       "   TP53_type SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2  \\\n",
       "84       NaN        NaN         NaN          NaN            NaN   \n",
       "85       NaN        NaN         NaN          NaN            NaN   \n",
       "86       NaN        NaN         NaN          NaN            NaN   \n",
       "87       NaN        NaN         NaN          NaN            NaN   \n",
       "88       NaN        NaN         NaN          NaN            NaN   \n",
       "89       NaN        NaN         NaN          NaN            NaN   \n",
       "90       NaN        NaN         NaN          NaN            NaN   \n",
       "91       NaN        NaN         NaN          NaN            NaN   \n",
       "92       NaN        NaN         NaN          NaN            NaN   \n",
       "93       NaN        NaN         NaN          NaN            NaN   \n",
       "\n",
       "   Germline_FANCC Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "84            NaN             NaN            NaN               NaN        NaN   \n",
       "85            NaN             NaN            NaN               NaN        NaN   \n",
       "86            NaN             NaN            NaN               NaN        NaN   \n",
       "87            NaN             NaN            NaN               NaN        NaN   \n",
       "88            NaN             NaN            NaN               NaN        NaN   \n",
       "89            NaN             NaN            NaN               NaN        NaN   \n",
       "90            NaN             NaN            NaN               NaN        NaN   \n",
       "91            NaN             NaN            NaN               NaN        NaN   \n",
       "92            NaN             NaN            NaN               NaN        NaN   \n",
       "93            NaN             NaN            NaN               NaN        NaN   \n",
       "\n",
       "    Time_from_diagnosis_to_surgery_(weeks) Cycles Location Grade  Size_(cm)  \\\n",
       "84                                     NaN    NaN      NaN   NaN        NaN   \n",
       "85                                     NaN    NaN      NaN   NaN        NaN   \n",
       "86                                     NaN    NaN      NaN   NaN        NaN   \n",
       "87                                     NaN    NaN      NaN   NaN        NaN   \n",
       "88                                     NaN    NaN      NaN   NaN        NaN   \n",
       "89                                     NaN    NaN      NaN   NaN        NaN   \n",
       "90                                     NaN    NaN      NaN   NaN        NaN   \n",
       "91                                     NaN    NaN      NaN   NaN        NaN   \n",
       "92                                     NaN    NaN      NaN   NaN        NaN   \n",
       "93                                     NaN    NaN      NaN   NaN        NaN   \n",
       "\n",
       "   Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "84   NaN      NaN          NaN                     NaN                  NaN   \n",
       "85   NaN      NaN          NaN                     NaN                  NaN   \n",
       "86   NaN      NaN          NaN                     NaN                  NaN   \n",
       "87   NaN      NaN          NaN                     NaN                  NaN   \n",
       "88   NaN      NaN          NaN                     NaN                  NaN   \n",
       "89   NaN      NaN          NaN                     NaN                  NaN   \n",
       "90   NaN      NaN          NaN                     NaN                  NaN   \n",
       "91   NaN      NaN          NaN                     NaN                  NaN   \n",
       "92   NaN      NaN          NaN                     NaN                  NaN   \n",
       "93   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "   Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "84    NaN     NaN             NaN            NaN   \n",
       "85    NaN     NaN             NaN            NaN   \n",
       "86    NaN     NaN             NaN            NaN   \n",
       "87    NaN     NaN             NaN            NaN   \n",
       "88    NaN     NaN             NaN            NaN   \n",
       "89    NaN     NaN             NaN            NaN   \n",
       "90    NaN     NaN             NaN            NaN   \n",
       "91    NaN     NaN             NaN            NaN   \n",
       "92    NaN     NaN             NaN            NaN   \n",
       "93    NaN     NaN             NaN            NaN   \n",
       "\n",
       "    Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "84                            NaN          NaN                NaN   \n",
       "85                            NaN          NaN                NaN   \n",
       "86                            NaN          NaN                NaN   \n",
       "87                            NaN          NaN                NaN   \n",
       "88                            NaN          NaN                NaN   \n",
       "89                            NaN          NaN                NaN   \n",
       "90                            NaN          NaN                NaN   \n",
       "91                            NaN          NaN                NaN   \n",
       "92                            NaN          NaN                NaN   \n",
       "93                            NaN          NaN                NaN   \n",
       "\n",
       "    Overall Survival Procedure  Age  BMI Treatment perineural invasion  \\\n",
       "84               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "85               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "86               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "87               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "88               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "89               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "90               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "91               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "92               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "93               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "\n",
       "   vascular invasion peripancreatic infiltarion  \n",
       "84               NaN                        NaN  \n",
       "85               NaN                        NaN  \n",
       "86               NaN                        NaN  \n",
       "87               NaN                        NaN  \n",
       "88               NaN                        NaN  \n",
       "89               NaN                        NaN  \n",
       "90               NaN                        NaN  \n",
       "91               NaN                        NaN  \n",
       "92               NaN                        NaN  \n",
       "93               NaN                        NaN  "
      ]
     },
     "execution_count": 843,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset == 'Caronni']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 845,
   "id": "3d238cf5-7d29-41e2-b526-ea0f87c808ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 845,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(caronni.ID) & set(df_id_merged.ID_harmonised))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 847,
   "id": "89c9f5d3-99ee-42f6-a28c-a640c39e3af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "caronni_map = caronni.set_index('ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 849,
   "id": "34486c85-d983-48ce-94e8-c6701dcb0743",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Location', 'Grade', 'TNM_classification', 'Tissue']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 850,
   "id": "347fbd50-c098-4e6c-9aa1-e0bddc76e307",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in interesing_cols:\n",
    "    if col in df_id_merged.columns:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(caronni_map[col]).fillna(df_id_merged[col])\n",
    "    else:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(caronni_map[col])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 851,
   "id": "506244df-1ede-45c3-b50f-2d52f507edac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>TNM_classification</th>\n",
       "      <th>T1cN0M0</th>\n",
       "      <th>T1cN1M0</th>\n",
       "      <th>T1cN2M0</th>\n",
       "      <th>T2; N1</th>\n",
       "      <th>T2; N1; M0</th>\n",
       "      <th>T2; N2</th>\n",
       "      <th>T2; N2; M0</th>\n",
       "      <th>T2N0M0</th>\n",
       "      <th>T2N1M0</th>\n",
       "      <th>T2N2M0</th>\n",
       "      <th>T3N0M0</th>\n",
       "      <th>T3N1M0</th>\n",
       "      <th>T3N2M0</th>\n",
       "      <th>T4N2M0</th>\n",
       "      <th>pT1a pN0</th>\n",
       "      <th>pT1c, pN2</th>\n",
       "      <th>pT2 pN0</th>\n",
       "      <th>pT2 pN1</th>\n",
       "      <th>pT2 pN2</th>\n",
       "      <th>pT3 pN2</th>\n",
       "      <th>rpT2 moderately differentiated pN1</th>\n",
       "      <th>yT2 pN1</th>\n",
       "      <th>ypT1 moderately differentiated positive for perineural invasion negative for lymphovascular pN0</th>\n",
       "      <th>ypT1aN0M0</th>\n",
       "      <th>ypT1aN1M0</th>\n",
       "      <th>ypT1cN0M0</th>\n",
       "      <th>ypT1cN1M0</th>\n",
       "      <th>ypT2 pN0</th>\n",
       "      <th>ypT2 pN1</th>\n",
       "      <th>ypT2 pN2</th>\n",
       "      <th>ypT2N0M0</th>\n",
       "      <th>ypT2N1M0</th>\n",
       "      <th>ypT2N2M0</th>\n",
       "      <th>ypT3 pN0</th>\n",
       "      <th>ypT3 pN2</th>\n",
       "      <th>ypT3N0M0</th>\n",
       "      <th>ypT3N1M0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Caronni</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hwang</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng</th>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_Normal</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "TNM_classification  T1cN0M0  T1cN1M0  T1cN2M0  T2; N1  T2; N1; M0  T2; N2  \\\n",
       "Dataset                                                                     \n",
       "Caronni                 NaN      NaN      NaN     1.0         2.0     1.0   \n",
       "Ding                    NaN      NaN      NaN     NaN         NaN     NaN   \n",
       "Ding_snRNA-seq          NaN      NaN      NaN     NaN         NaN     NaN   \n",
       "Hwang                   NaN      NaN      NaN     NaN         NaN     NaN   \n",
       "Peng                    1.0      2.0      1.0     NaN         NaN     NaN   \n",
       "Peng_Normal             NaN      NaN      NaN     NaN         NaN     NaN   \n",
       "\n",
       "TNM_classification  T2; N2; M0  T2N0M0  T2N1M0  T2N2M0  T3N0M0  T3N1M0  \\\n",
       "Dataset                                                                  \n",
       "Caronni                    5.0     NaN     NaN     NaN     NaN     NaN   \n",
       "Ding                       NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "Ding_snRNA-seq             NaN     NaN     NaN     NaN     NaN     NaN   \n",
       "Hwang                      NaN     2.0     6.0     1.0     3.0     3.0   \n",
       "Peng                       NaN     8.0     6.0     NaN     1.0     3.0   \n",
       "Peng_Normal                NaN     NaN     NaN     NaN     1.0     NaN   \n",
       "\n",
       "TNM_classification  T3N2M0  T4N2M0  pT1a pN0  pT1c, pN2  pT2 pN0  pT2 pN1  \\\n",
       "Dataset                                                                     \n",
       "Caronni                NaN     NaN       NaN        1.0      NaN      NaN   \n",
       "Ding                   NaN     NaN       NaN        NaN      NaN      2.0   \n",
       "Ding_snRNA-seq         NaN     NaN       1.0        NaN      2.0      1.0   \n",
       "Hwang                  3.0     NaN       NaN        NaN      NaN      NaN   \n",
       "Peng                   1.0     1.0       NaN        NaN      NaN      NaN   \n",
       "Peng_Normal            NaN     NaN       NaN        NaN      NaN      NaN   \n",
       "\n",
       "TNM_classification  pT2 pN2  pT3 pN2  rpT2 moderately differentiated pN1  \\\n",
       "Dataset                                                                    \n",
       "Caronni                 NaN      NaN                                 NaN   \n",
       "Ding                    4.0      1.0                                 NaN   \n",
       "Ding_snRNA-seq          NaN      NaN                                 1.0   \n",
       "Hwang                   NaN      NaN                                 NaN   \n",
       "Peng                    NaN      NaN                                 NaN   \n",
       "Peng_Normal             NaN      NaN                                 NaN   \n",
       "\n",
       "TNM_classification  yT2 pN1  \\\n",
       "Dataset                       \n",
       "Caronni                 NaN   \n",
       "Ding                    1.0   \n",
       "Ding_snRNA-seq          NaN   \n",
       "Hwang                   NaN   \n",
       "Peng                    NaN   \n",
       "Peng_Normal             NaN   \n",
       "\n",
       "TNM_classification  ypT1 moderately differentiated positive for perineural invasion negative for lymphovascular pN0  \\\n",
       "Dataset                                                                                                               \n",
       "Caronni                                                           NaN                                                 \n",
       "Ding                                                              NaN                                                 \n",
       "Ding_snRNA-seq                                                    1.0                                                 \n",
       "Hwang                                                             NaN                                                 \n",
       "Peng                                                              NaN                                                 \n",
       "Peng_Normal                                                       NaN                                                 \n",
       "\n",
       "TNM_classification  ypT1aN0M0  ypT1aN1M0  ypT1cN0M0  ypT1cN1M0  ypT2 pN0  \\\n",
       "Dataset                                                                    \n",
       "Caronni                   NaN        NaN        NaN        NaN       NaN   \n",
       "Ding                      NaN        NaN        NaN        NaN       3.0   \n",
       "Ding_snRNA-seq            NaN        NaN        NaN        NaN       1.0   \n",
       "Hwang                     2.0        1.0        2.0        1.0       NaN   \n",
       "Peng                      NaN        NaN        NaN        NaN       NaN   \n",
       "Peng_Normal               NaN        NaN        NaN        NaN       NaN   \n",
       "\n",
       "TNM_classification  ypT2 pN1  ypT2 pN2  ypT2N0M0  ypT2N1M0  ypT2N2M0  \\\n",
       "Dataset                                                                \n",
       "Caronni                  NaN       NaN       NaN       NaN       NaN   \n",
       "Ding                     5.0       3.0       NaN       NaN       NaN   \n",
       "Ding_snRNA-seq           1.0       NaN       NaN       NaN       NaN   \n",
       "Hwang                    NaN       NaN       6.0       2.0       1.0   \n",
       "Peng                     NaN       NaN       NaN       NaN       NaN   \n",
       "Peng_Normal              NaN       NaN       NaN       NaN       NaN   \n",
       "\n",
       "TNM_classification  ypT3 pN0  ypT3 pN2  ypT3N0M0  ypT3N1M0  \n",
       "Dataset                                                     \n",
       "Caronni                  NaN       NaN       NaN       NaN  \n",
       "Ding                     NaN       2.0       NaN       NaN  \n",
       "Ding_snRNA-seq           2.0       1.0       NaN       NaN  \n",
       "Hwang                    NaN       NaN       8.0       2.0  \n",
       "Peng                     NaN       NaN       NaN       NaN  \n",
       "Peng_Normal              NaN       NaN       NaN       NaN  "
      ]
     },
     "execution_count": 851,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'TNM_classification']).size().unstack()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "634e04bc-9e55-4545-950b-48853d5a2582",
   "metadata": {},
   "source": [
    "# Linn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 853,
   "id": "e28b6ce8-fbdb-4a56-a833-dc54e9288b7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "linn = pd.read_excel('metadata_publications/Lin_metadata.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 866,
   "id": "3985cc11-0300-4005-8a94-41dc9aba6228",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Diagnosis</th>\n",
       "      <th>Primary of Metastasis</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Grade</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Lin_P01</td>\n",
       "      <td>68</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>III</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Lin_P02</td>\n",
       "      <td>64</td>\n",
       "      <td>F</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>III</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lin_P03</td>\n",
       "      <td>72</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>IIB</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Lin_P04</td>\n",
       "      <td>57</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>primary</td>\n",
       "      <td>IIB</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Lin_P05</td>\n",
       "      <td>80</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>IIB</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Lin_P06</td>\n",
       "      <td>41</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>IIB</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Lin_P07</td>\n",
       "      <td>56</td>\n",
       "      <td>F</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>primary</td>\n",
       "      <td>IIA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Lin_P08</td>\n",
       "      <td>67</td>\n",
       "      <td>F</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>IIA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Lin_P09</td>\n",
       "      <td>53</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Primary</td>\n",
       "      <td>IIB</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Lin_P10</td>\n",
       "      <td>74</td>\n",
       "      <td>F</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>pri ma ry</td>\n",
       "      <td>IIA</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Lin_M01</td>\n",
       "      <td>43</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Liver Met</td>\n",
       "      <td>IV</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Lin_M02</td>\n",
       "      <td>61</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Liver Met</td>\n",
       "      <td>IV</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Lin_M03</td>\n",
       "      <td>39</td>\n",
       "      <td>M</td>\n",
       "      <td>pDAC</td>\n",
       "      <td>Omentum Met</td>\n",
       "      <td>IV</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Lin_M04</td>\n",
       "      <td>47</td>\n",
       "      <td>F</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Liver Met</td>\n",
       "      <td>IV</td>\n",
       "      <td>2-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Lin_M05</td>\n",
       "      <td>46</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Liver Met</td>\n",
       "      <td>IV</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Lin_M06</td>\n",
       "      <td>67</td>\n",
       "      <td>M</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>Liver Met</td>\n",
       "      <td>IV</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Patient ID  Age Gender Diagnosis Primary of Metastasis Stage Grade\n",
       "0     Lin_P01   68      M      PDAC               Primary   III     4\n",
       "1     Lin_P02   64      F      PDAC               Primary   III     2\n",
       "2     Lin_P03   72      M      PDAC               Primary   IIB     4\n",
       "3     Lin_P04   57      M      PDAC               primary   IIB     3\n",
       "4     Lin_P05   80      M      PDAC               Primary   IIB     2\n",
       "5     Lin_P06   41      M      PDAC               Primary   IIB     2\n",
       "6     Lin_P07   56      F      PDAC               primary   IIA     3\n",
       "7     Lin_P08   67      F      PDAC               Primary   IIA     2\n",
       "8     Lin_P09   53      M      PDAC               Primary   IIB     2\n",
       "9     Lin_P10   74      F      PDAC             pri ma ry   IIA     2\n",
       "10    Lin_M01   43      M      PDAC             Liver Met    IV    ND\n",
       "11    Lin_M02   61      M      PDAC             Liver Met    IV     2\n",
       "12    Lin_M03   39      M      pDAC           Omentum Met    IV    ND\n",
       "13    Lin_M04   47      F      PDAC             Liver Met    IV   2-3\n",
       "14    Lin_M05   46      M      PDAC             Liver Met    IV    ND\n",
       "15    Lin_M06   67      M      PDAC             Liver Met    IV    ND"
      ]
     },
     "execution_count": 866,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "linn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 857,
   "id": "bf55aa71-e931-478b-a1ee-e212469b4cae",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged.ID_harmonised = np.where(df_id_merged.Dataset.str.contains('Lin'), 'Lin' + '_' + df_id_merged.ID_harmonised, df_id_merged.ID_harmonised)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 858,
   "id": "ac534371-295a-4665-9fd1-7e87ddc15cb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Lin_P02</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Lin_P10</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679541</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Lin_P01</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679532</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Lin_P05</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679536</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Lin_P04</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679535</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Lin_P08</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>Lin_P06</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679537</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>Lin_P07</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>Lin_P03</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679534</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>Lin_P09</td>\n",
       "      <td>Lin</td>\n",
       "      <td>GSM4679540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ID_harmonised Dataset   Unique_ID Case KRAS_mutation KRAS_type TP53_type  \\\n",
       "94        Lin_P02     Lin  GSM4679533  NaN           NaN       NaN       NaN   \n",
       "95        Lin_P10     Lin  GSM4679541  NaN           NaN       NaN       NaN   \n",
       "96        Lin_P01     Lin  GSM4679532  NaN           NaN       NaN       NaN   \n",
       "97        Lin_P05     Lin  GSM4679536  NaN           NaN       NaN       NaN   \n",
       "98        Lin_P04     Lin  GSM4679535  NaN           NaN       NaN       NaN   \n",
       "99        Lin_P08     Lin  GSM4679539  NaN           NaN       NaN       NaN   \n",
       "100       Lin_P06     Lin  GSM4679537  NaN           NaN       NaN       NaN   \n",
       "101       Lin_P07     Lin  GSM4679538  NaN           NaN       NaN       NaN   \n",
       "102       Lin_P03     Lin  GSM4679534  NaN           NaN       NaN       NaN   \n",
       "103       Lin_P09     Lin  GSM4679540  NaN           NaN       NaN       NaN   \n",
       "\n",
       "    SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "94         NaN         NaN          NaN            NaN            NaN   \n",
       "95         NaN         NaN          NaN            NaN            NaN   \n",
       "96         NaN         NaN          NaN            NaN            NaN   \n",
       "97         NaN         NaN          NaN            NaN            NaN   \n",
       "98         NaN         NaN          NaN            NaN            NaN   \n",
       "99         NaN         NaN          NaN            NaN            NaN   \n",
       "100        NaN         NaN          NaN            NaN            NaN   \n",
       "101        NaN         NaN          NaN            NaN            NaN   \n",
       "102        NaN         NaN          NaN            NaN            NaN   \n",
       "103        NaN         NaN          NaN            NaN            NaN   \n",
       "\n",
       "    Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "94              NaN            NaN               NaN        NaN   \n",
       "95              NaN            NaN               NaN        NaN   \n",
       "96              NaN            NaN               NaN        NaN   \n",
       "97              NaN            NaN               NaN        NaN   \n",
       "98              NaN            NaN               NaN        NaN   \n",
       "99              NaN            NaN               NaN        NaN   \n",
       "100             NaN            NaN               NaN        NaN   \n",
       "101             NaN            NaN               NaN        NaN   \n",
       "102             NaN            NaN               NaN        NaN   \n",
       "103             NaN            NaN               NaN        NaN   \n",
       "\n",
       "     Time_from_diagnosis_to_surgery_(weeks) Cycles Location Grade  Size_(cm)  \\\n",
       "94                                      NaN    NaN      NaN   NaN        NaN   \n",
       "95                                      NaN    NaN      NaN   NaN        NaN   \n",
       "96                                      NaN    NaN      NaN   NaN        NaN   \n",
       "97                                      NaN    NaN      NaN   NaN        NaN   \n",
       "98                                      NaN    NaN      NaN   NaN        NaN   \n",
       "99                                      NaN    NaN      NaN   NaN        NaN   \n",
       "100                                     NaN    NaN      NaN   NaN        NaN   \n",
       "101                                     NaN    NaN      NaN   NaN        NaN   \n",
       "102                                     NaN    NaN      NaN   NaN        NaN   \n",
       "103                                     NaN    NaN      NaN   NaN        NaN   \n",
       "\n",
       "    Stage Diabetes Pancreatitis Previous_cancer_history Other_co-morbidities  \\\n",
       "94    NaN      NaN          NaN                     NaN                  NaN   \n",
       "95    NaN      NaN          NaN                     NaN                  NaN   \n",
       "96    NaN      NaN          NaN                     NaN                  NaN   \n",
       "97    NaN      NaN          NaN                     NaN                  NaN   \n",
       "98    NaN      NaN          NaN                     NaN                  NaN   \n",
       "99    NaN      NaN          NaN                     NaN                  NaN   \n",
       "100   NaN      NaN          NaN                     NaN                  NaN   \n",
       "101   NaN      NaN          NaN                     NaN                  NaN   \n",
       "102   NaN      NaN          NaN                     NaN                  NaN   \n",
       "103   NaN      NaN          NaN                     NaN                  NaN   \n",
       "\n",
       "    Smoker Alcohol Pathology_notes Lymph_node_met  \\\n",
       "94     NaN     NaN             NaN            NaN   \n",
       "95     NaN     NaN             NaN            NaN   \n",
       "96     NaN     NaN             NaN            NaN   \n",
       "97     NaN     NaN             NaN            NaN   \n",
       "98     NaN     NaN             NaN            NaN   \n",
       "99     NaN     NaN             NaN            NaN   \n",
       "100    NaN     NaN             NaN            NaN   \n",
       "101    NaN     NaN             NaN            NaN   \n",
       "102    NaN     NaN             NaN            NaN   \n",
       "103    NaN     NaN             NaN            NaN   \n",
       "\n",
       "     Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "94                             NaN          NaN                NaN   \n",
       "95                             NaN          NaN                NaN   \n",
       "96                             NaN          NaN                NaN   \n",
       "97                             NaN          NaN                NaN   \n",
       "98                             NaN          NaN                NaN   \n",
       "99                             NaN          NaN                NaN   \n",
       "100                            NaN          NaN                NaN   \n",
       "101                            NaN          NaN                NaN   \n",
       "102                            NaN          NaN                NaN   \n",
       "103                            NaN          NaN                NaN   \n",
       "\n",
       "     Overall Survival Procedure  Age  BMI Treatment perineural invasion  \\\n",
       "94                NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "95                NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "96                NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "97                NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "98                NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "99                NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "100               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "101               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "102               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "103               NaN       NaN  NaN  NaN       nan                 NaN   \n",
       "\n",
       "    vascular invasion peripancreatic infiltarion Tissue  \n",
       "94                NaN                        NaN    NaN  \n",
       "95                NaN                        NaN    NaN  \n",
       "96                NaN                        NaN    NaN  \n",
       "97                NaN                        NaN    NaN  \n",
       "98                NaN                        NaN    NaN  \n",
       "99                NaN                        NaN    NaN  \n",
       "100               NaN                        NaN    NaN  \n",
       "101               NaN                        NaN    NaN  \n",
       "102               NaN                        NaN    NaN  \n",
       "103               NaN                        NaN    NaN  "
      ]
     },
     "execution_count": 858,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged[df_id_merged.Dataset.str.contains('Lin')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 861,
   "id": "4b476339-7045-4b9a-92fc-63c7a05360b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "linn['Patient ID'] = 'Lin' + '_' + linn['Patient ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 865,
   "id": "51504827-9a29-4aea-ad0a-536940bcbe7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 865,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(linn['Patient ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 864,
   "id": "ee542672-2613-4981-83d8-2ab9a2925e28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 864,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(linn['Patient ID']) & set(df_id_merged.ID_harmonised))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 869,
   "id": "cbd63f63-8a02-4288-8937-e012e2eb080b",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesing_cols = ['Stage', 'Grade']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 870,
   "id": "8a8d096d-292f-41ac-bc7d-00bb57c686ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "linn_map = linn.set_index('Patient ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 871,
   "id": "df3b5fca-ad8b-4ed2-afe7-25c590850102",
   "metadata": {},
   "outputs": [],
   "source": [
    "for col in interesing_cols:\n",
    "    if col in df_id_merged.columns:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(linn_map[col]).fillna(df_id_merged[col])\n",
    "    else:\n",
    "        df_id_merged[col] = df_id_merged['ID_harmonised'].map(linn_map[col])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 872,
   "id": "f77fa617-d3da-48f7-870a-e7071d772576",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Stage</th>\n",
       "      <th>IB</th>\n",
       "      <th>IIA</th>\n",
       "      <th>IIB</th>\n",
       "      <th>III</th>\n",
       "      <th>IIb</th>\n",
       "      <th>Ia</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ding</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lin</th>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng</th>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_Normal</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Stage            IB  IIA   IIB  III  IIb   Ia\n",
       "Dataset                                      \n",
       "Ding            NaN  NaN   NaN  NaN  2.0  NaN\n",
       "Ding_snRNA-seq  NaN  NaN   NaN  1.0  2.0  1.0\n",
       "Lin             NaN  3.0   5.0  2.0  NaN  NaN\n",
       "Peng            9.0  2.0  10.0  3.0  NaN  NaN\n",
       "Peng_Normal     NaN  1.0   NaN  NaN  NaN  NaN"
      ]
     },
     "execution_count": 872,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby(['Dataset', 'Stage']).size().unstack()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18797682-d6f6-4869-9860-1a99ab0a8381",
   "metadata": {},
   "source": [
    "# Schlesinger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 880,
   "id": "07504c06-3f2b-4f70-affb-6462b45225ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CELL</th>\n",
       "      <th>Patient</th>\n",
       "      <th>Type</th>\n",
       "      <th>Cell_type</th>\n",
       "      <th>celltype0</th>\n",
       "      <th>celltype1</th>\n",
       "      <th>celltype2</th>\n",
       "      <th>celltype3</th>\n",
       "      <th>Patient2</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>Project</th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>Percent_mito</th>\n",
       "      <th>Percent_ribo</th>\n",
       "      <th>Percent_hemo</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>predicted.id</th>\n",
       "      <th>prediction.score.Fibroblast.cell</th>\n",
       "      <th>prediction.score.Stellate.cell</th>\n",
       "      <th>prediction.score.Macrophage.cell</th>\n",
       "      <th>prediction.score.Endothelial.cell</th>\n",
       "      <th>prediction.score.T.cell</th>\n",
       "      <th>prediction.score.B.cell</th>\n",
       "      <th>prediction.score.Ductal.cell.type.2</th>\n",
       "      <th>prediction.score.Endocrine.cell</th>\n",
       "      <th>prediction.score.Ductal.cell.type.1</th>\n",
       "      <th>prediction.score.Acinar.cell</th>\n",
       "      <th>prediction.score.max</th>\n",
       "      <th>classical_score1</th>\n",
       "      <th>basal_score1</th>\n",
       "      <th>classical_score21</th>\n",
       "      <th>basal_score21</th>\n",
       "      <th>endocrine_score_1</th>\n",
       "      <th>immune_score_1</th>\n",
       "      <th>exocrine_score_1</th>\n",
       "      <th>activated_stroma_score_1</th>\n",
       "      <th>histone_score_1</th>\n",
       "      <th>normal_stroma_score_1</th>\n",
       "      <th>olfactory_score_1</th>\n",
       "      <th>na_score_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGCTTATCG.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Ductal cell type 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>12017.0</td>\n",
       "      <td>2768</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.124823</td>\n",
       "      <td>Ductal cell type 2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.026194</td>\n",
       "      <td>0.007170</td>\n",
       "      <td>-0.036023</td>\n",
       "      <td>0.079767</td>\n",
       "      <td>-0.011971</td>\n",
       "      <td>-0.086293</td>\n",
       "      <td>-0.065146</td>\n",
       "      <td>-0.050759</td>\n",
       "      <td>0.004761</td>\n",
       "      <td>-0.017200</td>\n",
       "      <td>-0.044291</td>\n",
       "      <td>-0.011333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGTTCCC.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Ductal cell type 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>7956.0</td>\n",
       "      <td>2205</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.255405</td>\n",
       "      <td>Ductal cell type 2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.984984</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.015016</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.984984</td>\n",
       "      <td>0.170036</td>\n",
       "      <td>0.072386</td>\n",
       "      <td>0.322161</td>\n",
       "      <td>0.278054</td>\n",
       "      <td>-0.018611</td>\n",
       "      <td>-0.086257</td>\n",
       "      <td>-0.071900</td>\n",
       "      <td>-0.104126</td>\n",
       "      <td>-0.042787</td>\n",
       "      <td>-0.025997</td>\n",
       "      <td>0.190484</td>\n",
       "      <td>0.010961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTACGACCC.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Macrophage cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>3865.0</td>\n",
       "      <td>1256</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.897801</td>\n",
       "      <td>Macrophage cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.052737</td>\n",
       "      <td>-0.043333</td>\n",
       "      <td>-0.092356</td>\n",
       "      <td>-0.069527</td>\n",
       "      <td>-0.014427</td>\n",
       "      <td>0.008943</td>\n",
       "      <td>-0.071959</td>\n",
       "      <td>-0.010036</td>\n",
       "      <td>-0.040611</td>\n",
       "      <td>-0.016507</td>\n",
       "      <td>-0.010267</td>\n",
       "      <td>0.014202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGGTCCTGCTT.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>T cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>1549.0</td>\n",
       "      <td>706</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.582311</td>\n",
       "      <td>T cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.038521</td>\n",
       "      <td>-0.031630</td>\n",
       "      <td>-0.069824</td>\n",
       "      <td>-0.086458</td>\n",
       "      <td>-0.011213</td>\n",
       "      <td>0.049802</td>\n",
       "      <td>-0.045535</td>\n",
       "      <td>-0.073924</td>\n",
       "      <td>-0.038334</td>\n",
       "      <td>-0.014738</td>\n",
       "      <td>-0.021979</td>\n",
       "      <td>-0.007318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGTCTCCGGTT.1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>10471.0</td>\n",
       "      <td>2860</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.991978</td>\n",
       "      <td>Endothelial cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.446262</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.553738</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.553738</td>\n",
       "      <td>-0.063611</td>\n",
       "      <td>-0.044657</td>\n",
       "      <td>-0.101788</td>\n",
       "      <td>-0.039715</td>\n",
       "      <td>-0.012631</td>\n",
       "      <td>-0.090387</td>\n",
       "      <td>-0.120219</td>\n",
       "      <td>0.044567</td>\n",
       "      <td>0.005782</td>\n",
       "      <td>0.007340</td>\n",
       "      <td>-0.041781</td>\n",
       "      <td>-0.013644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACAATCAGAA.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Macrophage cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>14317.0</td>\n",
       "      <td>3216</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.016205</td>\n",
       "      <td>Macrophage cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.064940</td>\n",
       "      <td>-0.039494</td>\n",
       "      <td>-0.096741</td>\n",
       "      <td>-0.044958</td>\n",
       "      <td>-0.020583</td>\n",
       "      <td>-0.016210</td>\n",
       "      <td>-0.122504</td>\n",
       "      <td>-0.053966</td>\n",
       "      <td>-0.049402</td>\n",
       "      <td>-0.008592</td>\n",
       "      <td>-0.049314</td>\n",
       "      <td>0.007289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCACATTGAGCT.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Ductal cell type 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>6242.0</td>\n",
       "      <td>1887</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.906440</td>\n",
       "      <td>Ductal cell type 2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.027045</td>\n",
       "      <td>0.184153</td>\n",
       "      <td>0.060189</td>\n",
       "      <td>0.569048</td>\n",
       "      <td>-0.014743</td>\n",
       "      <td>-0.088458</td>\n",
       "      <td>-0.083752</td>\n",
       "      <td>-0.060217</td>\n",
       "      <td>0.018828</td>\n",
       "      <td>-0.027924</td>\n",
       "      <td>-0.003724</td>\n",
       "      <td>-0.009523</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTACGCGC.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>T cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>621</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.544910</td>\n",
       "      <td>T cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.036852</td>\n",
       "      <td>-0.025715</td>\n",
       "      <td>-0.059922</td>\n",
       "      <td>-0.048385</td>\n",
       "      <td>-0.009793</td>\n",
       "      <td>0.082954</td>\n",
       "      <td>-0.057760</td>\n",
       "      <td>-0.038966</td>\n",
       "      <td>-0.030693</td>\n",
       "      <td>-0.009352</td>\n",
       "      <td>-0.019439</td>\n",
       "      <td>-0.004807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCAGTTCTGAAC.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Macrophage cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>2802.0</td>\n",
       "      <td>931</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.568879</td>\n",
       "      <td>Macrophage cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.042013</td>\n",
       "      <td>-0.012602</td>\n",
       "      <td>-0.062265</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>-0.008709</td>\n",
       "      <td>0.006425</td>\n",
       "      <td>-0.075182</td>\n",
       "      <td>-0.054605</td>\n",
       "      <td>-0.069540</td>\n",
       "      <td>0.000325</td>\n",
       "      <td>-0.024479</td>\n",
       "      <td>-0.004986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGTCATCTGCCAGG.3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>T cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>1152.0</td>\n",
       "      <td>524</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>GSM4293555</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.083333</td>\n",
       "      <td>T cell</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.046467</td>\n",
       "      <td>-0.021963</td>\n",
       "      <td>-0.074276</td>\n",
       "      <td>-0.038332</td>\n",
       "      <td>-0.009534</td>\n",
       "      <td>0.165481</td>\n",
       "      <td>-0.059087</td>\n",
       "      <td>-0.042106</td>\n",
       "      <td>-0.015409</td>\n",
       "      <td>-0.011928</td>\n",
       "      <td>-0.017464</td>\n",
       "      <td>-0.005118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4874 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   CELL     Patient   Type           Cell_type celltype0  \\\n",
       "AAACCTGAGCTTATCG.1  NaN  GSM4293555  Tumor  Ductal cell type 2       NaN   \n",
       "AAACCTGAGGGTTCCC.1  NaN  GSM4293555  Tumor  Ductal cell type 2       NaN   \n",
       "AAACCTGGTACGACCC.1  NaN  GSM4293555  Tumor     Macrophage cell       NaN   \n",
       "AAACCTGGTCCTGCTT.1  NaN  GSM4293555  Tumor              T cell       NaN   \n",
       "AAACCTGTCTCCGGTT.1  NaN  GSM4293555  Tumor    Endothelial cell       NaN   \n",
       "...                 ...         ...    ...                 ...       ...   \n",
       "TTTGTCACAATCAGAA.3  NaN  GSM4293555  Tumor     Macrophage cell       NaN   \n",
       "TTTGTCACATTGAGCT.3  NaN  GSM4293555  Tumor  Ductal cell type 2       NaN   \n",
       "TTTGTCAGTTACGCGC.3  NaN  GSM4293555  Tumor              T cell       NaN   \n",
       "TTTGTCAGTTCTGAAC.3  NaN  GSM4293555  Tumor     Macrophage cell       NaN   \n",
       "TTTGTCATCTGCCAGG.3  NaN  GSM4293555  Tumor              T cell       NaN   \n",
       "\n",
       "                   celltype1 celltype2 celltype3    Patient2  nCount_RNA  \\\n",
       "AAACCTGAGCTTATCG.1       NaN       NaN       NaN  GSM4293555     12017.0   \n",
       "AAACCTGAGGGTTCCC.1       NaN       NaN       NaN  GSM4293555      7956.0   \n",
       "AAACCTGGTACGACCC.1       NaN       NaN       NaN  GSM4293555      3865.0   \n",
       "AAACCTGGTCCTGCTT.1       NaN       NaN       NaN  GSM4293555      1549.0   \n",
       "AAACCTGTCTCCGGTT.1       NaN       NaN       NaN  GSM4293555     10471.0   \n",
       "...                      ...       ...       ...         ...         ...   \n",
       "TTTGTCACAATCAGAA.3       NaN       NaN       NaN  GSM4293555     14317.0   \n",
       "TTTGTCACATTGAGCT.3       NaN       NaN       NaN  GSM4293555      6242.0   \n",
       "TTTGTCAGTTACGCGC.3       NaN       NaN       NaN  GSM4293555      2004.0   \n",
       "TTTGTCAGTTCTGAAC.3       NaN       NaN       NaN  GSM4293555      2802.0   \n",
       "TTTGTCATCTGCCAGG.3       NaN       NaN       NaN  GSM4293555      1152.0   \n",
       "\n",
       "                    nFeature_RNA     Project  orig.ident Stage  Grade Gender  \\\n",
       "AAACCTGAGCTTATCG.1          2768  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "AAACCTGAGGGTTCCC.1          2205  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "AAACCTGGTACGACCC.1          1256  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "AAACCTGGTCCTGCTT.1           706  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "AAACCTGTCTCCGGTT.1          2860  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "...                          ...         ...         ...   ...    ...    ...   \n",
       "TTTGTCACAATCAGAA.3          3216  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "TTTGTCACATTGAGCT.3          1887  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "TTTGTCAGTTACGCGC.3           621  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "TTTGTCAGTTCTGAAC.3           931  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "TTTGTCATCTGCCAGG.3           524  GSM4293555  GSM4293555   NaN    NaN    NaN   \n",
       "\n",
       "                    Age  Percent_mito  Percent_ribo  Percent_hemo  percent.mt  \\\n",
       "AAACCTGAGCTTATCG.1  NaN           NaN           NaN           NaN    0.124823   \n",
       "AAACCTGAGGGTTCCC.1  NaN           NaN           NaN           NaN    3.255405   \n",
       "AAACCTGGTACGACCC.1  NaN           NaN           NaN           NaN    2.897801   \n",
       "AAACCTGGTCCTGCTT.1  NaN           NaN           NaN           NaN    2.582311   \n",
       "AAACCTGTCTCCGGTT.1  NaN           NaN           NaN           NaN    3.991978   \n",
       "...                 ...           ...           ...           ...         ...   \n",
       "TTTGTCACAATCAGAA.3  NaN           NaN           NaN           NaN    4.016205   \n",
       "TTTGTCACATTGAGCT.3  NaN           NaN           NaN           NaN    1.906440   \n",
       "TTTGTCAGTTACGCGC.3  NaN           NaN           NaN           NaN    2.544910   \n",
       "TTTGTCAGTTCTGAAC.3  NaN           NaN           NaN           NaN    3.568879   \n",
       "TTTGTCATCTGCCAGG.3  NaN           NaN           NaN           NaN    2.083333   \n",
       "\n",
       "                          predicted.id  prediction.score.Fibroblast.cell  \\\n",
       "AAACCTGAGCTTATCG.1  Ductal cell type 2                               0.0   \n",
       "AAACCTGAGGGTTCCC.1  Ductal cell type 2                               0.0   \n",
       "AAACCTGGTACGACCC.1     Macrophage cell                               0.0   \n",
       "AAACCTGGTCCTGCTT.1              T cell                               0.0   \n",
       "AAACCTGTCTCCGGTT.1    Endothelial cell                               0.0   \n",
       "...                                ...                               ...   \n",
       "TTTGTCACAATCAGAA.3     Macrophage cell                               0.0   \n",
       "TTTGTCACATTGAGCT.3  Ductal cell type 2                               0.0   \n",
       "TTTGTCAGTTACGCGC.3              T cell                               0.0   \n",
       "TTTGTCAGTTCTGAAC.3     Macrophage cell                               0.0   \n",
       "TTTGTCATCTGCCAGG.3              T cell                               0.0   \n",
       "\n",
       "                    prediction.score.Stellate.cell  \\\n",
       "AAACCTGAGCTTATCG.1                        0.000000   \n",
       "AAACCTGAGGGTTCCC.1                        0.000000   \n",
       "AAACCTGGTACGACCC.1                        0.000000   \n",
       "AAACCTGGTCCTGCTT.1                        0.000000   \n",
       "AAACCTGTCTCCGGTT.1                        0.446262   \n",
       "...                                            ...   \n",
       "TTTGTCACAATCAGAA.3                        0.000000   \n",
       "TTTGTCACATTGAGCT.3                        0.000000   \n",
       "TTTGTCAGTTACGCGC.3                        0.000000   \n",
       "TTTGTCAGTTCTGAAC.3                        0.000000   \n",
       "TTTGTCATCTGCCAGG.3                        0.000000   \n",
       "\n",
       "                    prediction.score.Macrophage.cell  \\\n",
       "AAACCTGAGCTTATCG.1                               0.0   \n",
       "AAACCTGAGGGTTCCC.1                               0.0   \n",
       "AAACCTGGTACGACCC.1                               1.0   \n",
       "AAACCTGGTCCTGCTT.1                               0.0   \n",
       "AAACCTGTCTCCGGTT.1                               0.0   \n",
       "...                                              ...   \n",
       "TTTGTCACAATCAGAA.3                               1.0   \n",
       "TTTGTCACATTGAGCT.3                               0.0   \n",
       "TTTGTCAGTTACGCGC.3                               0.0   \n",
       "TTTGTCAGTTCTGAAC.3                               1.0   \n",
       "TTTGTCATCTGCCAGG.3                               0.0   \n",
       "\n",
       "                    prediction.score.Endothelial.cell  \\\n",
       "AAACCTGAGCTTATCG.1                           0.000000   \n",
       "AAACCTGAGGGTTCCC.1                           0.000000   \n",
       "AAACCTGGTACGACCC.1                           0.000000   \n",
       "AAACCTGGTCCTGCTT.1                           0.000000   \n",
       "AAACCTGTCTCCGGTT.1                           0.553738   \n",
       "...                                               ...   \n",
       "TTTGTCACAATCAGAA.3                           0.000000   \n",
       "TTTGTCACATTGAGCT.3                           0.000000   \n",
       "TTTGTCAGTTACGCGC.3                           0.000000   \n",
       "TTTGTCAGTTCTGAAC.3                           0.000000   \n",
       "TTTGTCATCTGCCAGG.3                           0.000000   \n",
       "\n",
       "                    prediction.score.T.cell  prediction.score.B.cell  \\\n",
       "AAACCTGAGCTTATCG.1                      0.0                      0.0   \n",
       "AAACCTGAGGGTTCCC.1                      0.0                      0.0   \n",
       "AAACCTGGTACGACCC.1                      0.0                      0.0   \n",
       "AAACCTGGTCCTGCTT.1                      1.0                      0.0   \n",
       "AAACCTGTCTCCGGTT.1                      0.0                      0.0   \n",
       "...                                     ...                      ...   \n",
       "TTTGTCACAATCAGAA.3                      0.0                      0.0   \n",
       "TTTGTCACATTGAGCT.3                      0.0                      0.0   \n",
       "TTTGTCAGTTACGCGC.3                      1.0                      0.0   \n",
       "TTTGTCAGTTCTGAAC.3                      0.0                      0.0   \n",
       "TTTGTCATCTGCCAGG.3                      1.0                      0.0   \n",
       "\n",
       "                    prediction.score.Ductal.cell.type.2  \\\n",
       "AAACCTGAGCTTATCG.1                             1.000000   \n",
       "AAACCTGAGGGTTCCC.1                             0.984984   \n",
       "AAACCTGGTACGACCC.1                             0.000000   \n",
       "AAACCTGGTCCTGCTT.1                             0.000000   \n",
       "AAACCTGTCTCCGGTT.1                             0.000000   \n",
       "...                                                 ...   \n",
       "TTTGTCACAATCAGAA.3                             0.000000   \n",
       "TTTGTCACATTGAGCT.3                             1.000000   \n",
       "TTTGTCAGTTACGCGC.3                             0.000000   \n",
       "TTTGTCAGTTCTGAAC.3                             0.000000   \n",
       "TTTGTCATCTGCCAGG.3                             0.000000   \n",
       "\n",
       "                    prediction.score.Endocrine.cell  \\\n",
       "AAACCTGAGCTTATCG.1                              0.0   \n",
       "AAACCTGAGGGTTCCC.1                              0.0   \n",
       "AAACCTGGTACGACCC.1                              0.0   \n",
       "AAACCTGGTCCTGCTT.1                              0.0   \n",
       "AAACCTGTCTCCGGTT.1                              0.0   \n",
       "...                                             ...   \n",
       "TTTGTCACAATCAGAA.3                              0.0   \n",
       "TTTGTCACATTGAGCT.3                              0.0   \n",
       "TTTGTCAGTTACGCGC.3                              0.0   \n",
       "TTTGTCAGTTCTGAAC.3                              0.0   \n",
       "TTTGTCATCTGCCAGG.3                              0.0   \n",
       "\n",
       "                    prediction.score.Ductal.cell.type.1  \\\n",
       "AAACCTGAGCTTATCG.1                             0.000000   \n",
       "AAACCTGAGGGTTCCC.1                             0.015016   \n",
       "AAACCTGGTACGACCC.1                             0.000000   \n",
       "AAACCTGGTCCTGCTT.1                             0.000000   \n",
       "AAACCTGTCTCCGGTT.1                             0.000000   \n",
       "...                                                 ...   \n",
       "TTTGTCACAATCAGAA.3                             0.000000   \n",
       "TTTGTCACATTGAGCT.3                             0.000000   \n",
       "TTTGTCAGTTACGCGC.3                             0.000000   \n",
       "TTTGTCAGTTCTGAAC.3                             0.000000   \n",
       "TTTGTCATCTGCCAGG.3                             0.000000   \n",
       "\n",
       "                    prediction.score.Acinar.cell  prediction.score.max  \\\n",
       "AAACCTGAGCTTATCG.1                           0.0              1.000000   \n",
       "AAACCTGAGGGTTCCC.1                           0.0              0.984984   \n",
       "AAACCTGGTACGACCC.1                           0.0              1.000000   \n",
       "AAACCTGGTCCTGCTT.1                           0.0              1.000000   \n",
       "AAACCTGTCTCCGGTT.1                           0.0              0.553738   \n",
       "...                                          ...                   ...   \n",
       "TTTGTCACAATCAGAA.3                           0.0              1.000000   \n",
       "TTTGTCACATTGAGCT.3                           0.0              1.000000   \n",
       "TTTGTCAGTTACGCGC.3                           0.0              1.000000   \n",
       "TTTGTCAGTTCTGAAC.3                           0.0              1.000000   \n",
       "TTTGTCATCTGCCAGG.3                           0.0              1.000000   \n",
       "\n",
       "                    classical_score1  basal_score1  classical_score21  \\\n",
       "AAACCTGAGCTTATCG.1         -0.026194      0.007170          -0.036023   \n",
       "AAACCTGAGGGTTCCC.1          0.170036      0.072386           0.322161   \n",
       "AAACCTGGTACGACCC.1         -0.052737     -0.043333          -0.092356   \n",
       "AAACCTGGTCCTGCTT.1         -0.038521     -0.031630          -0.069824   \n",
       "AAACCTGTCTCCGGTT.1         -0.063611     -0.044657          -0.101788   \n",
       "...                              ...           ...                ...   \n",
       "TTTGTCACAATCAGAA.3         -0.064940     -0.039494          -0.096741   \n",
       "TTTGTCACATTGAGCT.3          0.027045      0.184153           0.060189   \n",
       "TTTGTCAGTTACGCGC.3         -0.036852     -0.025715          -0.059922   \n",
       "TTTGTCAGTTCTGAAC.3         -0.042013     -0.012602          -0.062265   \n",
       "TTTGTCATCTGCCAGG.3         -0.046467     -0.021963          -0.074276   \n",
       "\n",
       "                    basal_score21  endocrine_score_1  immune_score_1  \\\n",
       "AAACCTGAGCTTATCG.1       0.079767          -0.011971       -0.086293   \n",
       "AAACCTGAGGGTTCCC.1       0.278054          -0.018611       -0.086257   \n",
       "AAACCTGGTACGACCC.1      -0.069527          -0.014427        0.008943   \n",
       "AAACCTGGTCCTGCTT.1      -0.086458          -0.011213        0.049802   \n",
       "AAACCTGTCTCCGGTT.1      -0.039715          -0.012631       -0.090387   \n",
       "...                           ...                ...             ...   \n",
       "TTTGTCACAATCAGAA.3      -0.044958          -0.020583       -0.016210   \n",
       "TTTGTCACATTGAGCT.3       0.569048          -0.014743       -0.088458   \n",
       "TTTGTCAGTTACGCGC.3      -0.048385          -0.009793        0.082954   \n",
       "TTTGTCAGTTCTGAAC.3       0.000273          -0.008709        0.006425   \n",
       "TTTGTCATCTGCCAGG.3      -0.038332          -0.009534        0.165481   \n",
       "\n",
       "                    exocrine_score_1  activated_stroma_score_1  \\\n",
       "AAACCTGAGCTTATCG.1         -0.065146                 -0.050759   \n",
       "AAACCTGAGGGTTCCC.1         -0.071900                 -0.104126   \n",
       "AAACCTGGTACGACCC.1         -0.071959                 -0.010036   \n",
       "AAACCTGGTCCTGCTT.1         -0.045535                 -0.073924   \n",
       "AAACCTGTCTCCGGTT.1         -0.120219                  0.044567   \n",
       "...                              ...                       ...   \n",
       "TTTGTCACAATCAGAA.3         -0.122504                 -0.053966   \n",
       "TTTGTCACATTGAGCT.3         -0.083752                 -0.060217   \n",
       "TTTGTCAGTTACGCGC.3         -0.057760                 -0.038966   \n",
       "TTTGTCAGTTCTGAAC.3         -0.075182                 -0.054605   \n",
       "TTTGTCATCTGCCAGG.3         -0.059087                 -0.042106   \n",
       "\n",
       "                    histone_score_1  normal_stroma_score_1  olfactory_score_1  \\\n",
       "AAACCTGAGCTTATCG.1         0.004761              -0.017200          -0.044291   \n",
       "AAACCTGAGGGTTCCC.1        -0.042787              -0.025997           0.190484   \n",
       "AAACCTGGTACGACCC.1        -0.040611              -0.016507          -0.010267   \n",
       "AAACCTGGTCCTGCTT.1        -0.038334              -0.014738          -0.021979   \n",
       "AAACCTGTCTCCGGTT.1         0.005782               0.007340          -0.041781   \n",
       "...                             ...                    ...                ...   \n",
       "TTTGTCACAATCAGAA.3        -0.049402              -0.008592          -0.049314   \n",
       "TTTGTCACATTGAGCT.3         0.018828              -0.027924          -0.003724   \n",
       "TTTGTCAGTTACGCGC.3        -0.030693              -0.009352          -0.019439   \n",
       "TTTGTCAGTTCTGAAC.3        -0.069540               0.000325          -0.024479   \n",
       "TTTGTCATCTGCCAGG.3        -0.015409              -0.011928          -0.017464   \n",
       "\n",
       "                    na_score_1  \n",
       "AAACCTGAGCTTATCG.1   -0.011333  \n",
       "AAACCTGAGGGTTCCC.1    0.010961  \n",
       "AAACCTGGTACGACCC.1    0.014202  \n",
       "AAACCTGGTCCTGCTT.1   -0.007318  \n",
       "AAACCTGTCTCCGGTT.1   -0.013644  \n",
       "...                        ...  \n",
       "TTTGTCACAATCAGAA.3    0.007289  \n",
       "TTTGTCACATTGAGCT.3   -0.009523  \n",
       "TTTGTCAGTTACGCGC.3   -0.004807  \n",
       "TTTGTCAGTTCTGAAC.3   -0.004986  \n",
       "TTTGTCATCTGCCAGG.3   -0.005118  \n",
       "\n",
       "[4874 rows x 45 columns]"
      ]
     },
     "execution_count": 880,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "zenodo[zenodo.obs.Project == 'GSM4293555'].obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 881,
   "id": "03d728fb-7788-456b-a253-27eec31b4880",
   "metadata": {},
   "outputs": [],
   "source": [
    "# no metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 885,
   "id": "b40799b8-590c-4475-b1d8-a894f47c5a8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2834644/1103722888.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  df_id_merged.groupby('Dataset').apply(lambda x: x.notna().sum())\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "      <th>Tissue</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Caronni</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding</th>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>21</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>19</td>\n",
       "      <td>21</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hwang</th>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lee</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lin</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng</th>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_Normal</th>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schlesinger</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Simeone</th>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele</th>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID_harmonised  Dataset  Unique_ID  Case  KRAS_mutation  \\\n",
       "Dataset                                                                   \n",
       "Caronni                     10       10         10     0              0   \n",
       "Ding                        22       22         22    21             15   \n",
       "Ding_snRNA-seq              11       11         11     0              0   \n",
       "Hwang                       43       43         43     0              0   \n",
       "Lee                          5        5          5     0              0   \n",
       "Lin                         10       10         10     0              0   \n",
       "Peng                        24       24         24     0              0   \n",
       "Peng_Normal                 11       11         11     0              0   \n",
       "Schlesinger                  1        1          1     0              0   \n",
       "Simeone                     17       17         17     0             14   \n",
       "Steele                      16       16         16     0              0   \n",
       "Steele_Adj_Norm              3        3          3     0              0   \n",
       "\n",
       "                 KRAS_type  TP53_type  SMAD4_type  CDKN2A_type  Germline_ATM  \\\n",
       "Dataset                                                                        \n",
       "Caronni                  0          0           0            0             0   \n",
       "Ding                    16         16          16           16            16   \n",
       "Ding_snRNA-seq           0          0           0            0             0   \n",
       "Hwang                    0          0           0            0             0   \n",
       "Lee                      0          0           0            0             0   \n",
       "Lin                      0          0           0            0             0   \n",
       "Peng                     0          0           0            0             0   \n",
       "Peng_Normal              0          0           0            0             0   \n",
       "Schlesinger              0          0           0            0             0   \n",
       "Simeone                  0         14          14           14             0   \n",
       "Steele                   0          0           0            0             0   \n",
       "Steele_Adj_Norm          0          0           0            0             0   \n",
       "\n",
       "                 Germline_BRCA2  Germline_FANCC  Moffitt_Subtype  \\\n",
       "Dataset                                                            \n",
       "Caronni                       0               0                0   \n",
       "Ding                         16              16               16   \n",
       "Ding_snRNA-seq                0               0                0   \n",
       "Hwang                         0               0                0   \n",
       "Lee                           0               0                0   \n",
       "Lin                           0               0                0   \n",
       "Peng                          0               0                0   \n",
       "Peng_Normal                   0               0                0   \n",
       "Schlesinger                   0               0                0   \n",
       "Simeone                       0               0                0   \n",
       "Steele                        0               0                0   \n",
       "Steele_Adj_Norm               0               0                0   \n",
       "\n",
       "                 Bailey_Subtype  Collisson_Subtype  Patient_ID  \\\n",
       "Dataset                                                          \n",
       "Caronni                       0                  0           0   \n",
       "Ding                         16                 16          21   \n",
       "Ding_snRNA-seq                0                  0          11   \n",
       "Hwang                         0                  0           0   \n",
       "Lee                           0                  0           0   \n",
       "Lin                           0                  0           0   \n",
       "Peng                          0                  0           0   \n",
       "Peng_Normal                   0                  0           0   \n",
       "Schlesinger                   0                  0           0   \n",
       "Simeone                       0                  0           0   \n",
       "Steele                        0                  0           0   \n",
       "Steele_Adj_Norm               0                  0           0   \n",
       "\n",
       "                 Time_from_diagnosis_to_surgery_(weeks)  Cycles  Location  \\\n",
       "Dataset                                                                     \n",
       "Caronni                                               0       0        10   \n",
       "Ding                                                 21       9        19   \n",
       "Ding_snRNA-seq                                       11       6        11   \n",
       "Hwang                                                 0       0         0   \n",
       "Lee                                                   0       0         0   \n",
       "Lin                                                   0       0         0   \n",
       "Peng                                                  0       0         0   \n",
       "Peng_Normal                                           0       0         0   \n",
       "Schlesinger                                           0       0         0   \n",
       "Simeone                                               0       0         0   \n",
       "Steele                                                0       0         0   \n",
       "Steele_Adj_Norm                                       0       0         0   \n",
       "\n",
       "                 Grade  Size_(cm)  Stage  Diabetes  Pancreatitis  \\\n",
       "Dataset                                                            \n",
       "Caronni             10          0      0         0             0   \n",
       "Ding                19         20      2        19            13   \n",
       "Ding_snRNA-seq      10         11      4        11            11   \n",
       "Hwang               43          0      0         0             0   \n",
       "Lee                  0          0      0         0             0   \n",
       "Lin                 10          0     10         0             0   \n",
       "Peng                 0          0     24         0             0   \n",
       "Peng_Normal          0          0      1         0             0   \n",
       "Schlesinger          0          0      0         0             0   \n",
       "Simeone              0          0      0         0             0   \n",
       "Steele               0          0      0         0             0   \n",
       "Steele_Adj_Norm      0          0      0         0             0   \n",
       "\n",
       "                 Previous_cancer_history  Other_co-morbidities  Smoker  \\\n",
       "Dataset                                                                  \n",
       "Caronni                                0                     0       0   \n",
       "Ding                                  18                    21      21   \n",
       "Ding_snRNA-seq                        11                    11      11   \n",
       "Hwang                                  0                     0       0   \n",
       "Lee                                    0                     0       0   \n",
       "Lin                                    0                     0       0   \n",
       "Peng                                   0                     0       0   \n",
       "Peng_Normal                            0                     0       0   \n",
       "Schlesinger                            0                     0       0   \n",
       "Simeone                                0                     0       0   \n",
       "Steele                                 0                     0       0   \n",
       "Steele_Adj_Norm                        0                     0       0   \n",
       "\n",
       "                 Alcohol  Pathology_notes  Lymph_node_met  \\\n",
       "Dataset                                                     \n",
       "Caronni                0                0               0   \n",
       "Ding                  19               21              21   \n",
       "Ding_snRNA-seq        11               11              11   \n",
       "Hwang                  0                0               0   \n",
       "Lee                    0                0               0   \n",
       "Lin                    0                0               0   \n",
       "Peng                   0               24               0   \n",
       "Peng_Normal            0               11               0   \n",
       "Schlesinger            0                0               0   \n",
       "Simeone                0                0               0   \n",
       "Steele                 0                0               0   \n",
       "Steele_Adj_Norm        0                0               0   \n",
       "\n",
       "                 Survival_post_surgery_(weeks)  Vital_status  \\\n",
       "Dataset                                                        \n",
       "Caronni                                      0             0   \n",
       "Ding                                         1             1   \n",
       "Ding_snRNA-seq                               3             3   \n",
       "Hwang                                        0            43   \n",
       "Lee                                          0             0   \n",
       "Lin                                          0             0   \n",
       "Peng                                         0             0   \n",
       "Peng_Normal                                  0             0   \n",
       "Schlesinger                                  0             0   \n",
       "Simeone                                      0             0   \n",
       "Steele                                       0             0   \n",
       "Steele_Adj_Norm                              0             0   \n",
       "\n",
       "                 TNM_classification  Overall Survival  Procedure  Age  BMI  \\\n",
       "Dataset                                                                      \n",
       "Caronni                          10                 0          0    0    0   \n",
       "Ding                             21                 1          0    0    0   \n",
       "Ding_snRNA-seq                   11                 3          0    0    0   \n",
       "Hwang                            43                43          0    0    0   \n",
       "Lee                               0                 0          0    0    0   \n",
       "Lin                               0                 0          0    0    0   \n",
       "Peng                             24                 0         24   24    0   \n",
       "Peng_Normal                       1                 0         11   11    0   \n",
       "Schlesinger                       0                 0          0    0    0   \n",
       "Simeone                           0                 0         17    0    0   \n",
       "Steele                            0                 0          0   16   16   \n",
       "Steele_Adj_Norm                   0                 0          0    3    3   \n",
       "\n",
       "                 Treatment  perineural invasion  vascular invasion  \\\n",
       "Dataset                                                              \n",
       "Caronni                 10                    0                  0   \n",
       "Ding                    22                    0                  0   \n",
       "Ding_snRNA-seq          11                    0                  0   \n",
       "Hwang                   43                    0                  0   \n",
       "Lee                      5                    0                  0   \n",
       "Lin                     10                    0                  0   \n",
       "Peng                    24                   24                 24   \n",
       "Peng_Normal             11                   11                 11   \n",
       "Schlesinger              1                    0                  0   \n",
       "Simeone                 17                    0                  0   \n",
       "Steele                  16                    0                  0   \n",
       "Steele_Adj_Norm          3                    0                  0   \n",
       "\n",
       "                 peripancreatic infiltarion  Tissue  \n",
       "Dataset                                              \n",
       "Caronni                                   0      10  \n",
       "Ding                                      0       0  \n",
       "Ding_snRNA-seq                            0       0  \n",
       "Hwang                                     0       0  \n",
       "Lee                                       0       0  \n",
       "Lin                                       0       0  \n",
       "Peng                                     24       0  \n",
       "Peng_Normal                              11       0  \n",
       "Schlesinger                               0       0  \n",
       "Simeone                                   0       0  \n",
       "Steele                                    0       0  \n",
       "Steele_Adj_Norm                           0       0  "
      ]
     },
     "execution_count": 885,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.groupby('Dataset').apply(lambda x: x.notna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 907,
   "id": "18b382b3-f253-4a3b-b8f5-1390cf4315dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['nan', 'Treatment Naive'], dtype=object)"
      ]
     },
     "execution_count": 907,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged['Treatment'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 893,
   "id": "fac8ba61-2022-4447-ba66-c80aa229faeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged.CDKN2A_type = df_id_merged.CDKN2A_type.replace('WT', 'Wild_Type')\n",
    "df_id_merged.SMAD4_type = df_id_merged.SMAD4_type.replace('WT', 'Wild_Type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 895,
   "id": "31578b80-1ed3-4adf-b171-bf5d7e4eaf68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the dictionary for harmonizing the 'Location' values\n",
    "location_dict = {\n",
    "    'head': 'Head',\n",
    "    'Head': 'Head',\n",
    "    'head, uncinate_process': 'Head, Uncinated Process',\n",
    "    'Head, Uncinated process': 'Head, Uncinated Process',\n",
    "    'Uncinated process': 'Uncinated Process',\n",
    "    'neck': 'Neck',\n",
    "    'neck/body margin': 'Neck/Body',\n",
    "    'body': 'Body',\n",
    "    'Body': 'Body',\n",
    "    'body/tail': 'Body/Tail',\n",
    "    'Tail': 'Tail',\n",
    "    'Tail (proximal remanant)': 'Tail, Proximal Remanant',\n",
    "    'Tail (proximal remanant)': 'Tail, Proximal Remanant',\n",
    "    'Head/Uncinate': 'Head, Uncinated Process'\n",
    "}\n",
    "\n",
    "df_id_merged['Location'] = df_id_merged['Location'].map(location_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 896,
   "id": "4c4b5eb1-98c0-4664-9fab-b728455dd012",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Head, Uncinated Process', 'Head', 'Neck', nan, 'Body',\n",
       "       'Uncinated Process', 'Tail', 'Body/Tail',\n",
       "       'Tail, Proximal Remanant', 'Neck/Body'], dtype=object)"
      ]
     },
     "execution_count": 896,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged['Location'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 903,
   "id": "a2ef44f0-35d5-4d91-b83a-9f92c1f7a440",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the dictionary for harmonizing the 'Grade' values\n",
    "grade_dict = {\n",
    "    3.0: 'G3',\n",
    "    1.0: 'G1',\n",
    "    2.0: 'G2',\n",
    "    4: 'G4',\n",
    "    'G3': 'G3',\n",
    "    'G2': 'G2',\n",
    "    'g3': 'G3',\n",
    "    'g2': 'G2',\n",
    "    'g2-3': 'G2/G3',\n",
    "    'gX': 'GX'\n",
    "}\n",
    "\n",
    "# Harmonize the 'Grade' column in df_id_merged using the dictionary\n",
    "df_id_merged['Harmonized_Grade'] = df_id_merged['Grade'].map(grade_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 904,
   "id": "b376b39c-f1c9-404a-9f94-62c911548741",
   "metadata": {},
   "outputs": [],
   "source": [
    "tnm_dict = {\n",
    "    'ypT2 pN1': 'ypT2 pN1',\n",
    "    'pT2 pN1': 'pT2 pN1',\n",
    "    'pT2 pN2': 'pT2 pN2',\n",
    "    'ypT2 pN0': 'ypT2 pN0',\n",
    "    'ypT2 pN2': 'ypT2 pN2',\n",
    "    'pT3 pN2': 'pT3 pN2',\n",
    "    'ypT3 pN2': 'ypT3 pN2',\n",
    "    'yT2 pN1': 'yT2 pN1',\n",
    "    'T4N2M0': 'T4 pN2 M0',\n",
    "    'T1cN1M0': 'T1c pN1 M0',\n",
    "    'T2N0M0': 'T2 pN0 M0',\n",
    "    'T3N0M0': 'T3 pN0 M0',\n",
    "    'T3N1M0': 'T3 pN1 M0',\n",
    "    'T1cN2M0': 'T1c pN2 M0',\n",
    "    'T2N1M0': 'T2 pN1 M0',\n",
    "    'T3N2M0': 'T3 pN2 M0',\n",
    "    'T1cN0M0': 'T1c pN0 M0',\n",
    "    'T2; N1; M0': 'T2 pN1 M0',\n",
    "    'pT1c, pN2': 'pT1c pN2',\n",
    "    'T2; N2': 'T2 pN2',\n",
    "    'T2; N2; M0': 'T2 pN2 M0',\n",
    "    'T2; N1': 'T2 pN1',\n",
    "    'ypT2N1M0': 'ypT2 pN1 M0',\n",
    "    'ypT3N0M0': 'ypT3 pN0 M0',\n",
    "    'ypT1cN0M0': 'ypT1c pN0 M0',\n",
    "    'ypT2N0M0': 'ypT2 pN0 M0',\n",
    "    'ypT1aN1M0': 'ypT1a pN1 M0',\n",
    "    'ypT2N2M0': 'ypT2 pN2 M0',\n",
    "    'T2N2M0': 'T2 pN2 M0',\n",
    "    'ypT1aN0M0': 'ypT1a pN0 M0',\n",
    "    'ypT1cN1M0': 'ypT1c pN1 M0',\n",
    "    'ypT3N1M0': 'ypT3 pN1 M0',\n",
    "    'ypT3 pN0': 'ypT3 pN0',\n",
    "    'rpT2 moderately differentiated pN1': 'rpT2 pN1',\n",
    "    'pT2 pN0': 'pT2 pN0',\n",
    "    'pT1a pN0': 'pT1a pN0',\n",
    "    'ypT1 moderately differentiated positive for perineural invasion negative for lymphovascular pN0': 'ypT1 pN0'\n",
    "}\n",
    "\n",
    "# Harmonize the 'TNM_classification' column using the dictionary\n",
    "df_id_merged['Harmonized_TNM_Classification'] = df_id_merged['TNM_classification'].map(tnm_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 908,
   "id": "997d7ad7-4cf7-40c9-8f93-c63296a52e60",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged.to_csv('metadata_publications/metadata_all.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cb94250b-045e-4fab-8c58-28e933204f94",
   "metadata": {},
   "source": [
    "# Reload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "843a97ed-6c60-4dd0-b2a5-947b90b3f0de",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "pd.set_option('display.max_columns', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "431b9e28-1409-4b8c-bf24-036c88c5c21f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8e75d793-3943-473e-91b4-36e98d30ca93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ictstr01/groups/ml01/workspace/shrey.parikh/PDAC_Work_Dir/PDAC_Final/Downstream'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "id": "e458abac-36eb-4cc7-ba3d-e97f2c5fcb3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged = pd.read_csv('metadata_publications/metadata_all.csv', index_col='Unnamed: 0')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "b3878038-c4ef-4aae-87ac-7aef325bf5e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# change "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "id": "6eef83fe-824a-44a5-a034-78d42f907b74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Case</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>Age</th>\n",
       "      <th>BMI</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Harmonized_Grade</th>\n",
       "      <th>Harmonized_TNM_Classification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>Ding</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Classical</td>\n",
       "      <td>ADEX</td>\n",
       "      <td>Exocrine-like</td>\n",
       "      <td>HT060P1</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>IIb</td>\n",
       "      <td>Type II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, extends to duodenum and vascular groove</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ID_harmonised Dataset Unique_ID     Case KRAS_mutation KRAS_type  TP53_type  \\\n",
       "0       HT056P1    Ding   HT056P1  HT056P1        p.G12V  Missense   Missense   \n",
       "1       HT060P1    Ding   HT060P1  HT060P1        p.G12V  Missense  Wild_Type   \n",
       "\n",
       "  SMAD4_type CDKN2A_type Germline_ATM Germline_BRCA2 Germline_FANCC  \\\n",
       "0  Wild_Type   Wild_Type    Wild_Type      Wild_Type      Wild_Type   \n",
       "1  Wild_Type   Wild_Type     Nonsense       Nonsense      Wild_Type   \n",
       "\n",
       "  Moffitt_Subtype Bailey_Subtype Collisson_Subtype Patient_ID  \\\n",
       "0      Basal-like    Immunogenic  Quasimesenchymal    HT056P1   \n",
       "1       Classical           ADEX     Exocrine-like    HT060P1   \n",
       "\n",
       "   Time_from_diagnosis_to_surgery_(weeks) Cycles                 Location  \\\n",
       "0                                    20.0      6  Head, Uncinated Process   \n",
       "1                                    20.0      6  Head, Uncinated Process   \n",
       "\n",
       "  Grade  Size_(cm) Stage Diabetes Pancreatitis Previous_cancer_history  \\\n",
       "0   3.0        3.2   NaN       No          Yes                      No   \n",
       "1   3.0        3.2   IIb  Type II          NaN                     NaN   \n",
       "\n",
       "  Other_co-morbidities Smoker     Alcohol  \\\n",
       "0                   No     No  Occasional   \n",
       "1                   No     No  Occasional   \n",
       "\n",
       "                                 Pathology_notes Lymph_node_met  \\\n",
       "0         ypT2, focal sarcomatoid transformation      2/21, pN1   \n",
       "1  ypT2, extends to duodenum and vascular groove      2/25, pN1   \n",
       "\n",
       "   Survival_post_surgery_(weeks) Vital_status TNM_classification  \\\n",
       "0                            NaN          NaN           ypT2 pN1   \n",
       "1                            NaN          NaN           ypT2 pN1   \n",
       "\n",
       "   Overall Survival Procedure  Age  BMI Treatment perineural invasion  \\\n",
       "0               NaN       NaN  NaN  NaN       NaN                 NaN   \n",
       "1               NaN       NaN  NaN  NaN       NaN                 NaN   \n",
       "\n",
       "  vascular invasion peripancreatic infiltarion Tissue Harmonized_Grade  \\\n",
       "0               NaN                        NaN    NaN               G3   \n",
       "1               NaN                        NaN    NaN               G3   \n",
       "\n",
       "  Harmonized_TNM_Classification  \n",
       "0                      ypT2 pN1  \n",
       "1                      ypT2 pN1  "
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_id_merged.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "id": "fa8daf04-3f0e-4958-8540-415c876153d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2515696833.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  df = df_id_merged.groupby('Dataset').apply(lambda x: x.notna().sum())\n"
     ]
    }
   ],
   "source": [
    "df = df_id_merged.groupby('Dataset').apply(lambda x: x.notna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "e107f779-f20c-475c-a324-d365dcb64297",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read Anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "df6466cc-2566-4206-947e-59f69b40bc0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('../Binned_Data/adata_scanvi_neighbours_leiden.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a029e48-0766-46f1-9331-19c212d46fda",
   "metadata": {},
   "source": [
    "## check shared ID_harmonised"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "bdae6c35-62f5-41d9-a3bf-ed10bc7895d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/3706364925.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  shared_ids = adata.obs.groupby('ID_harmonised')['Dataset'].nunique()\n"
     ]
    }
   ],
   "source": [
    "shared_ids = adata.obs.groupby('ID_harmonised')['Dataset'].nunique()\n",
    "shared_ids_multiple_datasets = shared_ids[shared_ids > 1]\n",
    "shared_ids_multiple_datasets_info = adata.obs[adata.obs['ID_harmonised'].isin(shared_ids_multiple_datasets.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "22e3a57b-83ea-49cd-939b-ba3649dd33d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID_harmonised\n",
       "P03    2\n",
       "P04    2\n",
       "P05    2\n",
       "P06    2\n",
       "P07    2\n",
       "P08    2\n",
       "P09    2\n",
       "P10    2\n",
       "Name: Dataset, dtype: int64"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_ids_multiple_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "a32174fa-c7bc-40e3-a080-ebbacf7b8b07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Simeone', 'Lin']\n",
       "Categories (12, object): ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', ..., 'Schlesinger', 'Simeone', 'Steele', 'Steele_Adj_Norm']"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_ids_multiple_datasets_info.Dataset.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cff596ed-bdcd-4df0-81d2-013af9f7baf5",
   "metadata": {},
   "source": [
    "## Resolve by adding prefixes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "id": "bebdbaf3-0c62-4ba9-ab19-a000ed4d07e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['ID_harmonised'] = adata.obs['ID_harmonised'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "id": "37721d45-3793-4b34-b273-701c4954c579",
   "metadata": {},
   "outputs": [],
   "source": [
    "conditions = [\n",
    "    (adata.obs['Dataset'].str.contains('Peng')),\n",
    "    (adata.obs['Dataset'] == 'Lin')\n",
    "]\n",
    "values = ['Peng' + '_' + adata.obs['ID_harmonised'], 'Lin' + '_' + adata.obs['ID_harmonised']]\n",
    "adata.obs['ID_harmonised'] = np.select(conditions, values, default=adata.obs['ID_harmonised'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "id": "d8165d52-260f-496b-a9c4-9f61e67b0a34",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "173"
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(df_id_merged.ID_harmonised) & set(adata.obs.ID_harmonised))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a822611c-3723-4838-ae53-8248b5939194",
   "metadata": {},
   "source": [
    "## recheck"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "886fb9ae-e924-4c9c-b45c-59d5a86a27bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "shared_ids = adata.obs.groupby('ID_harmonised')['Dataset'].nunique()\n",
    "shared_ids_multiple_datasets = shared_ids[shared_ids > 1]\n",
    "shared_ids_multiple_datasets_info = adata.obs[adata.obs['ID_harmonised'].isin(shared_ids_multiple_datasets.index)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "e4c09260-839a-4d85-8133-1a6e79bf99ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: Dataset, dtype: int64)"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_ids_multiple_datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "ad3fc8ca-27c4-441c-86c9-71920d0515e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Series([], Name: Dataset, dtype: int64)"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "shared_ids_multiple_datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffd8e8c4-39bb-448b-ae83-669922ad71e0",
   "metadata": {},
   "source": [
    "# Check if you can add this biological metadata to obsm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "id": "ec7fc365-7a63-4ec2-980c-eeb20b7bff21",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_obs_subset = adata.obs[['Barcode','Dataset','ID_batch_covariate', 'ID_harmonised', 'Unique_ID', 'Level_1_refined','Tissue', 'Age', 'Sex', 'Diabetes', 'Treatment']].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03fb2ea9-397e-4576-89cf-b95f52ac48a4",
   "metadata": {},
   "source": [
    "## Add Steele Treatment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "ea7a9b23-61ff-42df-9456-8a1d10f848c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_obs_subset.Treatment = np.where(adata_obs_subset.Dataset == 'Steele', 'Treatment Naive', adata_obs_subset.Treatment) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "3df298f2-90af-4f0d-8916-193b891669bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_id_merged.drop('Treatment', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "d2b51207-23a4-4584-a3bb-b289058721cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "## can drop Case, Patient_ID, Diabetes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "5e5deb79-439f-4440-97e5-5ffed942ee26",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_id_merged.drop(['Case', 'Patient_ID', 'Diabetes'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a738bf61-1b25-497c-8677-7068bfc2a941",
   "metadata": {},
   "source": [
    "## Add Steele and Peng_Normal Age "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "8e64af52-feba-4247-a218-69767db4af4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_obs_subset.Age = np.where(adata_obs_subset.Dataset == 'Steele', '60s', adata_obs_subset.Age) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "id": "e9c95a4a-be15-47d0-9bf7-9197047b2769",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_obs_subset.Age = np.where(adata_obs_subset.Dataset == 'Steele_Adj_Norm', '60s', adata_obs_subset.Age) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "id": "a6101aef-e67c-4be6-87c2-7d8ce3c0cb86",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_obs_subset.Age = adata_obs_subset.ID_harmonised.map(dict(zip(df_id_merged[df_id_merged.Dataset == 'Peng'].ID_harmonised, \n",
    "         df_id_merged.loc[df_id_merged.Dataset == 'Peng', 'Age']\n",
    "         .apply(lambda x: f\"{int((x // 10) * 10)}s\")\n",
    "         .tolist()))).fillna(adata_obs_subset.Age)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "41240ec9-6157-435d-a5c5-f117a807947f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_obs_subset.Age = adata_obs_subset.ID_harmonised.map(dict(zip(df_id_merged[df_id_merged.Dataset == 'Peng_Normal'].ID_harmonised, \n",
    "         df_id_merged.loc[df_id_merged.Dataset == 'Peng_Normal', 'Age']\n",
    "         .apply(lambda x: f\"{int((x // 10) * 10)}s\")\n",
    "         .tolist()))).fillna(adata_obs_subset.Age)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "9f95e68b-9aeb-403e-bf2f-0b3322feddeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1052958436.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  adata_obs_subset[adata_obs_subset.Dataset.str.contains('Peng')].groupby(['Dataset', 'Age']).size().unstack()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Age</th>\n",
       "      <th>30s</th>\n",
       "      <th>40s</th>\n",
       "      <th>50s</th>\n",
       "      <th>60s</th>\n",
       "      <th>70s</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Caronni</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lee</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lin</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng</th>\n",
       "      <td>2301</td>\n",
       "      <td>1787</td>\n",
       "      <td>22201</td>\n",
       "      <td>11299</td>\n",
       "      <td>3823</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_Normal</th>\n",
       "      <td>4294</td>\n",
       "      <td>2270</td>\n",
       "      <td>3895</td>\n",
       "      <td>4209</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hwang</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schlesinger</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Simeone</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Age               30s   40s    50s    60s   70s\n",
       "Dataset                                        \n",
       "Caronni             0     0      0      0     0\n",
       "Ding                0     0      0      0     0\n",
       "Ding_snRNA-seq      0     0      0      0     0\n",
       "Lee                 0     0      0      0     0\n",
       "Lin                 0     0      0      0     0\n",
       "Peng             2301  1787  22201  11299  3823\n",
       "Peng_Normal      4294  2270   3895   4209     0\n",
       "Hwang               0     0      0      0     0\n",
       "Schlesinger         0     0      0      0     0\n",
       "Simeone             0     0      0      0     0\n",
       "Steele              0     0      0      0     0\n",
       "Steele_Adj_Norm     0     0      0      0     0"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_obs_subset[adata_obs_subset.Dataset.str.contains('Peng')].groupby(['Dataset', 'Age']).size().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "id": "28ebb9f7-548b-4bc6-91de-37c24caa52d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop Age from df_id_merged\n",
    "df_id_merged.drop('Age', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "id": "50bbd83b-3bd2-4bd9-83c2-9e94466d6d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop Tissue\n",
    "df_id_merged.drop('Tissue', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "b6489c51-8210-407c-8343-7de1dd03dcb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop Dataset, Unique_ID, \n",
    "df_id_merged.drop(['Dataset', 'Unique_ID'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "00aad20f-f39e-457c-bcbc-488a0aba7bb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata = adata_obs_subset.merge(df_id_merged, on='ID_harmonised', how='inner').copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 250,
   "id": "0647a732-3fbf-44d5-a8e3-e094ce64125d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 250,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "[i for i in biological_metadata.columns if re.search(r'_x|_y', i)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "id": "195d1c42-a323-4147-b97f-cc4bec06a72c",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.index = adata.obs.index "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "id": "d4536e30-be1d-4fd9-85e1-a3ed40702ded",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.drop('Barcode', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "id": "3710f853-6ae0-463c-a005-863e0413c169",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>ID_batch_covariate</th>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th>Unique_ID</th>\n",
       "      <th>Level_1_refined</th>\n",
       "      <th>Tissue</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>KRAS_mutation</th>\n",
       "      <th>KRAS_type</th>\n",
       "      <th>TP53_type</th>\n",
       "      <th>SMAD4_type</th>\n",
       "      <th>CDKN2A_type</th>\n",
       "      <th>Germline_ATM</th>\n",
       "      <th>Germline_BRCA2</th>\n",
       "      <th>Germline_FANCC</th>\n",
       "      <th>Moffitt_Subtype</th>\n",
       "      <th>Bailey_Subtype</th>\n",
       "      <th>Collisson_Subtype</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>TNM_classification</th>\n",
       "      <th>Overall Survival</th>\n",
       "      <th>Procedure</th>\n",
       "      <th>BMI</th>\n",
       "      <th>perineural invasion</th>\n",
       "      <th>vascular invasion</th>\n",
       "      <th>peripancreatic infiltarion</th>\n",
       "      <th>Harmonized_Grade</th>\n",
       "      <th>Harmonized_TNM_Classification</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset_Barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0</th>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1_Ding_scRNA-seq</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>T Cell</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>40s</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0</th>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1_Ding_scRNA-seq</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>40s</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0</th>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1_Ding_scRNA-seq</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>T Cell</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>40s</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0</th>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1_Ding_scRNA-seq</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>T Cell</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>40s</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0</th>\n",
       "      <td>Ding</td>\n",
       "      <td>HT056P1_Ding_scRNA-seq</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>HT056P1</td>\n",
       "      <td>T Cell</td>\n",
       "      <td>Pancreas</td>\n",
       "      <td>40s</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>p.G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Missense</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Wild_Type</td>\n",
       "      <td>Basal-like</td>\n",
       "      <td>Immunogenic</td>\n",
       "      <td>Quasimesenchymal</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>G3</td>\n",
       "      <td>ypT2 pN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn1_1-1-1</th>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1_Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Ductal Cell</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5, 1</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>GERD, hypertension</td>\n",
       "      <td>Yes, past</td>\n",
       "      <td>No</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>0/13, pN0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 pN0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn1_1-1-1</th>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1_Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Ductal Cell</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5, 1</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>GERD, hypertension</td>\n",
       "      <td>Yes, past</td>\n",
       "      <td>No</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>0/13, pN0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 pN0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn1_1-1-1</th>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1_Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Ductal Cell</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5, 1</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>GERD, hypertension</td>\n",
       "      <td>Yes, past</td>\n",
       "      <td>No</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>0/13, pN0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 pN0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn1_1-1-1</th>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1_Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Fibroblast</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5, 1</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>GERD, hypertension</td>\n",
       "      <td>Yes, past</td>\n",
       "      <td>No</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>0/13, pN0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 pN0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn1_1-1-1</th>\n",
       "      <td>Ding_snRNA-seq</td>\n",
       "      <td>HT412P1_Ding_snRNA-seq</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>HT412P1</td>\n",
       "      <td>Ductal Cell</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5, 1</td>\n",
       "      <td>Head, Uncinated Process</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>GERD, hypertension</td>\n",
       "      <td>Yes, past</td>\n",
       "      <td>No</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>0/13, pN0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 moderately differentiated positive for pe...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ypT1 pN0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>726054 rows × 45 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           Dataset  \\\n",
       "Dataset_Barcode                                                      \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                 Ding   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                 Ding   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                 Ding   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                 Ding   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                 Ding   \n",
       "...                                                            ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  Ding_snRNA-seq   \n",
       "\n",
       "                                                        ID_batch_covariate  \\\n",
       "Dataset_Barcode                                                              \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       HT056P1_Ding_scRNA-seq   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0       HT056P1_Ding_scRNA-seq   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0       HT056P1_Ding_scRNA-seq   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0       HT056P1_Ding_scRNA-seq   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0       HT056P1_Ding_scRNA-seq   \n",
       "...                                                                    ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  HT412P1_Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  HT412P1_Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  HT412P1_Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  HT412P1_Ding_snRNA-seq   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  HT412P1_Ding_snRNA-seq   \n",
       "\n",
       "                                                   ID_harmonised Unique_ID  \\\n",
       "Dataset_Barcode                                                              \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0            HT056P1   HT056P1   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0            HT056P1   HT056P1   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0            HT056P1   HT056P1   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0            HT056P1   HT056P1   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0            HT056P1   HT056P1   \n",
       "...                                                          ...       ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...       HT412P1   HT412P1   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...       HT412P1   HT412P1   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...       HT412P1   HT412P1   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...       HT412P1   HT412P1   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...       HT412P1   HT412P1   \n",
       "\n",
       "                                                   Level_1_refined    Tissue  \\\n",
       "Dataset_Barcode                                                                \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0               T Cell  Pancreas   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           Fibroblast  Pancreas   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0               T Cell  Pancreas   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0               T Cell  Pancreas   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0               T Cell  Pancreas   \n",
       "...                                                            ...       ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...     Ductal Cell   Unknown   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...     Ductal Cell   Unknown   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...     Ductal Cell   Unknown   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...      Fibroblast   Unknown   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...     Ductal Cell   Unknown   \n",
       "\n",
       "                                                        Age      Sex Diabetes  \\\n",
       "Dataset_Barcode                                                                 \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0           40s   Female       No   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           40s   Female       No   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0           40s   Female       No   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0           40s   Female       No   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0           40s   Female       No   \n",
       "...                                                     ...      ...      ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Unknown  Unknown  Unknown   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Unknown  Unknown  Unknown   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Unknown  Unknown  Unknown   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Unknown  Unknown  Unknown   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  Unknown  Unknown  Unknown   \n",
       "\n",
       "                                                     Treatment KRAS_mutation  \\\n",
       "Dataset_Barcode                                                                \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       FOLFIRINOX        p.G12V   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0       FOLFIRINOX        p.G12V   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0       FOLFIRINOX        p.G12V   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0       FOLFIRINOX        p.G12V   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0       FOLFIRINOX        p.G12V   \n",
       "...                                                        ...           ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...     Unknown           NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...     Unknown           NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...     Unknown           NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...     Unknown           NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...     Unknown           NaN   \n",
       "\n",
       "                                                   KRAS_type TP53_type  \\\n",
       "Dataset_Barcode                                                          \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       Missense  Missense   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0       Missense  Missense   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0       Missense  Missense   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0       Missense  Missense   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0       Missense  Missense   \n",
       "...                                                      ...       ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN       NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN       NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN       NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN       NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN       NaN   \n",
       "\n",
       "                                                   SMAD4_type CDKN2A_type  \\\n",
       "Dataset_Barcode                                                             \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       Wild_Type   Wild_Type   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0       Wild_Type   Wild_Type   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0       Wild_Type   Wild_Type   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0       Wild_Type   Wild_Type   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0       Wild_Type   Wild_Type   \n",
       "...                                                       ...         ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...        NaN         NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...        NaN         NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...        NaN         NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...        NaN         NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...        NaN         NaN   \n",
       "\n",
       "                                                   Germline_ATM  \\\n",
       "Dataset_Barcode                                                   \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0         Wild_Type   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0         Wild_Type   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0         Wild_Type   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0         Wild_Type   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0         Wild_Type   \n",
       "...                                                         ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "\n",
       "                                                   Germline_BRCA2  \\\n",
       "Dataset_Barcode                                                     \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "...                                                           ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "\n",
       "                                                   Germline_FANCC  \\\n",
       "Dataset_Barcode                                                     \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0           Wild_Type   \n",
       "...                                                           ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "\n",
       "                                                   Moffitt_Subtype  \\\n",
       "Dataset_Barcode                                                      \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0           Basal-like   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           Basal-like   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0           Basal-like   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0           Basal-like   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0           Basal-like   \n",
       "...                                                            ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...             NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...             NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...             NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...             NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...             NaN   \n",
       "\n",
       "                                                   Bailey_Subtype  \\\n",
       "Dataset_Barcode                                                     \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0         Immunogenic   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0         Immunogenic   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0         Immunogenic   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0         Immunogenic   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0         Immunogenic   \n",
       "...                                                           ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...            NaN   \n",
       "\n",
       "                                                   Collisson_Subtype  \\\n",
       "Dataset_Barcode                                                        \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       Quasimesenchymal   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0       Quasimesenchymal   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0       Quasimesenchymal   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0       Quasimesenchymal   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0       Quasimesenchymal   \n",
       "...                                                              ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "\n",
       "                                                    Time_from_diagnosis_to_surgery_(weeks)  \\\n",
       "Dataset_Barcode                                                                              \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                                         20.0   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                                         20.0   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                                         20.0   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                                         20.0   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                                         20.0   \n",
       "...                                                                                    ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...                                    12.0   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...                                    12.0   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...                                    12.0   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...                                    12.0   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...                                    12.0   \n",
       "\n",
       "                                                   Cycles  \\\n",
       "Dataset_Barcode                                             \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0           6   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           6   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0           6   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0           6   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0           6   \n",
       "...                                                   ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...   5, 1   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...   5, 1   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...   5, 1   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...   5, 1   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...   5, 1   \n",
       "\n",
       "                                                                   Location  \\\n",
       "Dataset_Barcode                                                               \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0       Head, Uncinated Process   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0       Head, Uncinated Process   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0       Head, Uncinated Process   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0       Head, Uncinated Process   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0       Head, Uncinated Process   \n",
       "...                                                                     ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Head, Uncinated Process   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Head, Uncinated Process   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Head, Uncinated Process   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Head, Uncinated Process   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  Head, Uncinated Process   \n",
       "\n",
       "                                                   Grade  Size_(cm) Stage  \\\n",
       "Dataset_Barcode                                                             \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0        3.0        3.2   NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0        3.0        3.2   NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0        3.0        3.2   NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0        3.0        3.2   NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0        3.0        3.2   NaN   \n",
       "...                                                  ...        ...   ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...   NaN        1.9   NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...   NaN        1.9   NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...   NaN        1.9   NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...   NaN        1.9   NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...   NaN        1.9   NaN   \n",
       "\n",
       "                                                   Pancreatitis  \\\n",
       "Dataset_Barcode                                                   \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0               Yes   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0               Yes   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0               Yes   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0               Yes   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0               Yes   \n",
       "...                                                         ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...           No   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...           No   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...           No   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...           No   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...           No   \n",
       "\n",
       "                                                   Previous_cancer_history  \\\n",
       "Dataset_Barcode                                                              \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                           No   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                           No   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                           No   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                           No   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                           No   \n",
       "...                                                                    ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...                      No   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...                      No   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...                      No   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...                      No   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...                      No   \n",
       "\n",
       "                                                   Other_co-morbidities  \\\n",
       "Dataset_Barcode                                                           \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                        No   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                        No   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                        No   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                        No   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                        No   \n",
       "...                                                                 ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...   GERD, hypertension   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...   GERD, hypertension   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...   GERD, hypertension   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...   GERD, hypertension   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...   GERD, hypertension   \n",
       "\n",
       "                                                       Smoker     Alcohol  \\\n",
       "Dataset_Barcode                                                             \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0              No  Occasional   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0              No  Occasional   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0              No  Occasional   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0              No  Occasional   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0              No  Occasional   \n",
       "...                                                       ...         ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Yes, past          No   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Yes, past          No   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  Yes, past          No   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  Yes, past          No   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  Yes, past          No   \n",
       "\n",
       "                                                                                      Pathology_notes  \\\n",
       "Dataset_Barcode                                                                                         \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                  ypT2, focal sarcomatoid transformation   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                  ypT2, focal sarcomatoid transformation   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                  ypT2, focal sarcomatoid transformation   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                  ypT2, focal sarcomatoid transformation   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                  ypT2, focal sarcomatoid transformation   \n",
       "...                                                                                               ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "\n",
       "                                                   Lymph_node_met  \\\n",
       "Dataset_Barcode                                                     \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0           2/21, pN1   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0           2/21, pN1   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0           2/21, pN1   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0           2/21, pN1   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0           2/21, pN1   \n",
       "...                                                           ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...      0/13, pN0   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...      0/13, pN0   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...      0/13, pN0   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...      0/13, pN0   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...      0/13, pN0   \n",
       "\n",
       "                                                    Survival_post_surgery_(weeks)  \\\n",
       "Dataset_Barcode                                                                     \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                                 NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                                 NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                                 NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                                 NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                                 NaN   \n",
       "...                                                                           ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...                            NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...                            NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...                            NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...                            NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...                            NaN   \n",
       "\n",
       "                                                   Vital_status  \\\n",
       "Dataset_Barcode                                                   \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0               NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0               NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0               NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0               NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0               NaN   \n",
       "...                                                         ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...          NaN   \n",
       "\n",
       "                                                                                   TNM_classification  \\\n",
       "Dataset_Barcode                                                                                         \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                                                ypT2 pN1   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                                                ypT2 pN1   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                                                ypT2 pN1   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                                                ypT2 pN1   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                                                ypT2 pN1   \n",
       "...                                                                                               ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...  ypT1 moderately differentiated positive for pe...   \n",
       "\n",
       "                                                    Overall Survival  \\\n",
       "Dataset_Barcode                                                        \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "...                                                              ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "\n",
       "                                                   Procedure  BMI  \\\n",
       "Dataset_Barcode                                                     \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0            NaN  NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0            NaN  NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0            NaN  NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0            NaN  NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0            NaN  NaN   \n",
       "...                                                      ...  ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN  NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN  NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN  NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN  NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...       NaN  NaN   \n",
       "\n",
       "                                                   perineural invasion  \\\n",
       "Dataset_Barcode                                                          \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                      NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                      NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                      NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                      NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                      NaN   \n",
       "...                                                                ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...                 NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...                 NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...                 NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...                 NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...                 NaN   \n",
       "\n",
       "                                                   vascular invasion  \\\n",
       "Dataset_Barcode                                                        \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                    NaN   \n",
       "...                                                              ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...               NaN   \n",
       "\n",
       "                                                   peripancreatic infiltarion  \\\n",
       "Dataset_Barcode                                                                 \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                             NaN   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                             NaN   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                             NaN   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                             NaN   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                             NaN   \n",
       "...                                                                       ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...                        NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...                        NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...                        NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...                        NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...                        NaN   \n",
       "\n",
       "                                                   Harmonized_Grade  \\\n",
       "Dataset_Barcode                                                       \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                    G3   \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                    G3   \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                    G3   \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                    G3   \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                    G3   \n",
       "...                                                             ...   \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...              NaN   \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...              NaN   \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...              NaN   \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...              NaN   \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...              NaN   \n",
       "\n",
       "                                                   Harmonized_TNM_Classification  \n",
       "Dataset_Barcode                                                                   \n",
       "Ding_AAACCCAAGTCATCGT-1-HT056P-S1PAA1-0-0-0-0                           ypT2 pN1  \n",
       "Ding_AAACCCACAACCACGC-1-HT056P-S1PAA1-0-0-0-0                           ypT2 pN1  \n",
       "Ding_AAACCCACAACTCCCT-1-HT056P-S1PAA1-0-0-0-0                           ypT2 pN1  \n",
       "Ding_AAACCCACAGACAATA-1-HT056P-S1PAA1-0-0-0-0                           ypT2 pN1  \n",
       "Ding_AAACCCACATGCCGAC-1-HT056P-S1PAA1-0-0-0-0                           ypT2 pN1  \n",
       "...                                                                          ...  \n",
       "Ding_TTTGTTGGTCTTACTA-1-HT412P1-S1H1A3N1Z1_1Bmn...                      ypT1 pN0  \n",
       "Ding_TTTGTTGGTTGAGCCG-1-HT412P1-S1H1A3N1Z1_1Bmn...                      ypT1 pN0  \n",
       "Ding_TTTGTTGGTTGCACGG-1-HT412P1-S1H1A3N1Z1_1Bmn...                      ypT1 pN0  \n",
       "Ding_TTTGTTGGTTGGTTGA-1-HT412P1-S1H1A3N1Z1_1Bmn...                      ypT1 pN0  \n",
       "Ding_TTTGTTGGTTTATCGC-1-HT412P1-S1H1A3N1Z1_1Bmn...                      ypT1 pN0  \n",
       "\n",
       "[726054 rows x 45 columns]"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "7da27b51-4005-47bf-9719-32ad7865f130",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1531259420.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  df = biological_metadata.groupby('Dataset').nunique() #.apply(lambda x: x.notna().nunique())\n"
     ]
    }
   ],
   "source": [
    "df = biological_metadata.groupby('Dataset').nunique() #.apply(lambda x: x.notna().nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "a8f73772-8373-4eb5-9378-f46e75a6929f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID_batch_covariate: 12/12\n",
      "ID_harmonised: 12/12\n",
      "Unique_ID: 12/12\n",
      "Level_1_refined: 12/12\n",
      "Tissue: 12/12\n",
      "Age: 12/12\n",
      "Sex: 12/12\n",
      "Diabetes: 12/12\n",
      "Treatment: 12/12\n",
      "TNM_classification: 6/12\n",
      "Harmonized_TNM_Classification: 6/12\n",
      "Grade: 5/12\n",
      "Stage: 5/12\n",
      "Harmonized_Grade: 5/12\n",
      "Pathology_notes: 4/12\n",
      "Location: 3/12\n",
      "Vital_status: 3/12\n",
      "Overall Survival: 3/12\n",
      "Procedure: 3/12\n",
      "KRAS_mutation: 2/12\n",
      "TP53_type: 2/12\n",
      "SMAD4_type: 2/12\n",
      "CDKN2A_type: 2/12\n",
      "Time_from_diagnosis_to_surgery_(weeks): 2/12\n",
      "Cycles: 2/12\n",
      "Size_(cm): 2/12\n",
      "Pancreatitis: 2/12\n",
      "Previous_cancer_history: 2/12\n",
      "Other_co-morbidities: 2/12\n",
      "Smoker: 2/12\n",
      "Alcohol: 2/12\n",
      "Lymph_node_met: 2/12\n",
      "Survival_post_surgery_(weeks): 2/12\n",
      "BMI: 2/12\n",
      "perineural invasion: 2/12\n",
      "vascular invasion: 2/12\n",
      "peripancreatic infiltarion: 2/12\n",
      "KRAS_type: 1/12\n",
      "Germline_ATM: 1/12\n",
      "Germline_BRCA2: 1/12\n",
      "Germline_FANCC: 1/12\n",
      "Moffitt_Subtype: 1/12\n",
      "Bailey_Subtype: 1/12\n",
      "Collisson_Subtype: 1/12\n"
     ]
    }
   ],
   "source": [
    "counts = {}\n",
    "for col in df:\n",
    "    counts[col] = np.count_nonzero(df[col])  \n",
    "sorted_counts = sorted(counts.items(), key=lambda item: item[1], reverse=True)\n",
    "for col, count in sorted_counts:\n",
    "    print(f'{col}: {count}/{len(df[col].index)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "feaadc22-87e2-4bd8-b65f-741abac39cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/687848908.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Treatment']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treatment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abraxane</th>\n",
       "      <td>4093</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX</th>\n",
       "      <td>9731</td>\n",
       "      <td>61111</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45734</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + Gem/Abraxane</th>\n",
       "      <td>0</td>\n",
       "      <td>2383</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + Losartan + Radiotherapy + Capecitabine/5-FU</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6957</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + RT</th>\n",
       "      <td>0</td>\n",
       "      <td>2738</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + Radiotherapy + Capecitabine/5-FU</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66876</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + SBRT + Losartan + Nivolumab</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + SBRT + Nivolumab</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4994</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gem/Abraxane</th>\n",
       "      <td>14240</td>\n",
       "      <td>42646</td>\n",
       "      <td>0</td>\n",
       "      <td>1337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13819</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15781</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treatment Naive</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35928</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>58888</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14668</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Untreated</th>\n",
       "      <td>43921</td>\n",
       "      <td>36060</td>\n",
       "      <td>0</td>\n",
       "      <td>6799</td>\n",
       "      <td>8019</td>\n",
       "      <td>41411</td>\n",
       "      <td>0</td>\n",
       "      <td>90399</td>\n",
       "      <td>0</td>\n",
       "      <td>82435</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                                             Caronni   Ding  \\\n",
       "Treatment                                                            \n",
       "Abraxane                                               4093      0   \n",
       "FOLFIRINOX                                             9731  61111   \n",
       "FOLFIRINOX + Gem/Abraxane                                 0   2383   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...        0      0   \n",
       "FOLFIRINOX + RT                                           0   2738   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU             0      0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                  0      0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                             0      0   \n",
       "Gem/Abraxane                                          14240  42646   \n",
       "Other                                                     0      0   \n",
       "Treatment Naive                                           0      0   \n",
       "Unknown                                                   0   2045   \n",
       "Untreated                                             43921  36060   \n",
       "\n",
       "Dataset                                             Ding_snRNA-seq   Lee  \\\n",
       "Treatment                                                                  \n",
       "Abraxane                                                         0     0   \n",
       "FOLFIRINOX                                                       0     0   \n",
       "FOLFIRINOX + Gem/Abraxane                                        0     0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...               0     0   \n",
       "FOLFIRINOX + RT                                                  0     0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU                    0     0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                         0     0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                                    0     0   \n",
       "Gem/Abraxane                                                     0  1337   \n",
       "Other                                                            0     0   \n",
       "Treatment Naive                                                  0     0   \n",
       "Unknown                                                      58888     0   \n",
       "Untreated                                                        0  6799   \n",
       "\n",
       "Dataset                                              Lin   Peng  Peng_Normal  \\\n",
       "Treatment                                                                      \n",
       "Abraxane                                               0      0            0   \n",
       "FOLFIRINOX                                             0      0            0   \n",
       "FOLFIRINOX + Gem/Abraxane                              0      0            0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...     0      0            0   \n",
       "FOLFIRINOX + RT                                        0      0            0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU          0      0            0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab               0      0            0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                          0      0            0   \n",
       "Gem/Abraxane                                           0      0            0   \n",
       "Other                                                  0      0            0   \n",
       "Treatment Naive                                        0      0            0   \n",
       "Unknown                                                0      0        14668   \n",
       "Untreated                                           8019  41411            0   \n",
       "\n",
       "Dataset                                             Hwang  Schlesinger  \\\n",
       "Treatment                                                                \n",
       "Abraxane                                                0            0   \n",
       "FOLFIRINOX                                              0            0   \n",
       "FOLFIRINOX + Gem/Abraxane                               0            0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...   6957            0   \n",
       "FOLFIRINOX + RT                                         0            0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU       66876            0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab             1323            0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                        4994            0   \n",
       "Gem/Abraxane                                            0            0   \n",
       "Other                                               15781            0   \n",
       "Treatment Naive                                         0            0   \n",
       "Unknown                                                 0         5387   \n",
       "Untreated                                           90399            0   \n",
       "\n",
       "Dataset                                             Simeone  Steele  \\\n",
       "Treatment                                                             \n",
       "Abraxane                                                  0       0   \n",
       "FOLFIRINOX                                            45734       0   \n",
       "FOLFIRINOX + Gem/Abraxane                                 0       0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...        0       0   \n",
       "FOLFIRINOX + RT                                           0       0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU             0       0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                  0       0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                             0       0   \n",
       "Gem/Abraxane                                          13819       0   \n",
       "Other                                                     0       0   \n",
       "Treatment Naive                                           0   35928   \n",
       "Unknown                                                   0       0   \n",
       "Untreated                                             82435       0   \n",
       "\n",
       "Dataset                                             Steele_Adj_Norm  \n",
       "Treatment                                                            \n",
       "Abraxane                                                          0  \n",
       "FOLFIRINOX                                                        0  \n",
       "FOLFIRINOX + Gem/Abraxane                                         0  \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...                0  \n",
       "FOLFIRINOX + RT                                                   0  \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU                     0  \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                          0  \n",
       "FOLFIRINOX + SBRT + Nivolumab                                     0  \n",
       "Gem/Abraxane                                                      0  \n",
       "Other                                                             0  \n",
       "Treatment Naive                                                   0  \n",
       "Unknown                                                        6331  \n",
       "Untreated                                                         0  "
      ]
     },
     "execution_count": 264,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Treatment']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d20a1fd-65c8-411b-956d-fa48cf198621",
   "metadata": {},
   "source": [
    "# Resolved Ding Treatment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "66ccfbf9-9792-4469-8383-0d628e9b9328",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Treatment</th>\n",
       "      <th>FOLFIRINOX</th>\n",
       "      <th>FOLFIRINOX + Gem/Abraxane</th>\n",
       "      <th>FOLFIRINOX + RT</th>\n",
       "      <th>Gem/Abraxane</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Untreated</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HT056P1</th>\n",
       "      <td>7375.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT060P1</th>\n",
       "      <td>5699.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT061P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3457.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT064P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3889.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT071P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2738.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT085P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1267.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT115P1</th>\n",
       "      <td>5308.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT121P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3663.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT122P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8642.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT123P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11478.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT124P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3238.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT125P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18187.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT138P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10570.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT140P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9068.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT166P1</th>\n",
       "      <td>10533.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT168P1</th>\n",
       "      <td>13830.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT185B1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2045.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT185P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2383.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT190P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5247.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT191P1</th>\n",
       "      <td>7065.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT200P1</th>\n",
       "      <td>5744.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT204P1</th>\n",
       "      <td>5557.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT224P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2201.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT231P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4172.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT232P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4226.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT242P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5826.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT259P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4472.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT264P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3816.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT270P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6794.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT284P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5084.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT288P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7994.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT306P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5876.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT412P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8427.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Treatment      FOLFIRINOX  FOLFIRINOX + Gem/Abraxane  FOLFIRINOX + RT  \\\n",
       "ID_harmonised                                                           \n",
       "HT056P1            7375.0                        0.0              0.0   \n",
       "HT060P1            5699.0                        0.0              0.0   \n",
       "HT061P1               0.0                        0.0              0.0   \n",
       "HT064P1               0.0                        0.0              0.0   \n",
       "HT071P1               0.0                        0.0           2738.0   \n",
       "HT085P1               0.0                        0.0              0.0   \n",
       "HT115P1            5308.0                        0.0              0.0   \n",
       "HT121P1               0.0                        0.0              0.0   \n",
       "HT122P1               0.0                        0.0              0.0   \n",
       "HT123P1               0.0                        0.0              0.0   \n",
       "HT124P1               0.0                        0.0              0.0   \n",
       "HT125P1               0.0                        0.0              0.0   \n",
       "HT138P1               0.0                        0.0              0.0   \n",
       "HT140P1               0.0                        0.0              0.0   \n",
       "HT166P1           10533.0                        0.0              0.0   \n",
       "HT168P1           13830.0                        0.0              0.0   \n",
       "HT185B1               0.0                        0.0              0.0   \n",
       "HT185P1               0.0                     2383.0              0.0   \n",
       "HT190P1               0.0                        0.0              0.0   \n",
       "HT191P1            7065.0                        0.0              0.0   \n",
       "HT200P1            5744.0                        0.0              0.0   \n",
       "HT204P1            5557.0                        0.0              0.0   \n",
       "HT224P1               0.0                        0.0              0.0   \n",
       "HT231P1               0.0                        0.0              0.0   \n",
       "HT232P1               0.0                        0.0              0.0   \n",
       "HT242P1               0.0                        0.0              0.0   \n",
       "HT259P1               0.0                        0.0              0.0   \n",
       "HT264P1               0.0                        0.0              0.0   \n",
       "HT270P1               0.0                        0.0              0.0   \n",
       "HT284P1               0.0                        0.0              0.0   \n",
       "HT288P1               0.0                        0.0              0.0   \n",
       "HT306P1               0.0                        0.0              0.0   \n",
       "HT412P1               0.0                        0.0              0.0   \n",
       "\n",
       "Treatment      Gem/Abraxane  Unknown  Untreated  \n",
       "ID_harmonised                                    \n",
       "HT056P1                 0.0      0.0        0.0  \n",
       "HT060P1                 0.0      0.0        0.0  \n",
       "HT061P1                 0.0      0.0     3457.0  \n",
       "HT064P1                 0.0      0.0     3889.0  \n",
       "HT071P1                 0.0      0.0        0.0  \n",
       "HT085P1                 0.0      0.0     1267.0  \n",
       "HT115P1                 0.0      0.0        0.0  \n",
       "HT121P1                 0.0      0.0     3663.0  \n",
       "HT122P1              8642.0      0.0        0.0  \n",
       "HT123P1                 0.0      0.0    11478.0  \n",
       "HT124P1                 0.0      0.0     3238.0  \n",
       "HT125P1             18187.0      0.0        0.0  \n",
       "HT138P1             10570.0      0.0        0.0  \n",
       "HT140P1                 0.0      0.0     9068.0  \n",
       "HT166P1                 0.0      0.0        0.0  \n",
       "HT168P1                 0.0      0.0        0.0  \n",
       "HT185B1                 0.0   2045.0        0.0  \n",
       "HT185P1                 0.0      0.0        0.0  \n",
       "HT190P1              5247.0      0.0        0.0  \n",
       "HT191P1                 0.0      0.0        0.0  \n",
       "HT200P1                 0.0      0.0        0.0  \n",
       "HT204P1                 0.0      0.0        0.0  \n",
       "HT224P1                 0.0   2201.0        0.0  \n",
       "HT231P1                 0.0   4172.0        0.0  \n",
       "HT232P1                 0.0   4226.0        0.0  \n",
       "HT242P1                 0.0   5826.0        0.0  \n",
       "HT259P1                 0.0   4472.0        0.0  \n",
       "HT264P1                 0.0   3816.0        0.0  \n",
       "HT270P1                 0.0   6794.0        0.0  \n",
       "HT284P1                 0.0   5084.0        0.0  \n",
       "HT288P1                 0.0   7994.0        0.0  \n",
       "HT306P1                 0.0   5876.0        0.0  \n",
       "HT412P1                 0.0   8427.0        0.0  "
      ]
     },
     "execution_count": 284,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata[biological_metadata.Dataset.str.contains('Ding')].groupby(['ID_harmonised', 'Treatment']).size().unstack() #.fillna(0).apply(pd.to_numeric, errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "id": "66691c21-0e24-48bf-adaa-685c87fcbb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_ids = biological_metadata[(biological_metadata.Treatment == 'Unknown') & (biological_metadata.Dataset.str.contains('Ding'))].ID_harmonised.unique().tolist() #.groupby('ID_harmonised').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "id": "88c4debe-8a17-49f8-8c71-496be9aa4aeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 291,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ding_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "id": "3266e932-5785-4778-a3ce-4167ad8870da",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding = pd.read_excel('../Downstream/metadata_publications/Ding_metadata.xlsx', sheet_name='Table1_Clinical_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "149bb6be-55d2-485e-b007-74a64e7f5870",
   "metadata": {},
   "outputs": [],
   "source": [
    "overlapping_ding_id = list(set(ding.Patient_ID) & set(ding_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 300,
   "id": "f552b505-6a81-46f7-b767-73f54e1bb494",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21    HT264P1\n",
       "22    HT270P1\n",
       "23    HT284P1\n",
       "24    HT288P1\n",
       "25    HT224P1\n",
       "26    HT231P1\n",
       "27    HT232P1\n",
       "28    HT259P1\n",
       "29    HT306P1\n",
       "30    HT242P1\n",
       "31    HT412P1\n",
       "Name: Patient_ID, dtype: object"
      ]
     },
     "execution_count": 300,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding[ding.Patient_ID.isin(overlapping_ding_id)].Patient_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 301,
   "id": "101fb23e-5bba-4a0c-bf65-8f860a6d52f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HT264P1': 'naive',\n",
       " 'HT270P1': 'naive',\n",
       " 'HT284P1': 'naive',\n",
       " 'HT288P1': 'Naive',\n",
       " 'HT224P1': 'FOLIRINOX, Gem/Abraxane',\n",
       " 'HT231P1': 'FOLFIRINOX, SBRT',\n",
       " 'HT232P1': 'mFOLFIRINOX, FOLFIRI',\n",
       " 'HT259P1': 'FOLFIRINOX',\n",
       " 'HT306P1': 'FOLFIRINOX',\n",
       " 'HT242P1': nan,\n",
       " 'HT412P1': 'FOLFIRONOX, Gemcitabine'}"
      ]
     },
     "execution_count": 301,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(zip(ding[ding.Patient_ID.isin(overlapping_ding_id)].Patient_ID, ding[ding.Patient_ID.isin(overlapping_ding_id)].Treatment))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 304,
   "id": "81c1d255-8f87-4618-b77a-09a16ea08ae3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FOLFIRINOX',\n",
       " 'Untreated',\n",
       " 'FOLFIRINOX + RT',\n",
       " 'Gem/Abraxane',\n",
       " 'Unknown',\n",
       " 'FOLFIRINOX + Gem/Abraxane',\n",
       " 'Treatment Naive',\n",
       " 'Abraxane',\n",
       " 'Other',\n",
       " 'FOLFIRINOX + Radiotherapy + Capecitabine/5-FU',\n",
       " 'FOLFIRINOX + Losartan + Radiotherapy + Capecitabine/5-FU',\n",
       " 'FOLFIRINOX + SBRT + Nivolumab',\n",
       " 'FOLFIRINOX + SBRT + Losartan + Nivolumab']"
      ]
     },
     "execution_count": 304,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.Treatment.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 306,
   "id": "b4e9b3db-cccb-4cd1-a58d-887ed800cdab",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_treament_dict = {'HT264P1': 'Treatment Naive',\n",
    " 'HT270P1': 'Treatment Naive',\n",
    " 'HT284P1': 'Treatment Naive',\n",
    " 'HT288P1': 'Treatment Naive',\n",
    " 'HT224P1': 'FOLIRINOX + Gem/Abraxane',\n",
    " 'HT231P1': 'FOLFIRINOX + SBRT',\n",
    " 'HT232P1': 'mFOLFIRINOX + FOLFIRI',\n",
    " 'HT259P1': 'FOLFIRINOX',\n",
    " 'HT306P1': 'FOLFIRINOX',\n",
    " 'HT242P1': 'Unknown',\n",
    " 'HT412P1': 'FOLFIRONO + Gemcitabine'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 307,
   "id": "b896a48e-7401-4034-b5a3-26be803298d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Treatment = biological_metadata.ID_harmonised.map(ding_treament_dict).fillna(biological_metadata.Treatment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 314,
   "id": "a67af8ab-7960-431a-b812-22f5e2e0edaa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Treatment</th>\n",
       "      <th>FOLFIRINOX</th>\n",
       "      <th>FOLFIRINOX + Gem/Abraxane</th>\n",
       "      <th>FOLFIRINOX + RT</th>\n",
       "      <th>FOLFIRINOX + SBRT</th>\n",
       "      <th>FOLFIRONO + Gemcitabine</th>\n",
       "      <th>FOLIRINOX + Gem/Abraxane</th>\n",
       "      <th>Gem/Abraxane</th>\n",
       "      <th>Treatment Naive</th>\n",
       "      <th>Unknown</th>\n",
       "      <th>Untreated</th>\n",
       "      <th>mFOLFIRINOX + FOLFIRI</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ID_harmonised</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HT231P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4172.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT232P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4226.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT242P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5826.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT259P1</th>\n",
       "      <td>4472.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT264P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3816.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT270P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6794.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT284P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5084.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT288P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7994.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT306P1</th>\n",
       "      <td>5876.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HT412P1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8427.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Treatment      FOLFIRINOX  FOLFIRINOX + Gem/Abraxane  FOLFIRINOX + RT  \\\n",
       "ID_harmonised                                                           \n",
       "HT231P1               0.0                        0.0              0.0   \n",
       "HT232P1               0.0                        0.0              0.0   \n",
       "HT242P1               0.0                        0.0              0.0   \n",
       "HT259P1            4472.0                        0.0              0.0   \n",
       "HT264P1               0.0                        0.0              0.0   \n",
       "HT270P1               0.0                        0.0              0.0   \n",
       "HT284P1               0.0                        0.0              0.0   \n",
       "HT288P1               0.0                        0.0              0.0   \n",
       "HT306P1            5876.0                        0.0              0.0   \n",
       "HT412P1               0.0                        0.0              0.0   \n",
       "\n",
       "Treatment      FOLFIRINOX + SBRT  FOLFIRONO + Gemcitabine  \\\n",
       "ID_harmonised                                               \n",
       "HT231P1                   4172.0                      0.0   \n",
       "HT232P1                      0.0                      0.0   \n",
       "HT242P1                      0.0                      0.0   \n",
       "HT259P1                      0.0                      0.0   \n",
       "HT264P1                      0.0                      0.0   \n",
       "HT270P1                      0.0                      0.0   \n",
       "HT284P1                      0.0                      0.0   \n",
       "HT288P1                      0.0                      0.0   \n",
       "HT306P1                      0.0                      0.0   \n",
       "HT412P1                      0.0                   8427.0   \n",
       "\n",
       "Treatment      FOLIRINOX + Gem/Abraxane  Gem/Abraxane  Treatment Naive  \\\n",
       "ID_harmonised                                                            \n",
       "HT231P1                             0.0           0.0              0.0   \n",
       "HT232P1                             0.0           0.0              0.0   \n",
       "HT242P1                             0.0           0.0              0.0   \n",
       "HT259P1                             0.0           0.0              0.0   \n",
       "HT264P1                             0.0           0.0           3816.0   \n",
       "HT270P1                             0.0           0.0           6794.0   \n",
       "HT284P1                             0.0           0.0           5084.0   \n",
       "HT288P1                             0.0           0.0           7994.0   \n",
       "HT306P1                             0.0           0.0              0.0   \n",
       "HT412P1                             0.0           0.0              0.0   \n",
       "\n",
       "Treatment      Unknown  Untreated  mFOLFIRINOX + FOLFIRI  \n",
       "ID_harmonised                                             \n",
       "HT231P1            0.0        0.0                    0.0  \n",
       "HT232P1            0.0        0.0                 4226.0  \n",
       "HT242P1         5826.0        0.0                    0.0  \n",
       "HT259P1            0.0        0.0                    0.0  \n",
       "HT264P1            0.0        0.0                    0.0  \n",
       "HT270P1            0.0        0.0                    0.0  \n",
       "HT284P1            0.0        0.0                    0.0  \n",
       "HT288P1            0.0        0.0                    0.0  \n",
       "HT306P1            0.0        0.0                    0.0  \n",
       "HT412P1            0.0        0.0                    0.0  "
      ]
     },
     "execution_count": 314,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata[biological_metadata.Dataset.str.contains('Ding')].groupby(['ID_harmonised', 'Treatment']).size().unstack().fillna(0).apply(pd.to_numeric, errors='coerce').tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "faccdb23-7314-4913-bfb2-243ec4244be2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HT185B1', 'HT242P1'], dtype=object)"
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata[(biological_metadata.Dataset.str.contains('Ding')) & (biological_metadata.Treatment == 'Unknown')].ID_harmonised.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "id": "03405152-80c5-458b-8f1e-d36e0bc20adc",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Treatment = biological_metadata.Treatment.replace('Untreated', 'Treatment Naive')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "6af5a247-97c8-4549-a599-48f20fbe0a2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/687848908.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Treatment']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treatment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Abraxane</th>\n",
       "      <td>4093</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX</th>\n",
       "      <td>9731</td>\n",
       "      <td>61111</td>\n",
       "      <td>10348</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45734</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + Gem/Abraxane</th>\n",
       "      <td>0</td>\n",
       "      <td>2383</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + Losartan + Radiotherapy + Capecitabine/5-FU</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6957</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + RT</th>\n",
       "      <td>0</td>\n",
       "      <td>2738</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + Radiotherapy + Capecitabine/5-FU</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66876</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + SBRT</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4172</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + SBRT + Losartan + Nivolumab</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1323</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRINOX + SBRT + Nivolumab</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4994</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLFIRONO + Gemcitabine</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8427</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOLIRINOX + Gem/Abraxane</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2201</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gem/Abraxane</th>\n",
       "      <td>14240</td>\n",
       "      <td>42646</td>\n",
       "      <td>0</td>\n",
       "      <td>1337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13819</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Other</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15781</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treatment Naive</th>\n",
       "      <td>43921</td>\n",
       "      <td>36060</td>\n",
       "      <td>23688</td>\n",
       "      <td>6799</td>\n",
       "      <td>8019</td>\n",
       "      <td>41411</td>\n",
       "      <td>0</td>\n",
       "      <td>90399</td>\n",
       "      <td>0</td>\n",
       "      <td>82435</td>\n",
       "      <td>35928</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>5826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14668</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mFOLFIRINOX + FOLFIRI</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4226</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                                             Caronni   Ding  \\\n",
       "Treatment                                                            \n",
       "Abraxane                                               4093      0   \n",
       "FOLFIRINOX                                             9731  61111   \n",
       "FOLFIRINOX + Gem/Abraxane                                 0   2383   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...        0      0   \n",
       "FOLFIRINOX + RT                                           0   2738   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU             0      0   \n",
       "FOLFIRINOX + SBRT                                         0      0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                  0      0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                             0      0   \n",
       "FOLFIRONO + Gemcitabine                                   0      0   \n",
       "FOLIRINOX + Gem/Abraxane                                  0      0   \n",
       "Gem/Abraxane                                          14240  42646   \n",
       "Other                                                     0      0   \n",
       "Treatment Naive                                       43921  36060   \n",
       "Unknown                                                   0   2045   \n",
       "mFOLFIRINOX + FOLFIRI                                     0      0   \n",
       "\n",
       "Dataset                                             Ding_snRNA-seq   Lee  \\\n",
       "Treatment                                                                  \n",
       "Abraxane                                                         0     0   \n",
       "FOLFIRINOX                                                   10348     0   \n",
       "FOLFIRINOX + Gem/Abraxane                                        0     0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...               0     0   \n",
       "FOLFIRINOX + RT                                                  0     0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU                    0     0   \n",
       "FOLFIRINOX + SBRT                                             4172     0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                         0     0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                                    0     0   \n",
       "FOLFIRONO + Gemcitabine                                       8427     0   \n",
       "FOLIRINOX + Gem/Abraxane                                      2201     0   \n",
       "Gem/Abraxane                                                     0  1337   \n",
       "Other                                                            0     0   \n",
       "Treatment Naive                                              23688  6799   \n",
       "Unknown                                                       5826     0   \n",
       "mFOLFIRINOX + FOLFIRI                                         4226     0   \n",
       "\n",
       "Dataset                                              Lin   Peng  Peng_Normal  \\\n",
       "Treatment                                                                      \n",
       "Abraxane                                               0      0            0   \n",
       "FOLFIRINOX                                             0      0            0   \n",
       "FOLFIRINOX + Gem/Abraxane                              0      0            0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...     0      0            0   \n",
       "FOLFIRINOX + RT                                        0      0            0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU          0      0            0   \n",
       "FOLFIRINOX + SBRT                                      0      0            0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab               0      0            0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                          0      0            0   \n",
       "FOLFIRONO + Gemcitabine                                0      0            0   \n",
       "FOLIRINOX + Gem/Abraxane                               0      0            0   \n",
       "Gem/Abraxane                                           0      0            0   \n",
       "Other                                                  0      0            0   \n",
       "Treatment Naive                                     8019  41411            0   \n",
       "Unknown                                                0      0        14668   \n",
       "mFOLFIRINOX + FOLFIRI                                  0      0            0   \n",
       "\n",
       "Dataset                                             Hwang  Schlesinger  \\\n",
       "Treatment                                                                \n",
       "Abraxane                                                0            0   \n",
       "FOLFIRINOX                                              0            0   \n",
       "FOLFIRINOX + Gem/Abraxane                               0            0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...   6957            0   \n",
       "FOLFIRINOX + RT                                         0            0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU       66876            0   \n",
       "FOLFIRINOX + SBRT                                       0            0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab             1323            0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                        4994            0   \n",
       "FOLFIRONO + Gemcitabine                                 0            0   \n",
       "FOLIRINOX + Gem/Abraxane                                0            0   \n",
       "Gem/Abraxane                                            0            0   \n",
       "Other                                               15781            0   \n",
       "Treatment Naive                                     90399            0   \n",
       "Unknown                                                 0         5387   \n",
       "mFOLFIRINOX + FOLFIRI                                   0            0   \n",
       "\n",
       "Dataset                                             Simeone  Steele  \\\n",
       "Treatment                                                             \n",
       "Abraxane                                                  0       0   \n",
       "FOLFIRINOX                                            45734       0   \n",
       "FOLFIRINOX + Gem/Abraxane                                 0       0   \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...        0       0   \n",
       "FOLFIRINOX + RT                                           0       0   \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU             0       0   \n",
       "FOLFIRINOX + SBRT                                         0       0   \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                  0       0   \n",
       "FOLFIRINOX + SBRT + Nivolumab                             0       0   \n",
       "FOLFIRONO + Gemcitabine                                   0       0   \n",
       "FOLIRINOX + Gem/Abraxane                                  0       0   \n",
       "Gem/Abraxane                                          13819       0   \n",
       "Other                                                     0       0   \n",
       "Treatment Naive                                       82435   35928   \n",
       "Unknown                                                   0       0   \n",
       "mFOLFIRINOX + FOLFIRI                                     0       0   \n",
       "\n",
       "Dataset                                             Steele_Adj_Norm  \n",
       "Treatment                                                            \n",
       "Abraxane                                                          0  \n",
       "FOLFIRINOX                                                        0  \n",
       "FOLFIRINOX + Gem/Abraxane                                         0  \n",
       "FOLFIRINOX + Losartan + Radiotherapy + Capecita...                0  \n",
       "FOLFIRINOX + RT                                                   0  \n",
       "FOLFIRINOX + Radiotherapy + Capecitabine/5-FU                     0  \n",
       "FOLFIRINOX + SBRT                                                 0  \n",
       "FOLFIRINOX + SBRT + Losartan + Nivolumab                          0  \n",
       "FOLFIRINOX + SBRT + Nivolumab                                     0  \n",
       "FOLFIRONO + Gemcitabine                                           0  \n",
       "FOLIRINOX + Gem/Abraxane                                          0  \n",
       "Gem/Abraxane                                                      0  \n",
       "Other                                                             0  \n",
       "Treatment Naive                                                   0  \n",
       "Unknown                                                        6331  \n",
       "mFOLFIRINOX + FOLFIRI                                             0  "
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Treatment']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "171f80d3-a60d-4efd-9091-55b7ec882f9e",
   "metadata": {},
   "source": [
    "# Tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "id": "fe9867a9-8333-4c1e-8a40-c44bc274c65d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1132252041.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Tissue']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tissue</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pancreas</th>\n",
       "      <td>71985</td>\n",
       "      <td>144938</td>\n",
       "      <td>0</td>\n",
       "      <td>8136</td>\n",
       "      <td>8019</td>\n",
       "      <td>41411</td>\n",
       "      <td>0</td>\n",
       "      <td>186330</td>\n",
       "      <td>0</td>\n",
       "      <td>141988</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>58888</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14668</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset   Caronni    Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Tissue                                                                      \n",
       "Pancreas    71985  144938               0  8136  8019  41411            0   \n",
       "Unknown         0    2045           58888     0     0      0        14668   \n",
       "\n",
       "Dataset    Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Tissue                                                           \n",
       "Pancreas  186330            0   141988       0                0  \n",
       "Unknown        0         5387        0   35928             6331  "
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Tissue']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "id": "27f7d705-5e2c-40ec-a1dc-58415b303f85",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Tissue = np.where(biological_metadata.Dataset.str.contains('Ding'), 'Pancreas', biological_metadata.Tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "id": "20ae3bab-35e1-49b3-bd4b-899133d34ef3",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Tissue = np.where(biological_metadata.Dataset.str.contains('Peng'), 'Pancreas', biological_metadata.Tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "id": "ef623f2a-8c70-4251-992f-ae202f8897b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Tissue = np.where(biological_metadata.Dataset.str.contains('Schlesinger'), 'Pancreas', biological_metadata.Tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "id": "2e31c77f-dc04-48c7-aea1-3fd0b5acd335",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Tissue = np.where(biological_metadata.Dataset.str.contains('Steele'), 'Pancreas', biological_metadata.Tissue)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "id": "5d80d313-b371-46d2-9b6a-34a355a5899e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1132252041.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Tissue']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tissue</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Pancreas</th>\n",
       "      <td>71985</td>\n",
       "      <td>146983</td>\n",
       "      <td>58888</td>\n",
       "      <td>8136</td>\n",
       "      <td>8019</td>\n",
       "      <td>41411</td>\n",
       "      <td>14668</td>\n",
       "      <td>186330</td>\n",
       "      <td>5387</td>\n",
       "      <td>141988</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset   Caronni    Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Tissue                                                                      \n",
       "Pancreas    71985  146983           58888  8136  8019  41411        14668   \n",
       "\n",
       "Dataset    Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Tissue                                                           \n",
       "Pancreas  186330         5387   141988   35928             6331  "
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Tissue']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "810b2c6a-cd8c-4e4e-8b26-3f2833042732",
   "metadata": {},
   "source": [
    "# Age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "id": "d8babf25-6949-48b9-b3fa-34947fab5b7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1424734676.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Age']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20s</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5256</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30s</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2301</td>\n",
       "      <td>4294</td>\n",
       "      <td>4553</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40s</th>\n",
       "      <td>0</td>\n",
       "      <td>7375</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>752</td>\n",
       "      <td>1787</td>\n",
       "      <td>2270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4687</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50s</th>\n",
       "      <td>4093</td>\n",
       "      <td>40171</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2374</td>\n",
       "      <td>22201</td>\n",
       "      <td>3895</td>\n",
       "      <td>15567</td>\n",
       "      <td>0</td>\n",
       "      <td>10150</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60s</th>\n",
       "      <td>35019</td>\n",
       "      <td>57655</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1714</td>\n",
       "      <td>11299</td>\n",
       "      <td>4209</td>\n",
       "      <td>62724</td>\n",
       "      <td>0</td>\n",
       "      <td>81404</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70s</th>\n",
       "      <td>32873</td>\n",
       "      <td>36499</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2345</td>\n",
       "      <td>3823</td>\n",
       "      <td>0</td>\n",
       "      <td>77037</td>\n",
       "      <td>0</td>\n",
       "      <td>10992</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80s</th>\n",
       "      <td>0</td>\n",
       "      <td>3238</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>834</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21193</td>\n",
       "      <td>0</td>\n",
       "      <td>34755</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>58888</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Caronni   Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Age                                                                       \n",
       "20s            0      0               0     0     0      0            0   \n",
       "30s            0      0               0     0     0   2301         4294   \n",
       "40s            0   7375               0     0   752   1787         2270   \n",
       "50s         4093  40171               0     0  2374  22201         3895   \n",
       "60s        35019  57655               0     0  1714  11299         4209   \n",
       "70s        32873  36499               0     0  2345   3823            0   \n",
       "80s            0   3238               0     0   834      0            0   \n",
       "Unknown        0   2045           58888  8136     0      0            0   \n",
       "\n",
       "Dataset  Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Age                                                            \n",
       "20s       5256            0        0       0                0  \n",
       "30s       4553            0        0       0                0  \n",
       "40s          0            0     4687       0                0  \n",
       "50s      15567            0    10150       0                0  \n",
       "60s      62724            0    81404   35928             6331  \n",
       "70s      77037            0    10992       0                0  \n",
       "80s      21193            0    34755       0                0  \n",
       "Unknown      0         5387        0       0                0  "
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Age']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "3a473681-0b0c-417a-a938-e2d64d9953cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_age_ids = list(biological_metadata[(biological_metadata.Dataset.str.contains('Ding')) & (biological_metadata.Age == 'Unknown')].ID_harmonised.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "d8fdee70-6c86-4146-a374-b7c8d4e15c89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient_ID</th>\n",
       "      <th>Age</th>\n",
       "      <th>Race</th>\n",
       "      <th>Ethnicity</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Time_from_diagnosis_to_surgery_(weeks)</th>\n",
       "      <th>Treatment</th>\n",
       "      <th>Cycles</th>\n",
       "      <th>Treatment_notes</th>\n",
       "      <th>Cancer_type</th>\n",
       "      <th>Location</th>\n",
       "      <th>Grade</th>\n",
       "      <th>Size_(cm)</th>\n",
       "      <th>Tumor_viability_%</th>\n",
       "      <th>Pathology_notes</th>\n",
       "      <th>Lymph_node_met</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Diabetes</th>\n",
       "      <th>Pancreatitis</th>\n",
       "      <th>Previous_cancer_history</th>\n",
       "      <th>Other_co-morbidities</th>\n",
       "      <th>Smoker</th>\n",
       "      <th>Alcohol</th>\n",
       "      <th>Family_history_HPB_first_degree</th>\n",
       "      <th>Other_family_history_first_degree</th>\n",
       "      <th>Vital_status</th>\n",
       "      <th>Survival_post_surgery_(weeks)</th>\n",
       "      <th>Other_Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HT056P1</td>\n",
       "      <td>48</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>6</td>\n",
       "      <td>Last dose 7 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>60.0</td>\n",
       "      <td>ypT2, focal sarcomatoid transformation</td>\n",
       "      <td>2/21, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HT060P1</td>\n",
       "      <td>61</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>FOLFIRINOX</td>\n",
       "      <td>6</td>\n",
       "      <td>Last dose 4 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head, uncinate_process</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>70.0</td>\n",
       "      <td>ypT2, extends to duodenum and vascular groove</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>IIb</td>\n",
       "      <td>Type II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>breast cancer, diabetes-sister</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HT061P1</td>\n",
       "      <td>74</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>male</td>\n",
       "      <td>12</td>\n",
       "      <td>Naive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2, poorly differentiated tumor arising in IPMN</td>\n",
       "      <td>3/7, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Multiple myeloma (remission)</td>\n",
       "      <td>A-fib, chronic kidney disease</td>\n",
       "      <td>Former</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HT064P1</td>\n",
       "      <td>66</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>male</td>\n",
       "      <td>16</td>\n",
       "      <td>Naive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pT2, invasive carcinoma intermixed with IPMN</td>\n",
       "      <td>19/32, pN2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>breast cancer-sister, colon cancer-mother, lun...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HT071P1</td>\n",
       "      <td>68</td>\n",
       "      <td>white</td>\n",
       "      <td>non-hispanic</td>\n",
       "      <td>female</td>\n",
       "      <td>16</td>\n",
       "      <td>Chemo-RT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5-FU and radiation 11 weeks prior to surgery</td>\n",
       "      <td>PDAC</td>\n",
       "      <td>head</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.4</td>\n",
       "      <td>70.0</td>\n",
       "      <td>ypT2, poorly differentiated</td>\n",
       "      <td>2/25, pN1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Breast cancer (chemo 13 years prior)</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Occasional</td>\n",
       "      <td>No</td>\n",
       "      <td>Hodgkin's lymphoma-fater, neuroblastoma-brother</td>\n",
       "      <td>Deceased</td>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient_ID  Age   Race     Ethnicity     Sex  \\\n",
       "0    HT056P1   48  white  non-hispanic  female   \n",
       "1    HT060P1   61  white  non-hispanic  female   \n",
       "2    HT061P1   74  white  non-hispanic    male   \n",
       "3    HT064P1   66  white  non-hispanic    male   \n",
       "4    HT071P1   68  white  non-hispanic  female   \n",
       "\n",
       "   Time_from_diagnosis_to_surgery_(weeks)   Treatment Cycles  \\\n",
       "0                                      20  FOLFIRINOX      6   \n",
       "1                                      20  FOLFIRINOX      6   \n",
       "2                                      12       Naive    NaN   \n",
       "3                                      16       Naive    NaN   \n",
       "4                                      16    Chemo-RT    NaN   \n",
       "\n",
       "                                Treatment_notes Cancer_type  \\\n",
       "0            Last dose 7 weeks prior to surgery        PDAC   \n",
       "1            Last dose 4 weeks prior to surgery        PDAC   \n",
       "2                                           NaN        PDAC   \n",
       "3                                           NaN        PDAC   \n",
       "4  5-FU and radiation 11 weeks prior to surgery        PDAC   \n",
       "\n",
       "                 Location  Grade  Size_(cm)  Tumor_viability_%  \\\n",
       "0  head, uncinate_process    3.0        3.2               60.0   \n",
       "1  head, uncinate_process    3.0        3.2               70.0   \n",
       "2                    head    3.0        3.5                NaN   \n",
       "3                    head    1.0        3.2                NaN   \n",
       "4                    head    3.0        3.4               70.0   \n",
       "\n",
       "                                    Pathology_notes Lymph_node_met Stage  \\\n",
       "0            ypT2, focal sarcomatoid transformation      2/21, pN1   NaN   \n",
       "1     ypT2, extends to duodenum and vascular groove      2/25, pN1   IIb   \n",
       "2  pT2, poorly differentiated tumor arising in IPMN       3/7, pN1   NaN   \n",
       "3      pT2, invasive carcinoma intermixed with IPMN     19/32, pN2   NaN   \n",
       "4                       ypT2, poorly differentiated      2/25, pN1   NaN   \n",
       "\n",
       "  Diabetes Pancreatitis               Previous_cancer_history  \\\n",
       "0       No          Yes                                    No   \n",
       "1  Type II          NaN                                   NaN   \n",
       "2       No          NaN          Multiple myeloma (remission)   \n",
       "3       No          Yes                                    No   \n",
       "4       No          NaN  Breast cancer (chemo 13 years prior)   \n",
       "\n",
       "            Other_co-morbidities  Smoker     Alcohol  \\\n",
       "0                             No      No  Occasional   \n",
       "1                             No      No  Occasional   \n",
       "2  A-fib, chronic kidney disease  Former          No   \n",
       "3                             No      No          No   \n",
       "4                             No      No  Occasional   \n",
       "\n",
       "  Family_history_HPB_first_degree  \\\n",
       "0                              No   \n",
       "1                              No   \n",
       "2                              No   \n",
       "3                              No   \n",
       "4                              No   \n",
       "\n",
       "                   Other_family_history_first_degree Vital_status  \\\n",
       "0                                                 No          NaN   \n",
       "1                     breast cancer, diabetes-sister          NaN   \n",
       "2                                                NaN          NaN   \n",
       "3  breast cancer-sister, colon cancer-mother, lun...          NaN   \n",
       "4    Hodgkin's lymphoma-fater, neuroblastoma-brother     Deceased   \n",
       "\n",
       "   Survival_post_surgery_(weeks) Other_Notes  \n",
       "0                            NaN         NaN  \n",
       "1                            NaN         NaN  \n",
       "2                            NaN         NaN  \n",
       "3                            NaN         NaN  \n",
       "4                            5.0         NaN  "
      ]
     },
     "execution_count": 346,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "8cbaee71-f315-429c-b986-20b6f9956a5a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "ding_dict_age = dict(zip(ding.Patient_ID, ding.Age.apply(lambda x: f\"{int((x // 10) * 10)}s\").tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "id": "d0f998b3-1e1f-4d2d-8896-2c02c12376d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Age = biological_metadata.ID_harmonised.map(ding_dict_age).fillna(biological_metadata.Age)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "id": "a6893652-1dee-47bf-a49b-87516cbb7569",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1424734676.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Age']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Age</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20s</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5256</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30s</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2301</td>\n",
       "      <td>4294</td>\n",
       "      <td>4553</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40s</th>\n",
       "      <td>0</td>\n",
       "      <td>7375</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>752</td>\n",
       "      <td>1787</td>\n",
       "      <td>2270</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4687</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50s</th>\n",
       "      <td>4093</td>\n",
       "      <td>40171</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2374</td>\n",
       "      <td>22201</td>\n",
       "      <td>3895</td>\n",
       "      <td>15567</td>\n",
       "      <td>0</td>\n",
       "      <td>10150</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60s</th>\n",
       "      <td>35019</td>\n",
       "      <td>57655</td>\n",
       "      <td>21865</td>\n",
       "      <td>0</td>\n",
       "      <td>1714</td>\n",
       "      <td>11299</td>\n",
       "      <td>4209</td>\n",
       "      <td>62724</td>\n",
       "      <td>0</td>\n",
       "      <td>81404</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70s</th>\n",
       "      <td>32873</td>\n",
       "      <td>36499</td>\n",
       "      <td>37023</td>\n",
       "      <td>0</td>\n",
       "      <td>2345</td>\n",
       "      <td>3823</td>\n",
       "      <td>0</td>\n",
       "      <td>77037</td>\n",
       "      <td>0</td>\n",
       "      <td>10992</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80s</th>\n",
       "      <td>0</td>\n",
       "      <td>3238</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>834</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>21193</td>\n",
       "      <td>0</td>\n",
       "      <td>34755</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>0</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Caronni   Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Age                                                                       \n",
       "20s            0      0               0     0     0      0            0   \n",
       "30s            0      0               0     0     0   2301         4294   \n",
       "40s            0   7375               0     0   752   1787         2270   \n",
       "50s         4093  40171               0     0  2374  22201         3895   \n",
       "60s        35019  57655           21865     0  1714  11299         4209   \n",
       "70s        32873  36499           37023     0  2345   3823            0   \n",
       "80s            0   3238               0     0   834      0            0   \n",
       "Unknown        0   2045               0  8136     0      0            0   \n",
       "\n",
       "Dataset  Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Age                                                            \n",
       "20s       5256            0        0       0                0  \n",
       "30s       4553            0        0       0                0  \n",
       "40s          0            0     4687       0                0  \n",
       "50s      15567            0    10150       0                0  \n",
       "60s      62724            0    81404   35928             6331  \n",
       "70s      77037            0    10992       0                0  \n",
       "80s      21193            0    34755       0                0  \n",
       "Unknown      0         5387        0       0                0  "
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Age']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4968725-947c-4512-b780-ecd98f2186c6",
   "metadata": {},
   "source": [
    "# Sex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "id": "6e834088-d4e4-4ee2-be4f-5f1e99c84be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "ding_dict_sex = dict(zip(ding.Patient_ID, ding.Sex.apply(lambda x: x.capitalize())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "id": "e737fb9b-5dea-446d-a8dd-dbb9f7c5ab6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'HT056P1': 'Female',\n",
       " 'HT060P1': 'Female',\n",
       " 'HT061P1': 'Male',\n",
       " 'HT064P1': 'Male',\n",
       " 'HT071P1': 'Female',\n",
       " 'HT085P1': 'Female',\n",
       " 'HT115P1': 'Male',\n",
       " 'HT121P1': 'Male',\n",
       " 'HT122P1': 'Male',\n",
       " 'HT123P1': 'Female',\n",
       " 'HT124P1': 'Male',\n",
       " 'HT125P1': 'Female',\n",
       " 'HT138P1': 'Male',\n",
       " 'HT140P1': 'Male',\n",
       " 'HT166P1': 'Female',\n",
       " 'HT168P1': 'Male',\n",
       " 'HT185P1': 'Male',\n",
       " 'HT190P1': 'Female',\n",
       " 'HT191P1': 'Male',\n",
       " 'HT200P1': 'Female',\n",
       " 'HT204P1': 'Male',\n",
       " 'HT264P1': 'Female',\n",
       " 'HT270P1': 'Male',\n",
       " 'HT284P1': 'Male',\n",
       " 'HT288P1': 'Male',\n",
       " 'HT224P1': 'Female',\n",
       " 'HT231P1': 'Female',\n",
       " 'HT232P1': 'Female',\n",
       " 'HT259P1': 'Female',\n",
       " 'HT306P1': 'Female',\n",
       " 'HT242P1': 'Male',\n",
       " 'HT412P1': 'Male'}"
      ]
     },
     "execution_count": 365,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding_dict_sex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 366,
   "id": "356e0462-2ebf-436d-bf47-25e0b208682c",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Sex = biological_metadata.ID_harmonised.map(ding_dict_sex).fillna(biological_metadata.Sex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "id": "646ec2f8-337c-4401-b08a-a18e664a21d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2182428834.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Sex']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sex</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Female</th>\n",
       "      <td>30739</td>\n",
       "      <td>68268</td>\n",
       "      <td>24763</td>\n",
       "      <td>0</td>\n",
       "      <td>3862</td>\n",
       "      <td>22009</td>\n",
       "      <td>0</td>\n",
       "      <td>79903</td>\n",
       "      <td>0</td>\n",
       "      <td>88966</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Male</th>\n",
       "      <td>41246</td>\n",
       "      <td>76670</td>\n",
       "      <td>34125</td>\n",
       "      <td>0</td>\n",
       "      <td>4157</td>\n",
       "      <td>19402</td>\n",
       "      <td>0</td>\n",
       "      <td>106427</td>\n",
       "      <td>0</td>\n",
       "      <td>53022</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>0</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14668</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Caronni   Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Sex                                                                       \n",
       "Female     30739  68268           24763     0  3862  22009            0   \n",
       "Male       41246  76670           34125     0  4157  19402            0   \n",
       "Unknown        0   2045               0  8136     0      0        14668   \n",
       "\n",
       "Dataset   Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Sex                                                             \n",
       "Female    79903            0    88966       0                0  \n",
       "Male     106427            0    53022       0                0  \n",
       "Unknown       0         5387        0   35928             6331  "
      ]
     },
     "execution_count": 367,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Sex']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "id": "344fe3dd-1639-44f3-ac77-e77ba18bb26e",
   "metadata": {},
   "outputs": [],
   "source": [
    "peng_sex_dict = {'Peng_N1':'Female',\n",
    "'Peng_N2':'Male',\n",
    "'Peng_N3':'Male',\n",
    "'Peng_N4':'Male',\n",
    "'Peng_N5':'Female',\n",
    "'Peng_N6':'Female',\n",
    "'Peng_N7':'Female',\n",
    "'Peng_N8':'Male',\n",
    "'Peng_N9':'Male',\n",
    "'Peng_N10':'Female',\n",
    "'Peng_N11':'Female'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 372,
   "id": "16f63730-7325-4a53-b1fa-ca204db5c7ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Sex = biological_metadata.ID_harmonised.map(peng_sex_dict).fillna(biological_metadata.Sex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "id": "2a835ca3-0d42-4a82-8eb2-e24cc633db3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2182428834.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Sex']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sex</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Female</th>\n",
       "      <td>30739</td>\n",
       "      <td>68268</td>\n",
       "      <td>24763</td>\n",
       "      <td>0</td>\n",
       "      <td>3862</td>\n",
       "      <td>22009</td>\n",
       "      <td>8271</td>\n",
       "      <td>79903</td>\n",
       "      <td>0</td>\n",
       "      <td>88966</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Male</th>\n",
       "      <td>41246</td>\n",
       "      <td>76670</td>\n",
       "      <td>34125</td>\n",
       "      <td>0</td>\n",
       "      <td>4157</td>\n",
       "      <td>19402</td>\n",
       "      <td>6397</td>\n",
       "      <td>106427</td>\n",
       "      <td>0</td>\n",
       "      <td>53022</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>0</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Caronni   Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Sex                                                                       \n",
       "Female     30739  68268           24763     0  3862  22009         8271   \n",
       "Male       41246  76670           34125     0  4157  19402         6397   \n",
       "Unknown        0   2045               0  8136     0      0            0   \n",
       "\n",
       "Dataset   Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Sex                                                             \n",
       "Female    79903            0    88966       0                0  \n",
       "Male     106427            0    53022       0                0  \n",
       "Unknown       0         5387        0   35928             6331  "
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Sex']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2160df25-bcd1-4d8e-9168-89dc9d25ad46",
   "metadata": {},
   "source": [
    "# TNM Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 380,
   "id": "bba867a3-24aa-42f1-8954-5d354ac61380",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ypT2 pN1', 'pT2 pN1', 'pT2 pN2', 'ypT2 pN0', 'ypT2 pN2',\n",
       "       'pT3 pN2', 'ypT3 pN2', nan, 'yT2 pN1', 'T4 pN2 M0', 'T1c pN1 M0',\n",
       "       'T2 pN0 M0', 'T3 pN0 M0', 'T3 pN1 M0', 'T1c pN2 M0', 'T2 pN1 M0',\n",
       "       'T3 pN2 M0', 'T1c pN0 M0', 'pT1c pN2', 'T2 pN2', 'T2 pN2 M0',\n",
       "       'T2 pN1', 'ypT2 pN1 M0', 'ypT3 pN0 M0', 'ypT1c pN0 M0',\n",
       "       'ypT2 pN0 M0', 'ypT1a pN1 M0', 'ypT2 pN2 M0', 'ypT1a pN0 M0',\n",
       "       'ypT1c pN1 M0', 'ypT3 pN1 M0', 'ypT3 pN0', 'rpT2 pN1', 'pT2 pN0',\n",
       "       'pT1a pN0', 'ypT1 pN0'], dtype=object)"
      ]
     },
     "execution_count": 380,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.Harmonized_TNM_Classification.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 406,
   "id": "aace9c93-7878-4950-8026-96d5e4b5ade2",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Harmonized_TNM_Classification = biological_metadata.Harmonized_TNM_Classification.replace('NaN', 'Unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 407,
   "id": "a358337f-bff2-4640-98aa-a02dce7392c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2873525045.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Harmonized_TNM_Classification']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Harmonized_TNM_Classification</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>T1c pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1787</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1c pN1 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4060</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1c pN2 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>660</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2 pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14138</td>\n",
       "      <td>0</td>\n",
       "      <td>8197</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2 pN1</th>\n",
       "      <td>12343</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2 pN1 M0</th>\n",
       "      <td>14240</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11170</td>\n",
       "      <td>0</td>\n",
       "      <td>31710</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2 pN2</th>\n",
       "      <td>8154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T2 pN2 M0</th>\n",
       "      <td>34194</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6023</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3 pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1868</td>\n",
       "      <td>1405</td>\n",
       "      <td>15600</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3 pN1 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4340</td>\n",
       "      <td>0</td>\n",
       "      <td>12268</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T3 pN2 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2231</td>\n",
       "      <td>0</td>\n",
       "      <td>16601</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T4 pN2 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1157</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>2045</td>\n",
       "      <td>0</td>\n",
       "      <td>8136</td>\n",
       "      <td>8019</td>\n",
       "      <td>0</td>\n",
       "      <td>13263</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>141988</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT1a pN0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT1c pN2</th>\n",
       "      <td>3054</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11810</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN1</th>\n",
       "      <td>0</td>\n",
       "      <td>6695</td>\n",
       "      <td>6794</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT2 pN2</th>\n",
       "      <td>0</td>\n",
       "      <td>20297</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pT3 pN2</th>\n",
       "      <td>0</td>\n",
       "      <td>9068</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rpT2 pN1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yT2 pN1</th>\n",
       "      <td>0</td>\n",
       "      <td>2383</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1 pN0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8427</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1a pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5714</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1a pN1 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4994</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1c pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>11141</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT1c pN1 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3264</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN0</th>\n",
       "      <td>0</td>\n",
       "      <td>28991</td>\n",
       "      <td>5876</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>15063</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN1</th>\n",
       "      <td>0</td>\n",
       "      <td>29762</td>\n",
       "      <td>4472</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN1 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN2</th>\n",
       "      <td>0</td>\n",
       "      <td>31465</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT2 pN2 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5800</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8398</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN0 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36976</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN1 M0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5820</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ypT3 pN2</th>\n",
       "      <td>0</td>\n",
       "      <td>16277</td>\n",
       "      <td>2201</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                        Caronni   Ding  Ding_snRNA-seq   Lee   Lin  \\\n",
       "Harmonized_TNM_Classification                                               \n",
       "T1c pN0 M0                           0      0               0     0     0   \n",
       "T1c pN1 M0                           0      0               0     0     0   \n",
       "T1c pN2 M0                           0      0               0     0     0   \n",
       "T2 pN0 M0                            0      0               0     0     0   \n",
       "T2 pN1                           12343      0               0     0     0   \n",
       "T2 pN1 M0                        14240      0               0     0     0   \n",
       "T2 pN2                            8154      0               0     0     0   \n",
       "T2 pN2 M0                        34194      0               0     0     0   \n",
       "T3 pN0 M0                            0      0               0     0     0   \n",
       "T3 pN1 M0                            0      0               0     0     0   \n",
       "T3 pN2 M0                            0      0               0     0     0   \n",
       "T4 pN2 M0                            0      0               0     0     0   \n",
       "Unknown                              0   2045               0  8136  8019   \n",
       "pT1a pN0                             0      0            5084     0     0   \n",
       "pT1c pN2                          3054      0               0     0     0   \n",
       "pT2 pN0                              0      0           11810     0     0   \n",
       "pT2 pN1                              0   6695            6794     0     0   \n",
       "pT2 pN2                              0  20297               0     0     0   \n",
       "pT3 pN2                              0   9068               0     0     0   \n",
       "rpT2 pN1                             0      0            5826     0     0   \n",
       "yT2 pN1                              0   2383               0     0     0   \n",
       "ypT1 pN0                             0      0            8427     0     0   \n",
       "ypT1a pN0 M0                         0      0               0     0     0   \n",
       "ypT1a pN1 M0                         0      0               0     0     0   \n",
       "ypT1c pN0 M0                         0      0               0     0     0   \n",
       "ypT1c pN1 M0                         0      0               0     0     0   \n",
       "ypT2 pN0                             0  28991            5876     0     0   \n",
       "ypT2 pN0 M0                          0      0               0     0     0   \n",
       "ypT2 pN1                             0  29762            4472     0     0   \n",
       "ypT2 pN1 M0                          0      0               0     0     0   \n",
       "ypT2 pN2                             0  31465               0     0     0   \n",
       "ypT2 pN2 M0                          0      0               0     0     0   \n",
       "ypT3 pN0                             0      0            8398     0     0   \n",
       "ypT3 pN0 M0                          0      0               0     0     0   \n",
       "ypT3 pN1 M0                          0      0               0     0     0   \n",
       "ypT3 pN2                             0  16277            2201     0     0   \n",
       "\n",
       "Dataset                         Peng  Peng_Normal  Hwang  Schlesinger  \\\n",
       "Harmonized_TNM_Classification                                           \n",
       "T1c pN0 M0                      1787            0      0            0   \n",
       "T1c pN1 M0                      4060            0      0            0   \n",
       "T1c pN2 M0                       660            0      0            0   \n",
       "T2 pN0 M0                      14138            0   8197            0   \n",
       "T2 pN1                             0            0      0            0   \n",
       "T2 pN1 M0                      11170            0  31710            0   \n",
       "T2 pN2                             0            0      0            0   \n",
       "T2 pN2 M0                          0            0   6023            0   \n",
       "T3 pN0 M0                       1868         1405  15600            0   \n",
       "T3 pN1 M0                       4340            0  12268            0   \n",
       "T3 pN2 M0                       2231            0  16601            0   \n",
       "T4 pN2 M0                       1157            0      0            0   \n",
       "Unknown                            0        13263      0         5387   \n",
       "pT1a pN0                           0            0      0            0   \n",
       "pT1c pN2                           0            0      0            0   \n",
       "pT2 pN0                            0            0      0            0   \n",
       "pT2 pN1                            0            0      0            0   \n",
       "pT2 pN2                            0            0      0            0   \n",
       "pT3 pN2                            0            0      0            0   \n",
       "rpT2 pN1                           0            0      0            0   \n",
       "yT2 pN1                            0            0      0            0   \n",
       "ypT1 pN0                           0            0      0            0   \n",
       "ypT1a pN0 M0                       0            0   5714            0   \n",
       "ypT1a pN1 M0                       0            0   4994            0   \n",
       "ypT1c pN0 M0                       0            0  11141            0   \n",
       "ypT1c pN1 M0                       0            0   3264            0   \n",
       "ypT2 pN0                           0            0      0            0   \n",
       "ypT2 pN0 M0                        0            0  15063            0   \n",
       "ypT2 pN1                           0            0      0            0   \n",
       "ypT2 pN1 M0                        0            0   7159            0   \n",
       "ypT2 pN2                           0            0      0            0   \n",
       "ypT2 pN2 M0                        0            0   5800            0   \n",
       "ypT3 pN0                           0            0      0            0   \n",
       "ypT3 pN0 M0                        0            0  36976            0   \n",
       "ypT3 pN1 M0                        0            0   5820            0   \n",
       "ypT3 pN2                           0            0      0            0   \n",
       "\n",
       "Dataset                        Simeone  Steele  Steele_Adj_Norm  \n",
       "Harmonized_TNM_Classification                                    \n",
       "T1c pN0 M0                           0       0                0  \n",
       "T1c pN1 M0                           0       0                0  \n",
       "T1c pN2 M0                           0       0                0  \n",
       "T2 pN0 M0                            0       0                0  \n",
       "T2 pN1                               0       0                0  \n",
       "T2 pN1 M0                            0       0                0  \n",
       "T2 pN2                               0       0                0  \n",
       "T2 pN2 M0                            0       0                0  \n",
       "T3 pN0 M0                            0       0                0  \n",
       "T3 pN1 M0                            0       0                0  \n",
       "T3 pN2 M0                            0       0                0  \n",
       "T4 pN2 M0                            0       0                0  \n",
       "Unknown                         141988   35928             6331  \n",
       "pT1a pN0                             0       0                0  \n",
       "pT1c pN2                             0       0                0  \n",
       "pT2 pN0                              0       0                0  \n",
       "pT2 pN1                              0       0                0  \n",
       "pT2 pN2                              0       0                0  \n",
       "pT3 pN2                              0       0                0  \n",
       "rpT2 pN1                             0       0                0  \n",
       "yT2 pN1                              0       0                0  \n",
       "ypT1 pN0                             0       0                0  \n",
       "ypT1a pN0 M0                         0       0                0  \n",
       "ypT1a pN1 M0                         0       0                0  \n",
       "ypT1c pN0 M0                         0       0                0  \n",
       "ypT1c pN1 M0                         0       0                0  \n",
       "ypT2 pN0                             0       0                0  \n",
       "ypT2 pN0 M0                          0       0                0  \n",
       "ypT2 pN1                             0       0                0  \n",
       "ypT2 pN1 M0                          0       0                0  \n",
       "ypT2 pN2                             0       0                0  \n",
       "ypT2 pN2 M0                          0       0                0  \n",
       "ypT3 pN0                             0       0                0  \n",
       "ypT3 pN0 M0                          0       0                0  \n",
       "ypT3 pN1 M0                          0       0                0  \n",
       "ypT3 pN2                             0       0                0  "
      ]
     },
     "execution_count": 407,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Harmonized_TNM_Classification']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f640fc4-549b-458a-9dcd-082db66b37f0",
   "metadata": {},
   "source": [
    "# Grade"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "id": "76ff4327-eb60-4719-82c3-ba966dacc549",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['3.0', '1.0', '2.0', nan, 'G3', 'G2', '2', '4', '3', 'g3', 'g2',\n",
       "       'g2-3', 'gX'], dtype=object)"
      ]
     },
     "execution_count": 384,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.Grade.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "bb6f7590-6168-48f5-9eeb-a6e65b35daa4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2892831985.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Grade']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Grade</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1.0</th>\n",
       "      <td>0</td>\n",
       "      <td>31490</td>\n",
       "      <td>8900</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5299</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2.0</th>\n",
       "      <td>0</td>\n",
       "      <td>62279</td>\n",
       "      <td>22464</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1479</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3.0</th>\n",
       "      <td>0</td>\n",
       "      <td>43478</td>\n",
       "      <td>19097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1241</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G2</th>\n",
       "      <td>35461</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G3</th>\n",
       "      <td>36524</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103177</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g2-3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19056</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32243</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31854</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Caronni   Ding  Ding_snRNA-seq  Lee   Lin  Peng  Peng_Normal   Hwang  \\\n",
       "Grade                                                                           \n",
       "1.0            0  31490            8900    0     0     0            0       0   \n",
       "2              0      0               0    0  5299     0            0       0   \n",
       "2.0            0  62279           22464    0     0     0            0       0   \n",
       "3              0      0               0    0  1479     0            0       0   \n",
       "3.0            0  43478           19097    0     0     0            0       0   \n",
       "4              0      0               0    0  1241     0            0       0   \n",
       "G2         35461      0               0    0     0     0            0       0   \n",
       "G3         36524      0               0    0     0     0            0       0   \n",
       "g2             0      0               0    0     0     0            0  103177   \n",
       "g2-3           0      0               0    0     0     0            0   19056   \n",
       "g3             0      0               0    0     0     0            0   32243   \n",
       "gX             0      0               0    0     0     0            0   31854   \n",
       "\n",
       "Dataset  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Grade                                                   \n",
       "1.0                0        0       0                0  \n",
       "2                  0        0       0                0  \n",
       "2.0                0        0       0                0  \n",
       "3                  0        0       0                0  \n",
       "3.0                0        0       0                0  \n",
       "4                  0        0       0                0  \n",
       "G2                 0        0       0                0  \n",
       "G3                 0        0       0                0  \n",
       "g2                 0        0       0                0  \n",
       "g2-3               0        0       0                0  \n",
       "g3                 0        0       0                0  \n",
       "gX                 0        0       0                0  "
      ]
     },
     "execution_count": 383,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Grade']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "id": "b51cba35-1f10-457d-ae93-6fec24dfd0ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/229157401.py:20: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Harmonized_Grade']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Harmonized_Grade</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G1</th>\n",
       "      <td>0</td>\n",
       "      <td>31490</td>\n",
       "      <td>8900</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G2</th>\n",
       "      <td>35461</td>\n",
       "      <td>62279</td>\n",
       "      <td>22464</td>\n",
       "      <td>0</td>\n",
       "      <td>5299</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>103177</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G2-G3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19056</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G3</th>\n",
       "      <td>36524</td>\n",
       "      <td>43478</td>\n",
       "      <td>19097</td>\n",
       "      <td>0</td>\n",
       "      <td>2720</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>32243</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GX</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31854</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>0</td>\n",
       "      <td>9736</td>\n",
       "      <td>8427</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>41411</td>\n",
       "      <td>14668</td>\n",
       "      <td>0</td>\n",
       "      <td>5387</td>\n",
       "      <td>141988</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset           Caronni   Ding  Ding_snRNA-seq   Lee   Lin   Peng  \\\n",
       "Harmonized_Grade                                                      \n",
       "G1                      0  31490            8900     0     0      0   \n",
       "G2                  35461  62279           22464     0  5299      0   \n",
       "G2-G3                   0      0               0     0     0      0   \n",
       "G3                  36524  43478           19097     0  2720      0   \n",
       "GX                      0      0               0     0     0      0   \n",
       "Unknown                 0   9736            8427  8136     0  41411   \n",
       "\n",
       "Dataset           Peng_Normal   Hwang  Schlesinger  Simeone  Steele  \\\n",
       "Harmonized_Grade                                                      \n",
       "G1                          0       0            0        0       0   \n",
       "G2                          0  103177            0        0       0   \n",
       "G2-G3                       0   19056            0        0       0   \n",
       "G3                          0   32243            0        0       0   \n",
       "GX                          0   31854            0        0       0   \n",
       "Unknown                 14668       0         5387   141988   35928   \n",
       "\n",
       "Dataset           Steele_Adj_Norm  \n",
       "Harmonized_Grade                   \n",
       "G1                              0  \n",
       "G2                              0  \n",
       "G2-G3                           0  \n",
       "G3                              0  \n",
       "GX                              0  \n",
       "Unknown                      6331  "
      ]
     },
     "execution_count": 390,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grade_mapping = {\n",
    "    '1.0': 'G1',\n",
    "    '2.0': 'G2',\n",
    "    '3.0': 'G3',\n",
    "    '1': 'G1',\n",
    "    '2': 'G2',\n",
    "    '3': 'G3',\n",
    "    '4': 'G3',  # Assuming grade 4 maps to G3\n",
    "    'G1': 'G1',\n",
    "    'G2': 'G2',\n",
    "    'G3': 'G3',\n",
    "    'g1': 'G1',\n",
    "    'g2': 'G2',\n",
    "    'g3': 'G3',\n",
    "    'g2-3': 'G2-G3',  # Assuming g2-3 means G2 or G3\n",
    "    'gX': 'GX',  # Handle unknown grades\n",
    "    np.nan: 'Unknown'  # Handle missing values\n",
    "}\n",
    "biological_metadata['Harmonized_Grade'] = biological_metadata['Grade'].map(grade_mapping)\n",
    "biological_metadata.groupby(['Dataset', 'Harmonized_Grade']).size().unstack().T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fc4c970-e6d3-48c5-bca0-42613f5d1883",
   "metadata": {},
   "source": [
    "# Stage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 385,
   "id": "b872c813-2dad-48f5-8cfd-9e1a1ed851e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([nan, 'IIb', 'III', 'IIB', 'IB', 'IIA', 'Ia'], dtype=object)"
      ]
     },
     "execution_count": 385,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.Stage.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "id": "3ad38a78-4587-4b70-b548-77c5b2e6adbe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2146864415.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Stage']).size().unstack().T.fillna(0).apply(pd.to_numeric, errors='coerce')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stage</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IB</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3723</td>\n",
       "      <td>4169</td>\n",
       "      <td>1405</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIB</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3708</td>\n",
       "      <td>18743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>III</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4226</td>\n",
       "      <td>0</td>\n",
       "      <td>588</td>\n",
       "      <td>4048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIb</th>\n",
       "      <td>0</td>\n",
       "      <td>8082</td>\n",
       "      <td>10966</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ia</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>71985</td>\n",
       "      <td>138901</td>\n",
       "      <td>38612</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13263</td>\n",
       "      <td>186330</td>\n",
       "      <td>5387</td>\n",
       "      <td>141988</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset  Caronni    Ding  Ding_snRNA-seq   Lee   Lin   Peng  Peng_Normal  \\\n",
       "Stage                                                                      \n",
       "IB             0       0               0     0     0  14451            0   \n",
       "IIA            0       0               0     0  3723   4169         1405   \n",
       "IIB            0       0               0     0  3708  18743            0   \n",
       "III            0       0            4226     0   588   4048            0   \n",
       "IIb            0    8082           10966     0     0      0            0   \n",
       "Ia             0       0            5084     0     0      0            0   \n",
       "Unknown    71985  138901           38612  8136     0      0        13263   \n",
       "\n",
       "Dataset   Hwang  Schlesinger  Simeone  Steele  Steele_Adj_Norm  \n",
       "Stage                                                           \n",
       "IB            0            0        0       0                0  \n",
       "IIA           0            0        0       0                0  \n",
       "IIB           0            0        0       0                0  \n",
       "III           0            0        0       0                0  \n",
       "IIb           0            0        0       0                0  \n",
       "Ia            0            0        0       0                0  \n",
       "Unknown  186330         5387   141988   35928             6331  "
      ]
     },
     "execution_count": 393,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Stage']).size().unstack().T.fillna(0).apply(pd.to_numeric, errors='coerce')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "id": "f42ac84a-6ff5-41b5-82ed-bb730fca7556",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata.Stage = biological_metadata.Stage.replace(np.nan, 'Unknown')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "id": "3e3bdf8d-c493-438f-b8eb-24c30ada0304",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2278554412.py:11: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Harmonized_Stage']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Harmonized_Stage</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IB</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3723</td>\n",
       "      <td>4169</td>\n",
       "      <td>1405</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIB</th>\n",
       "      <td>0</td>\n",
       "      <td>8082</td>\n",
       "      <td>10966</td>\n",
       "      <td>0</td>\n",
       "      <td>3708</td>\n",
       "      <td>18743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>III</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4226</td>\n",
       "      <td>0</td>\n",
       "      <td>588</td>\n",
       "      <td>4048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>71985</td>\n",
       "      <td>138901</td>\n",
       "      <td>38612</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13263</td>\n",
       "      <td>186330</td>\n",
       "      <td>5387</td>\n",
       "      <td>141988</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset           Caronni    Ding  Ding_snRNA-seq   Lee   Lin   Peng  \\\n",
       "Harmonized_Stage                                                       \n",
       "IA                      0       0            5084     0     0      0   \n",
       "IB                      0       0               0     0     0  14451   \n",
       "IIA                     0       0               0     0  3723   4169   \n",
       "IIB                     0    8082           10966     0  3708  18743   \n",
       "III                     0       0            4226     0   588   4048   \n",
       "Unknown             71985  138901           38612  8136     0      0   \n",
       "\n",
       "Dataset           Peng_Normal   Hwang  Schlesinger  Simeone  Steele  \\\n",
       "Harmonized_Stage                                                      \n",
       "IA                          0       0            0        0       0   \n",
       "IB                          0       0            0        0       0   \n",
       "IIA                      1405       0            0        0       0   \n",
       "IIB                         0       0            0        0       0   \n",
       "III                         0       0            0        0       0   \n",
       "Unknown                 13263  186330         5387   141988   35928   \n",
       "\n",
       "Dataset           Steele_Adj_Norm  \n",
       "Harmonized_Stage                   \n",
       "IA                              0  \n",
       "IB                              0  \n",
       "IIA                             0  \n",
       "IIB                             0  \n",
       "III                             0  \n",
       "Unknown                      6331  "
      ]
     },
     "execution_count": 399,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stage_mapping = {\n",
    "    'Ia': 'IA',\n",
    "    'IB': 'IB',\n",
    "    'IIA': 'IIA',\n",
    "    'IIB': 'IIB',\n",
    "    'IIb': 'IIB',\n",
    "    'III': 'III',\n",
    "    np.nan: 'Unknown',  # Handle missing values\n",
    "}\n",
    "biological_metadata['Harmonized_Stage'] = biological_metadata['Stage'].map(stage_mapping).fillna('Unknown')\n",
    "biological_metadata.groupby(['Dataset', 'Harmonized_Stage']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 426,
   "id": "80e57c52-7bbf-49cc-9125-4dbb565fc00a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/3023972922.py:31: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Harmonized_Stage']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Harmonized_Stage</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>IA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IB</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>14451</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>45409</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIA</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3723</td>\n",
       "      <td>4169</td>\n",
       "      <td>1405</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IIB</th>\n",
       "      <td>0</td>\n",
       "      <td>8082</td>\n",
       "      <td>10966</td>\n",
       "      <td>0</td>\n",
       "      <td>3708</td>\n",
       "      <td>18743</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12200</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>III</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4226</td>\n",
       "      <td>0</td>\n",
       "      <td>588</td>\n",
       "      <td>4048</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31142</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IV</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53237</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown</th>\n",
       "      <td>71985</td>\n",
       "      <td>138901</td>\n",
       "      <td>38612</td>\n",
       "      <td>8136</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13263</td>\n",
       "      <td>186330</td>\n",
       "      <td>5387</td>\n",
       "      <td>0</td>\n",
       "      <td>35928</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset           Caronni    Ding  Ding_snRNA-seq   Lee   Lin   Peng  \\\n",
       "Harmonized_Stage                                                       \n",
       "IA                      0       0            5084     0     0      0   \n",
       "IB                      0       0               0     0     0  14451   \n",
       "IIA                     0       0               0     0  3723   4169   \n",
       "IIB                     0    8082           10966     0  3708  18743   \n",
       "III                     0       0            4226     0   588   4048   \n",
       "IV                      0       0               0     0     0      0   \n",
       "Unknown             71985  138901           38612  8136     0      0   \n",
       "\n",
       "Dataset           Peng_Normal   Hwang  Schlesinger  Simeone  Steele  \\\n",
       "Harmonized_Stage                                                      \n",
       "IA                          0       0            0        0       0   \n",
       "IB                          0       0            0    45409       0   \n",
       "IIA                      1405       0            0        0       0   \n",
       "IIB                         0       0            0    12200       0   \n",
       "III                         0       0            0    31142       0   \n",
       "IV                          0       0            0    53237       0   \n",
       "Unknown                 13263  186330         5387        0   35928   \n",
       "\n",
       "Dataset           Steele_Adj_Norm  \n",
       "Harmonized_Stage                   \n",
       "IA                              0  \n",
       "IB                              0  \n",
       "IIA                             0  \n",
       "IIB                             0  \n",
       "III                             0  \n",
       "IV                              0  \n",
       "Unknown                      6331  "
      ]
     },
     "execution_count": 426,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "simeone_stage = {'P01': 'IV',\n",
    "'P02': 'IV',\n",
    "'P03': 'IV',\n",
    "'P04': 'IB',\n",
    "'P05': 'III',\n",
    "'P06': 'III',\n",
    "'P07': 'IB',\n",
    "'P08': 'IIB',\n",
    "'P09': 'IV',\n",
    "'P10': 'IIB',\n",
    "'P11': 'IV',\n",
    "'P12': 'IV',\n",
    "'P13': 'III',\n",
    "'P14': 'IB',\n",
    "'P15': 'III',\n",
    "'P16': 'IV',\n",
    "'P17': 'IV',\n",
    "'P18': 'IV',\n",
    "'P19': 'IB',\n",
    "'P20': 'IV',\n",
    "'P21': 'IV',\n",
    "'P22': 'III',\n",
    "'P23': 'IB',\n",
    "'P24': 'IV',\n",
    "'P25': 'IV',\n",
    "'P26': 'IV',\n",
    "'P26': 'IV',\n",
    "'P27': 'IV'}\n",
    "\n",
    "biological_metadata['Harmonized_Stage'] = biological_metadata['ID_harmonised'].map(simeone_stage).fillna(biological_metadata.Harmonized_Stage)\n",
    "biological_metadata.groupby(['Dataset', 'Harmonized_Stage']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 427,
   "id": "c92fc3a7-f841-415a-804a-ac5d69ae70a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/4019530717.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  df = biological_metadata.groupby('Dataset').nunique() #.apply(lambda x: x.notna().nunique())\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID_batch_covariate: 12/12\n",
      "ID_harmonised: 12/12\n",
      "Unique_ID: 12/12\n",
      "Level_1_refined: 12/12\n",
      "Tissue: 12/12\n",
      "Age: 12/12\n",
      "Sex: 12/12\n",
      "Diabetes: 12/12\n",
      "Treatment: 12/12\n",
      "Stage: 12/12\n",
      "Harmonized_Grade: 12/12\n",
      "Harmonized_TNM_Classification: 12/12\n",
      "Harmonized_Stage: 12/12\n",
      "TNM_classification: 6/12\n",
      "Location: 5/12\n",
      "Grade: 5/12\n",
      "Pathology_notes: 4/12\n",
      "Vital_status: 3/12\n",
      "Overall Survival: 3/12\n",
      "Procedure: 3/12\n",
      "KRAS_mutation: 2/12\n",
      "TP53_type: 2/12\n",
      "SMAD4_type: 2/12\n",
      "CDKN2A_type: 2/12\n",
      "Time_from_diagnosis_to_surgery_(weeks): 2/12\n",
      "Cycles: 2/12\n",
      "Size_(cm): 2/12\n",
      "Pancreatitis: 2/12\n",
      "Previous_cancer_history: 2/12\n",
      "Other_co-morbidities: 2/12\n",
      "Smoker: 2/12\n",
      "Alcohol: 2/12\n",
      "Lymph_node_met: 2/12\n",
      "Survival_post_surgery_(weeks): 2/12\n",
      "BMI: 2/12\n",
      "perineural invasion: 2/12\n",
      "vascular invasion: 2/12\n",
      "peripancreatic infiltarion: 2/12\n",
      "KRAS_type: 1/12\n",
      "Germline_ATM: 1/12\n",
      "Germline_BRCA2: 1/12\n",
      "Germline_FANCC: 1/12\n",
      "Moffitt_Subtype: 1/12\n",
      "Bailey_Subtype: 1/12\n",
      "Collisson_Subtype: 1/12\n"
     ]
    }
   ],
   "source": [
    "df = biological_metadata.groupby('Dataset').nunique() #.apply(lambda x: x.notna().nunique())\n",
    "counts = {}\n",
    "for col in df:\n",
    "    counts[col] = np.count_nonzero(df[col])  \n",
    "sorted_counts = sorted(counts.items(), key=lambda item: item[1], reverse=True)\n",
    "for col, count in sorted_counts:\n",
    "    print(f'{col}: {count}/{len(df[col].index)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af5f7731-3005-43c5-84ab-1c135d2a1ce4",
   "metadata": {},
   "source": [
    "# Location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "id": "a98d1707-5aec-42db-865c-9bd3a63ee5a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Head, Uncinated Process', 'Head', 'Neck', nan, 'Body',\n",
       "       'Uncinated Process', 'Tail', 'Body/Tail',\n",
       "       'Tail, Proximal Remanant', 'Neck/Body'], dtype=object)"
      ]
     },
     "execution_count": 413,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.Location.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 411,
   "id": "89b2050f-fd04-462a-b6bc-4bc74326a164",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1357878787.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Location']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Location</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Body</th>\n",
       "      <td>4093</td>\n",
       "      <td>7065</td>\n",
       "      <td>3816</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Body/Tail</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4226</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Head</th>\n",
       "      <td>48824</td>\n",
       "      <td>102195</td>\n",
       "      <td>27337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Head, Uncinated Process</th>\n",
       "      <td>10914</td>\n",
       "      <td>14341</td>\n",
       "      <td>8427</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neck</th>\n",
       "      <td>0</td>\n",
       "      <td>10533</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neck/Body</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tail</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4172</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tail, Proximal Remanant</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uncinated Process</th>\n",
       "      <td>8154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                  Caronni    Ding  Ding_snRNA-seq  Lee  Lin  Peng  \\\n",
       "Location                                                                   \n",
       "Body                        4093    7065            3816    0    0     0   \n",
       "Body/Tail                      0       0            4226    0    0     0   \n",
       "Head                       48824  102195           27337    0    0     0   \n",
       "Head, Uncinated Process    10914   14341            8427    0    0     0   \n",
       "Neck                           0   10533               0    0    0     0   \n",
       "Neck/Body                      0       0            5084    0    0     0   \n",
       "Tail                           0       0            4172    0    0     0   \n",
       "Tail, Proximal Remanant        0       0            5826    0    0     0   \n",
       "Uncinated Process           8154       0               0    0    0     0   \n",
       "\n",
       "Dataset                  Peng_Normal  Hwang  Schlesinger  Simeone  Steele  \\\n",
       "Location                                                                    \n",
       "Body                               0      0            0        0       0   \n",
       "Body/Tail                          0      0            0        0       0   \n",
       "Head                               0      0            0        0       0   \n",
       "Head, Uncinated Process            0      0            0        0       0   \n",
       "Neck                               0      0            0        0       0   \n",
       "Neck/Body                          0      0            0        0       0   \n",
       "Tail                               0      0            0        0       0   \n",
       "Tail, Proximal Remanant            0      0            0        0       0   \n",
       "Uncinated Process                  0      0            0        0       0   \n",
       "\n",
       "Dataset                  Steele_Adj_Norm  \n",
       "Location                                  \n",
       "Body                                   0  \n",
       "Body/Tail                              0  \n",
       "Head                                   0  \n",
       "Head, Uncinated Process                0  \n",
       "Neck                                   0  \n",
       "Neck/Body                              0  \n",
       "Tail                                   0  \n",
       "Tail, Proximal Remanant                0  \n",
       "Uncinated Process                      0  "
      ]
     },
     "execution_count": 411,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'Location']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 414,
   "id": "27e672fa-faed-4e9c-9604-256e1d6d889d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/3455355903.py:39: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'Location']).size().unstack().T\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Dataset</th>\n",
       "      <th>Caronni</th>\n",
       "      <th>Ding</th>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <th>Lee</th>\n",
       "      <th>Lin</th>\n",
       "      <th>Peng</th>\n",
       "      <th>Peng_Normal</th>\n",
       "      <th>Hwang</th>\n",
       "      <th>Schlesinger</th>\n",
       "      <th>Simeone</th>\n",
       "      <th>Steele</th>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Location</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Body</th>\n",
       "      <td>4093</td>\n",
       "      <td>7065</td>\n",
       "      <td>3816</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8237</td>\n",
       "      <td>2068</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Body/Tail</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4226</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5180</td>\n",
       "      <td>3016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Common Bile Duct</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1405</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Descending Duodenum</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2039</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Head</th>\n",
       "      <td>48824</td>\n",
       "      <td>102195</td>\n",
       "      <td>27337</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>20017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Head, Uncinated Process</th>\n",
       "      <td>10914</td>\n",
       "      <td>14341</td>\n",
       "      <td>8427</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neck</th>\n",
       "      <td>0</td>\n",
       "      <td>10533</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Neck/Body</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5084</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tail</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>4172</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1868</td>\n",
       "      <td>6140</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tail, Proximal Remanant</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5826</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uncinated Process</th>\n",
       "      <td>8154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6109</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Dataset                  Caronni    Ding  Ding_snRNA-seq  Lee  Lin   Peng  \\\n",
       "Location                                                                    \n",
       "Body                        4093    7065            3816    0    0   8237   \n",
       "Body/Tail                      0       0            4226    0    0   5180   \n",
       "Common Bile Duct               0       0               0    0    0      0   \n",
       "Descending Duodenum            0       0               0    0    0      0   \n",
       "Head                       48824  102195           27337    0    0  20017   \n",
       "Head, Uncinated Process    10914   14341            8427    0    0      0   \n",
       "Neck                           0   10533               0    0    0      0   \n",
       "Neck/Body                      0       0            5084    0    0      0   \n",
       "Tail                           0       0            4172    0    0   1868   \n",
       "Tail, Proximal Remanant        0       0            5826    0    0      0   \n",
       "Uncinated Process           8154       0               0    0    0   6109   \n",
       "\n",
       "Dataset                  Peng_Normal  Hwang  Schlesinger  Simeone  Steele  \\\n",
       "Location                                                                    \n",
       "Body                            2068      0            0        0       0   \n",
       "Body/Tail                       3016      0            0        0       0   \n",
       "Common Bile Duct                1405      0            0        0       0   \n",
       "Descending Duodenum             2039      0            0        0       0   \n",
       "Head                               0      0            0        0       0   \n",
       "Head, Uncinated Process            0      0            0        0       0   \n",
       "Neck                               0      0            0        0       0   \n",
       "Neck/Body                          0      0            0        0       0   \n",
       "Tail                            6140      0            0        0       0   \n",
       "Tail, Proximal Remanant            0      0            0        0       0   \n",
       "Uncinated Process                  0      0            0        0       0   \n",
       "\n",
       "Dataset                  Steele_Adj_Norm  \n",
       "Location                                  \n",
       "Body                                   0  \n",
       "Body/Tail                              0  \n",
       "Common Bile Duct                       0  \n",
       "Descending Duodenum                    0  \n",
       "Head                                   0  \n",
       "Head, Uncinated Process                0  \n",
       "Neck                                   0  \n",
       "Neck/Body                              0  \n",
       "Tail                                   0  \n",
       "Tail, Proximal Remanant                0  \n",
       "Uncinated Process                      0  "
      ]
     },
     "execution_count": 414,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peng_location_mapping = {\n",
    "    'Peng_T1': 'Body',\n",
    "    'Peng_T2': 'Head',\n",
    "    'Peng_T3': 'Uncinated Process',\n",
    "    'Peng_T4': 'Body',\n",
    "    'Peng_T5': 'Uncinated Process',\n",
    "    'Peng_T6': 'Tail',\n",
    "    'Peng_T7': 'Body',\n",
    "    'Peng_T8': 'Uncinated Process',\n",
    "    'Peng_T9': 'Head',\n",
    "    'Peng_T10': 'Uncinated Process',\n",
    "    'Peng_T11': 'Body/Tail',\n",
    "    'Peng_T12': 'Uncinated Process',\n",
    "    'Peng_T13': 'Head',\n",
    "    'Peng_T14': 'Head',\n",
    "    'Peng_T15': 'Head',\n",
    "    'Peng_T16': 'Body',\n",
    "    'Peng_T17': 'Body/Tail',\n",
    "    'Peng_T18': 'Body',\n",
    "    'Peng_T19': 'Head',\n",
    "    'Peng_T20': 'Head',\n",
    "    'Peng_T21': 'Head',\n",
    "    'Peng_T22': 'Body',\n",
    "    'Peng_T23': 'Head',\n",
    "    'Peng_T24': 'Head',\n",
    "    'Peng_N1': 'Tail',\n",
    "    'Peng_N2': 'Descending Duodenum',\n",
    "    'Peng_N3': 'Descending Duodenum',\n",
    "    'Peng_N4': 'Body/Tail',\n",
    "    'Peng_N5': 'Body/Tail',\n",
    "    'Peng_N6': 'Body',\n",
    "    'Peng_N7': 'Tail',\n",
    "    'Peng_N8': 'Body/Tail',\n",
    "    'Peng_N9': 'Tail',\n",
    "    'Peng_N10': 'Common Bile Duct',\n",
    "    'Peng_N11': 'Body'\n",
    "}\n",
    "biological_metadata['Location'] = biological_metadata['ID_harmonised'].map(peng_location_mapping).fillna(biological_metadata.Location)\n",
    "biological_metadata.groupby(['Dataset', 'Location']).size().unstack().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 419,
   "id": "a3605f2f-2e5c-4cae-95cc-b5551db58412",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/3735475906.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata.groupby(['Dataset', 'BMI']).size().unstack()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>BMI</th>\n",
       "      <th>23.5</th>\n",
       "      <th>25.0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dataset</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Caronni</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ding_snRNA-seq</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lee</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lin</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Peng_Normal</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hwang</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Schlesinger</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Simeone</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele</th>\n",
       "      <td>35928</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Steele_Adj_Norm</th>\n",
       "      <td>0</td>\n",
       "      <td>6331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "BMI               23.5  25.0\n",
       "Dataset                     \n",
       "Caronni              0     0\n",
       "Ding                 0     0\n",
       "Ding_snRNA-seq       0     0\n",
       "Lee                  0     0\n",
       "Lin                  0     0\n",
       "Peng                 0     0\n",
       "Peng_Normal          0     0\n",
       "Hwang                0     0\n",
       "Schlesinger          0     0\n",
       "Simeone              0     0\n",
       "Steele           35928     0\n",
       "Steele_Adj_Norm      0  6331"
      ]
     },
     "execution_count": 419,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.groupby(['Dataset', 'BMI']).size().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 428,
   "id": "85f37523-c60a-4d7f-89de-dcc92749337a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/2166407596.py:13: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  unique_values_per_dataset = non_unknown_data.groupby('Dataset')[col].nunique()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dataset: 12/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "ID_batch_covariate: 12/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "ID_harmonised: 12/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Unique_ID: 12/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Level_1_refined: 12/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Tissue: 12/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Age: 10/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Treatment: 9/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Sex: 8/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "TNM_classification: 6/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Harmonized_TNM_Classification: 6/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Harmonized_Stage: 6/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Location: 5/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Grade: 5/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Stage: 5/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Harmonized_Grade: 5/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Pathology_notes: 4/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Vital_status: 3/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Overall Survival: 3/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Procedure: 3/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Diabetes: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "KRAS_mutation: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "TP53_type: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "SMAD4_type: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "CDKN2A_type: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Time_from_diagnosis_to_surgery_(weeks): 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Cycles: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Size_(cm): 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Pancreatitis: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Previous_cancer_history: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Other_co-morbidities: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Smoker: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Alcohol: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Lymph_node_met: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Survival_post_surgery_(weeks): 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "BMI: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "perineural invasion: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "vascular invasion: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "peripancreatic infiltarion: 2/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "KRAS_type: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Germline_ATM: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Germline_BRCA2: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Germline_FANCC: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Moffitt_Subtype: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Bailey_Subtype: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n",
      "Collisson_Subtype: 1/12 datasets\n",
      "Datasets: Caronni, Ding, Ding_snRNA-seq, Lee, Lin, Peng, Peng_Normal, Hwang, Schlesinger, Simeone, Steele, Steele_Adj_Norm\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Initialize a dictionary to store counts and the datasets for each column\n",
    "counts = {}\n",
    "datasets_for_columns = {}\n",
    "\n",
    "# Loop through columns in the biological_metadata DataFrame\n",
    "for col in biological_metadata.columns:\n",
    "    # Exclude 'Unknown' values and get the datasets where values are not 'Unknown'\n",
    "    non_unknown_data = biological_metadata[biological_metadata[col] != 'Unknown']\n",
    "    \n",
    "    # Group by 'Dataset' and count unique values excluding 'Unknown'\n",
    "    unique_values_per_dataset = non_unknown_data.groupby('Dataset')[col].nunique()\n",
    "\n",
    "    # Store the number of non-'Unknown' values in each dataset\n",
    "    counts[col] = np.count_nonzero(unique_values_per_dataset)\n",
    "    \n",
    "    # Get the datasets where the value is not 'Unknown'\n",
    "    datasets_for_columns[col] = unique_values_per_dataset.index.tolist()\n",
    "\n",
    "# Sort counts in descending order\n",
    "sorted_counts = sorted(counts.items(), key=lambda item: item[1], reverse=True)\n",
    "\n",
    "# Print the results\n",
    "for col, count in sorted_counts:\n",
    "    print(f'{col}: {count}/{len(biological_metadata[\"Dataset\"].unique())} datasets')\n",
    "    print(f'Datasets: {\", \".join(datasets_for_columns[col])}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "id": "255b8237-6e5c-43bd-83b3-f7d30f5cf435",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['biological_metadata'] = biological_metadata.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "id": "4a8a1f57-450f-407b-8ca4-d6fba5405654",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['biological_metadata'] = adata.obsm['biological_metadata'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "id": "80a586aa-52dc-4200-b555-9f71aec7fb7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('../Binned_Data/adata_scanvi_neighbours_leiden_biological_metadata.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "id": "6f424008-f5cb-4353-ba52-0aedeacdbbcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Ding',\n",
       " 'Lee',\n",
       " 'Simeone',\n",
       " 'Steele',\n",
       " 'Peng',\n",
       " 'Caronni',\n",
       " 'Lin',\n",
       " 'Schlesinger',\n",
       " 'Steele_Adj_Norm',\n",
       " 'Peng_Normal',\n",
       " 'Hwang',\n",
       " 'Ding_snRNA-seq']"
      ]
     },
     "execution_count": 435,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.Dataset.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "id": "9e9f513d-2009-4d67-a92e-f7b9fc3696d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Dataset', 'ID_batch_covariate', 'ID_harmonised', 'Unique_ID',\n",
       "       'Level_1_refined', 'Tissue', 'Age', 'Sex', 'Diabetes', 'Treatment',\n",
       "       'KRAS_mutation', 'KRAS_type', 'TP53_type', 'SMAD4_type', 'CDKN2A_type',\n",
       "       'Germline_ATM', 'Germline_BRCA2', 'Germline_FANCC', 'Moffitt_Subtype',\n",
       "       'Bailey_Subtype', 'Collisson_Subtype',\n",
       "       'Time_from_diagnosis_to_surgery_(weeks)', 'Cycles', 'Location', 'Grade',\n",
       "       'Size_(cm)', 'Stage', 'Pancreatitis', 'Previous_cancer_history',\n",
       "       'Other_co-morbidities', 'Smoker', 'Alcohol', 'Pathology_notes',\n",
       "       'Lymph_node_met', 'Survival_post_surgery_(weeks)', 'Vital_status',\n",
       "       'TNM_classification', 'Overall Survival', 'Procedure', 'BMI',\n",
       "       'perineural invasion', 'vascular invasion',\n",
       "       'peripancreatic infiltarion', 'Harmonized_Grade',\n",
       "       'Harmonized_TNM_Classification', 'Harmonized_Stage'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 438,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "id": "c14579a0-0029-4817-b178-9cab5af2cbc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "id": "de3539c5-cfdb-4836-85f8-e3e51e79e98c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/4133086291.py:1: FutureWarning: The behavior of Series.replace (and DataFrame.replace) with CategoricalDtype is deprecated. In a future version, replace will only be used for cases that preserve the categories. To change the categories, use ser.cat.rename_categories instead.\n",
      "  biological_metadata_only_na = biological_metadata.replace('Unknown', np.nan)\n"
     ]
    }
   ],
   "source": [
    "biological_metadata_only_na = biological_metadata.replace('Unknown', np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 471,
   "id": "290f5626-97ae-4b12-ac21-7019c7e50cc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/1092311295.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  df = biological_metadata_only_na.groupby('Dataset').nunique() #.apply(lambda x: x.notna().nunique())\n"
     ]
    }
   ],
   "source": [
    "df = biological_metadata_only_na.groupby('Dataset').nunique() #.apply(lambda x: x.notna().nunique())\n",
    "counts = {}\n",
    "for col in df:\n",
    "    counts[col] = df[df[col] > 0].index.tolist() #np.count_nonzero(df[col])  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 483,
   "id": "df1a523a-e10b-4154-8e9c-8176cfc46135",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;32mID_batch_covariate\u001b[0m: 12/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Schlesinger', 'Simeone', 'Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mID_harmonised\u001b[0m: 12/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Schlesinger', 'Simeone', 'Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mUnique_ID\u001b[0m: 12/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Schlesinger', 'Simeone', 'Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mLevel_1_refined\u001b[0m: 12/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Schlesinger', 'Simeone', 'Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mTissue\u001b[0m: 12/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Schlesinger', 'Simeone', 'Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mAge\u001b[0m: 10/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Simeone', 'Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mSex\u001b[0m: 8/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lin', 'Peng', 'Peng_Normal', 'Hwang', 'Simeone']\n",
      "\u001b[1;32mDiabetes\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Peng']\n",
      "\u001b[1;32mTreatment\u001b[0m: 9/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lee', 'Lin', 'Peng', 'Hwang', 'Simeone', 'Steele']\n",
      "\u001b[1;32mKRAS_mutation\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Simeone']\n",
      "\u001b[1;32mKRAS_type\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mTP53_type\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Simeone']\n",
      "\u001b[1;32mSMAD4_type\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Simeone']\n",
      "\u001b[1;32mCDKN2A_type\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Simeone']\n",
      "\u001b[1;32mGermline_ATM\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mGermline_BRCA2\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mGermline_FANCC\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mMoffitt_Subtype\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mBailey_Subtype\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mCollisson_Subtype\u001b[0m: 1/12\n",
      "Datasets present: ['Ding']\n",
      "\u001b[1;32mTime_from_diagnosis_to_surgery_(weeks)\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mCycles\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mLocation\u001b[0m: 5/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Peng', 'Peng_Normal']\n",
      "\u001b[1;32mGrade\u001b[0m: 5/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lin', 'Hwang']\n",
      "\u001b[1;32mSize_(cm)\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mStage\u001b[0m: 5/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq', 'Lin', 'Peng', 'Peng_Normal']\n",
      "\u001b[1;32mPancreatitis\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mPrevious_cancer_history\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mOther_co-morbidities\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mSmoker\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mAlcohol\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mPathology_notes\u001b[0m: 4/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq', 'Peng', 'Peng_Normal']\n",
      "\u001b[1;32mLymph_node_met\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mSurvival_post_surgery_(weeks)\u001b[0m: 2/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq']\n",
      "\u001b[1;32mVital_status\u001b[0m: 3/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq', 'Hwang']\n",
      "\u001b[1;32mTNM_classification\u001b[0m: 6/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Peng', 'Peng_Normal', 'Hwang']\n",
      "\u001b[1;32mOverall Survival\u001b[0m: 3/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq', 'Hwang']\n",
      "\u001b[1;32mProcedure\u001b[0m: 3/12\n",
      "Datasets present: ['Peng', 'Peng_Normal', 'Simeone']\n",
      "\u001b[1;32mBMI\u001b[0m: 2/12\n",
      "Datasets present: ['Steele', 'Steele_Adj_Norm']\n",
      "\u001b[1;32mperineural invasion\u001b[0m: 2/12\n",
      "Datasets present: ['Peng', 'Peng_Normal']\n",
      "\u001b[1;32mvascular invasion\u001b[0m: 2/12\n",
      "Datasets present: ['Peng', 'Peng_Normal']\n",
      "\u001b[1;32mperipancreatic infiltarion\u001b[0m: 2/12\n",
      "Datasets present: ['Peng', 'Peng_Normal']\n",
      "\u001b[1;32mHarmonized_Grade\u001b[0m: 5/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Lin', 'Hwang']\n",
      "\u001b[1;32mHarmonized_TNM_Classification\u001b[0m: 6/12\n",
      "Datasets present: ['Caronni', 'Ding', 'Ding_snRNA-seq', 'Peng', 'Peng_Normal', 'Hwang']\n",
      "\u001b[1;32mHarmonized_Stage\u001b[0m: 6/12\n",
      "Datasets present: ['Ding', 'Ding_snRNA-seq', 'Lin', 'Peng', 'Peng_Normal', 'Simeone']\n"
     ]
    }
   ],
   "source": [
    "# ANSI escape code for green text\n",
    "GREEN = '\\033[1;32m'\n",
    "RESET = '\\033[0m'\n",
    "\n",
    "sorted_counts = sorted(counts.items(), key=lambda item: len(item), reverse=False)\n",
    "\n",
    "for col, count in sorted_counts:\n",
    "    print(f'{GREEN}{col}{RESET}: {len(count)}/{len(df[col].index)}')\n",
    "    print(f'Datasets present: {count}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01e86bd4-38ee-4ea0-9658-8eb2b055f92f",
   "metadata": {},
   "outputs": [],
   "source": [
    "biological_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 493,
   "id": "ffe10d7c-faff-402f-8b2b-b6578ed3f991",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Patient_ID\n",
       "HT056P1                   [FOLFIRINOX]\n",
       "HT060P1                   [FOLFIRINOX]\n",
       "HT061P1                        [Naive]\n",
       "HT064P1                        [Naive]\n",
       "HT071P1                     [Chemo-RT]\n",
       "HT085P1                        [Naive]\n",
       "HT115P1                   [FOLFIRINOX]\n",
       "HT121P1                        [Naive]\n",
       "HT122P1                    [Gem_abrax]\n",
       "HT123P1                        [Naive]\n",
       "HT124P1                        [Naive]\n",
       "HT125P1                    [Gem_abrax]\n",
       "HT138P1                    [Gem_abrax]\n",
       "HT140P1                        [Naive]\n",
       "HT166P1                   [FOLFIRINOX]\n",
       "HT168P1                   [FOLFIRINOX]\n",
       "HT185P1    [FOLFIRINOX then Gem_abrax]\n",
       "HT190P1                    [Gem_abrax]\n",
       "HT191P1                   [FOLFIRINOX]\n",
       "HT200P1                   [FOLFIRINOX]\n",
       "HT204P1                   [FOLFIRINOX]\n",
       "HT224P1      [FOLIRINOX, Gem/Abraxane]\n",
       "HT231P1             [FOLFIRINOX, SBRT]\n",
       "HT232P1         [mFOLFIRINOX, FOLFIRI]\n",
       "HT242P1                          [nan]\n",
       "HT259P1                   [FOLFIRINOX]\n",
       "HT264P1                        [naive]\n",
       "HT270P1                        [naive]\n",
       "HT284P1                        [naive]\n",
       "HT288P1                        [Naive]\n",
       "HT306P1                   [FOLFIRINOX]\n",
       "HT412P1      [FOLFIRONOX, Gemcitabine]\n",
       "Name: Treatment, dtype: object"
      ]
     },
     "execution_count": 493,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ding.groupby('Patient_ID')['Treatment'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 490,
   "id": "73b2c67e-2322-4396-b0e4-06ba002f7a12",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_937282/396187640.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  biological_metadata[biological_metadata.Dataset.str.contains('Ding')].groupby('Unique_ID')['Treatment'].unique()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Unique_ID\n",
       "AdjNorm_TISSUE_1    []\n",
       "AdjNorm_TISSUE_2    []\n",
       "AdjNorm_TISSUE_3    []\n",
       "GSM4293555          []\n",
       "GSM4679532          []\n",
       "                    ..\n",
       "U5                  []\n",
       "U6                  []\n",
       "U7                  []\n",
       "U8                  []\n",
       "U9                  []\n",
       "Name: Treatment, Length: 149, dtype: object"
      ]
     },
     "execution_count": 490,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biological_metadata[biological_metadata.Dataset.str.contains('Ding')].groupby('Unique_ID')['Treatment'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 494,
   "id": "2e10809f-2957-4ded-976f-e77766a4a7a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 726054 × 2505\n",
       "    obs: 'Barcode', 'Dataset', 'ID_batch_covariate', 'Unique_ID', 'Technology', 'n_genes', 'n_counts', 'log_counts', 'mt_frac', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mito', 'log1p_total_counts_mito', 'pct_counts_mito', 'leiden', 'batch', 'leiden_0.2', 'leiden_0.2_annotation', 'leiden_subcluster', 'level0_leiden_subcluster', 'leiden_0.5', 'is_outlier_total_counts', 'outlier', 'infercnv_score_malignant', 'infercnv_score_malignant_refined', 'cnv_score_abs', 'treatment_status', 'Level_0', 'Level_1', 'Level_1_refined', 'MALAT1_lognorm', 'empty_droplet', 'ID_harmonised', 'Dataset_unique', 'Tissue', 'Age', 'Sex', 'Diabetes', 'Treatment', 'Acinar_leiden_0.75', 'Ambiguous_Epithelial_leiden_0.75', 'Intra_pancreatic_Neurons_leiden_0.75', 'Natural_Killer_leiden_0.75', 'Ambiguous_Immune_leiden_0.75', 'Neuronal_Cell_leiden_0.75', 'Schwann_Cell_leiden_0.75', 'Adipocyte_leiden_0.75', 'Erythroid_Cell_leiden_0.75', 'Smooth_Muscle_Cell_leiden_0.75', 'Ambiguous_Stromal_leiden_0.75', 'EMT_leiden_0.75', 'Pericyte_leiden_0.75', 'Endocrine_Cell_leiden_0.75', 'NK_Cell_leiden_0.75', 'B_Cell_leiden_0.75', 'Endothelial_Cell_leiden_0.75', 'Myeloid_Cell_leiden_0.75', 'Ductal_Cell_leiden_0.75', 'Malignant_leiden_0.75', 'Ductal_Cell_Malignant_leiden_0.75', 'Fibroblast_leiden_0.75', 'T_Cell_leiden_0.75'\n",
       "    var: 'n_cells', 'ensembl_id', 'start', 'end', 'chromosome', 'gene_name_adata_sc', 'highly_variable_adata_sc', 'means_adata_sc', 'dispersions_adata_sc', 'dispersions_norm_adata_sc', 'highly_variable_nbatches_adata_sc', 'highly_variable_intersection_adata_sc', 'n_cells_by_counts_adata_sc', 'mean_counts_adata_sc', 'log1p_mean_counts_adata_sc', 'pct_dropout_by_counts_adata_sc', 'total_counts_adata_sc', 'log1p_total_counts_adata_sc', 'mito_adata_sc', 'n_cells_by_counts_adata_sn', 'mean_counts_adata_sn', 'log1p_mean_counts_adata_sn', 'pct_dropout_by_counts_adata_sn', 'total_counts_adata_sn', 'log1p_total_counts_adata_sn', 'Manual_Genes'\n",
       "    uns: 'Condition_colors', 'Dataset_colors', 'Level_0_colors', 'Level_1_colors', 'Level_1_refined_colors', 'Technology_colors', 'leiden', 'neighbors', 'pca', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'drvi_emb', 'drvi_umap', 'expimap_emb', 'expimap_umap', 'scanvi_emb', 'scanvi_umap', 'scpoli_emb', 'scpoli_umap', 'scvi_emb', 'scvi_umap', 'sysvi_emb', 'sysvi_umap', 'biological_metadata'\n",
       "    varm: 'PCs'\n",
       "    layers: 'binned_data', 'counts', 'log_norm', 'raw'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 494,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64b4beab-d240-4463-a02b-4f4b27f43df6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scanpy",
   "language": "python",
   "name": "scanpy"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
